<!DOCTYPE html>
<html lang="en-CA"><head>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.6.39">

  <meta name="author" content="Alice Cima, Rachel Lobay, Daniel McDonald, Ryan Tibshirani">
  <title>InsightNet EpiData Workshop 2024 – day1-morning</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-6786ca87b8fc7b8b5f9e1cb937bb1622.css">
  <link rel="stylesheet" href="tachyons-minimal.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">


<section class="slide level2">

<div class="flex">
<div class="w-20">

</div>
<section id="section" class="slide level2 w-80" data-background-image="gfx/cover-art-1.svg" data-background-position="bottom">
<h2>Signal Discovery, Data Fetching, &amp; Panel Data</h2>
<h3 id="section-1">InsightNet Forecasting Workshop 2024</h3>
<p><br></p>
<h4 id="section-2">Alice Cima, Rachel Lobay, Daniel McDonald, Ryan Tibshirani</h4>
<p><span class="fstyle">with thanks to Delphi Tooling &amp; Forecasting Team: Logan Brooks, Nat DeFries, Dmitry Shemetov, David Webber</span></p>
<p>11 December – Morning</p>
</section>
</div>
</section>
<section id="outline" class="slide level2">
<h2>Outline</h2>
<ol type="1">
<li><p>Delphi</p></li>
<li><p>Overview of this Workshop</p></li>
<li><p>Panel Data</p></li>
<li><p>Versioned Data</p></li>
<li><p>Epidata API</p></li>
<li><p>Epidatr</p></li>
<li><p>Find Data Sources and Signals</p></li>
<li><p>Versioning in epidatr</p></li>
</ol>
</section>
<section>
<section id="delphi" class="title-slide slide level1 center" data-number="1">
<h1><span class="header-section-number">1</span> Delphi</h1>

</section>
<section id="about-delphi" class="slide level2">
<h2>About Delphi</h2>
<ul>
<li><p>Founded in 2012 at Carnegie Mellon University, now expanded to UC Berkeley, and University of British Columbia.</p></li>
<li><p>Currently 5 faculty, ~10 PhD students, ~15 staff (mostly software engineers).</p></li>
<li><p>Easy to join us from anywhere (lots of volunteers during Covid-19 pandemic).</p></li>
<li><p>We are:</p>
<ul>
<li>CDC Center of Excellence for Influenza and Covid-19 Forecasting (2019-24).</li>
<li>CDC Innovation Center for Outbreak Analytics and Disease Modeling (2024-29).</li>
</ul></li>
</ul>
<p><span class="primary"><strong>Our mission:</strong></span> To develop the theory and practice of <span class="primary">epidemic detection, tracking and forecasting</span>, and their use in decision making, both public and private.</p>
</section>
<section id="what-does-delphi-do" class="slide level2">
<h2>What does Delphi do?</h2>
<ul>
<li><p>Procure <span class="primary">real-time, aggregated data streams</span> informative of infectious diseases and syndromes, in collaboration with partners in industry and government.</p></li>
<li><p>Extract signals and make them widely available via the <span class="primary">Epidata platform &amp; API</span>.</p></li>
<li><p>Develop and deploy algorithms for <span class="primary">epidemic detection, tracking, forecasting</span>.</p></li>
<li><p>Develop and maintain statistical software packages for these tasks.</p></li>
<li><p>Make it all production-grade, maximally-accessible, and open-source (to serve CDC, state and local public health agencies, epi-forecasting researchers, data journalists, the public)</p></li>
</ul>
</section>
<section id="what-we-provide" class="slide level2">
<h2>What we provide</h2>

<img data-src="gfx/web_of_parts.svg" class="quarto-figure quarto-figure-center r-stretch"></section></section>
<section>
<section id="workshop-overview" class="title-slide slide level1 center" data-number="2">
<h1><span class="header-section-number">2</span> Workshop Overview</h1>

</section>
<section id="a-few-slides-here" class="slide level2">
<h2>A few slides here</h2>
</section>
<section id="a-few-slides-here-1" class="slide level2">
<h2>A few slides here</h2>
</section></section>
<section>
<section id="panel-data" class="title-slide slide level1 center" data-number="3">
<h1><span class="header-section-number">3</span> Panel Data</h1>

</section>
<section id="panel-data-1" class="slide level2">
<h2>Panel data</h2>
<ul>
<li><p><a href="https://en.wikipedia.org/wiki/Panel_data"><strong>Panel data</strong></a> or longitudinal data, contain cross-sectional measurements of subjects over time.</p></li>
<li><p>Since we’re working with aggregated data, the subjects are geographic units (e.g.&nbsp;counties, states)</p></li>
</ul>
<!-- and not individuals.-->
<ul>
<li><p>In table form, panel data is a time index + one or more locations/keys.</p></li>
<li><p><strong>Ex</strong>: The % of outpatient doctor visits that are COVID-related in WA from Dec.&nbsp;2021 to Feb.&nbsp;2022 (<a href="https://cmu-delphi.github.io/delphi-epidata/api/covidcast-signals/doctor-visits.html">docs</a>):</p></li>
</ul>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/panel-wa-ex_47fb13821011c17890705d9aa13f7ca2">
<div class="cell-output cell-output-stdout">
<pre><code>An `epi_df` object, 6 x 3 with metadata:
* geo_type  = state
* time_type = day
* as_of     = 2022-02-01

# A tibble: 6 × 3
  geo_value time_value percent_cli
* &lt;chr&gt;     &lt;date&gt;           &lt;dbl&gt;
1 wa        2021-12-01        4.70
2 wa        2021-12-02        4.60
3 wa        2021-12-03        4.56
4 wa        2021-12-04        4.93
5 wa        2021-12-05        4.17
6 wa        2021-12-06        4.12</code></pre>
</div>
</div>
<!-- Example: The estimated % of outpatient doctor visits that are COVID-related in WA from Dec. 2021 to Feb. 2022 -->
</section>
<section id="examples-of-panel-data---covid-19-cases" class="slide level2">
<h2>Examples of panel data - COVID-19 cases</h2>
<p><a href="https://cmu-delphi.github.io/delphi-epidata/api/covidcast-signals/jhu-csse.html"><span class="primary"><strong>JHU CSSE COVID cases per 100k </strong></span></a> estimates the daily number of new confirmed COVID-19 cases per 100,000 population, averaged over the past 7 days.</p>

<img data-src="day1-morning_files/figure-revealjs/examples-panel-covid2-1.svg" class="quarto-figure quarto-figure-center r-stretch"><aside class="notes">
<ul>
<li>WA switch to weekly reporting in 2022</li>
<li>FL reports “whenever” (weekly, biweekly, three days in a row, then 4 zeros, etc.)</li>
<li>API calculates change from cumulative, so no-report becomes a 0.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<!--

## Examples of panel data - CHNG-CLI

[[**Change Healthcare COVID-like illness (CHNG-CLI)**]{.primary}](https://cmu-delphi.github.io/delphi-epidata/api/covidcast-signals/chng.html) reports the percentage of outpatient visits for COVID-related symptoms, based on deidentified Change Healthcare claims data.


::: {.cell layout-align="center" hash='day1-morning_cache/revealjs/examples-chng-cli_f72ac8448ddf46e8b121a8b7a820d176'}
::: {.cell-output-display}
![](day1-morning_files/figure-revealjs/examples-chng-cli-1.svg){fig-align='center'}
:::
:::


## Examples of panel data - CHNG-COVID

[[**Change Healthcare COVID (CHNG-COVID)**]{.primary}](https://cmu-delphi.github.io/delphi-epidata/api/covidcast-signals/chng.html) reports the percentage of outpatient visits with confirmed COVID-19, based on Change Healthcare claims data.


::: {.cell layout-align="center" hash='day1-morning_cache/revealjs/examples-chng-covid_a73599bb55f5b57b76d646cb2e3e4251'}
::: {.cell-output-display}
![](day1-morning_files/figure-revealjs/examples-chng-covid-1.svg){fig-align='center'}
:::
:::


-->
<!--  Numerator = denote the Covid counts. Denominator be the total count of visits. Scaling by population is not necessary here because the signal is already normalized to the total number of visits, which acts as a proxy for population size. -->
</section>
<section id="examples-of-panel-data---hhs-admissions" class="slide level2">
<h2>Examples of panel data - HHS Admissions</h2>
<p><a href="https://cmu-delphi.github.io/delphi-epidata/api/covidcast-signals/hhs.html"><span class="primary"><strong>Confirmed COVID-19 Hospital Admissions per 100k</strong></span></a> estimates the daily sum of adult and pediatric confirmed COVID-19 hospital admissions, per 100,000 population, averaged over the past 7 days.</p>

<!-- 
## All together - Visualizing multiple panel data signals
Example: gathering different signals + visualizing panel data


::: {.cell layout-align="center" hash='day1-morning_cache/revealjs/multiple-signals-unscale_ae296b12fcf0939fb361e9ad6d4764ee'}

:::


## All together - Visualizing multiple panel data signals
Example: gathering different signals + **scaling** + visualizing panel data


::: {.cell layout-align="center" hash='day1-morning_cache/revealjs/multiple-signals-scale_8465f84e03a37f63c84fb30213c125d9'}

:::

::: {.cell layout-align="center" hash='day1-morning_cache/revealjs/scaled-signals_5d19ba1a9aba742e5994527f81d17ee7'}

:::

<div style="text-align: center;">
<small>[Figure 1 from Reinhart et al. (2021)](https://www.pnas.org/doi/10.1073/pnas.2111452118)</small>
</div>

**Takeaway:** The auxiliary signals track changes in the official reported cases quite well. This is clearer when they have all been placed on the same range as reported cases per 100,000 people.
-->
<img data-src="day1-morning_files/figure-revealjs/examples-hhs-admissions-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="examples-of-panel-data---covid-19-cases-and-deaths-in-ca" class="slide level2">
<h2>Examples of panel data - COVID-19 cases and deaths in CA</h2>
<ul>
<li><span class="primary"><strong>Takeaway</strong></span>: Cases appear to strongly correlate with deaths several weeks later.</li>
<li>We’ll see this again in an upcoming session…</li>
</ul>

<img data-src="day1-morning_files/figure-revealjs/plot-ca-cases-deaths-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section></section>
<section>
<section id="versioned-data" class="title-slide slide level1 center" data-number="4">
<h1><span class="header-section-number">4</span> Versioned Data</h1>

</section>
<section id="intro-to-versioned-data" class="slide level2">
<h2>Intro to versioned data</h2>
<ul>
<li><p>In panel data, we’ve seen that time is indicated by <code>time_value</code>.</p></li>
<li><p>Now, we add a second time index to indicate the data version…</p></li>
</ul>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/versioned-wa-ex_9e292916cbb4b3a0b7ed34c1043ca536">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  geo_value version    time_value percent_cli
  &lt;chr&gt;     &lt;date&gt;     &lt;date&gt;           &lt;dbl&gt;
1 wa        2021-12-05 2021-12-01       0.884
2 wa        2021-12-06 2021-12-01       0.917
3 wa        2021-12-07 2021-12-01       0.896
4 wa        2021-12-08 2021-12-01       0.985
5 wa        2021-12-09 2021-12-01       1.03 
6 wa        2021-12-10 2021-12-01       1.00 </code></pre>
</div>
</div>
<ul>
<li>Note that this feature can be indicated in different ways (ex. <code>version</code>, <code>issue</code>, <code>release</code>, <code>as_of</code>).</li>
</ul>
</section>
<section id="versioned-panel-data" class="slide level2">
<h2>Versioned panel data</h2>
<p>Estimated percentage of outpatient visits due to CLI in <span class="primary">FL</span> across multiple issue dates, with updates and revisions to past data as new issue dates are released:</p>

<img data-src="day1-morning_files/figure-revealjs/versioned-panel-multi-states-ex-2-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="latency-and-revision-in-signals" class="slide level2">
<h2>Latency and revision in signals</h2>
<ul>
<li><span class="primary"><strong>Latency</strong></span> refers to the delay between data collection and availability.</li>
</ul>
<p><strong>Example</strong>: A signal based on medical insurance claims may take several days to appear but is subject to delays as claims are processed over weeks.</p>
<ul>
<li><span class="primary"><strong>Revision</strong></span> occurs when data is updated or corrected after initial publication, due to late reporting or new information.</li>
</ul>
<p><strong>Example</strong>: COVID-19 case reports are revised frequently after initial publication as new data comes in or reporting backlogs are cleared.</p>
</section>
<section id="latency-and-revision-in-signals---example" class="slide level2">
<h2>Latency and revision in signals - Example</h2>
<ul>
<li><p>Recall the first example of panel &amp; versioned data we’ve seen…</p></li>
<li><p>This signal is 4 days <span class="primary"><strong>latent</strong></span> (min(<code>version</code> - <code>time_value</code>))</p></li>
</ul>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/latency-wa-ex_3504b5a9e38bbe3f65e2667383a5326a">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  geo_value version    time_value percent_cli version_time_diff
  &lt;chr&gt;     &lt;date&gt;     &lt;date&gt;           &lt;dbl&gt; &lt;drtn&gt;           
1 wa        2021-12-05 2021-12-01       0.884 4 days           
2 wa        2021-12-06 2021-12-02       0.737 4 days           
3 wa        2021-12-07 2021-12-03       0.662 4 days           
4 wa        2021-12-08 2021-12-04       0.663 4 days           
5 wa        2021-12-09 2021-12-05       0.872 4 days           
6 wa        2021-12-10 2021-12-06       0.642 4 days           </code></pre>
</div>
</div>
<ul>
<li>And clearly undergoes <span class="primary"><strong>revision</strong></span> over time (ex. consider Dec.&nbsp;1’s <code>percent_cli</code> across <code>version</code>):</li>
</ul>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/revision-wa-ex_37f042cea454fda9a187d86b78ad7091">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  geo_value time_value version    percent_cli version_time_diff
  &lt;chr&gt;     &lt;date&gt;     &lt;date&gt;           &lt;dbl&gt; &lt;drtn&gt;           
1 wa        2021-12-01 2021-12-05       0.884 4 days           
2 wa        2021-12-01 2021-12-06       0.917 5 days           
3 wa        2021-12-01 2021-12-07       0.896 6 days           
4 wa        2021-12-01 2021-12-08       0.985 7 days           
5 wa        2021-12-01 2021-12-09       1.03  8 days           
6 wa        2021-12-01 2021-12-10       1.00  9 days           </code></pre>
</div>
</div>
<!-- min_lag: the minimum time to any value min(as.integer(version) - as.integer(time_value)  -->
<!-- max_lag: the amount of time until the final (new) version -->
<!-- revision_summary computes some basic statistics about the revision behavior of an archive, returning a tibble summarizing the revisions per time_value+epi_key features. -->
</section>
<section id="revision-triangle-outpatient-visits-in-wa-2022" class="slide level2">
<h2>Revision triangle, Outpatient visits in WA 2022</h2>
<ul>
<li>7-day trailing average to smooth day-of-week effects</li>
</ul>

<img data-src="day1-morning_files/figure-revealjs/revision-triangle-1.png" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="revisions" class="slide level2">
<h2>Revisions</h2>
<p>Many data sources are subject to revisions:</p>
<ul>
<li><p>Case and death counts are frequently corrected or adjusted by authorities.</p></li>
<li><p>Medical claims data can take weeks to be submitted and processed.</p></li>
</ul>
<!--* In the previous slide's example because doctor's visits may be reported to the health system partners several days after they occur, these signals are typically available with several days of lag. This means that estimates for a specific day are only available several days later. -->
<ul>
<li><p>Lab tests and medical records can be backlogged for a variety of reasons.</p></li>
<li><p>Surveys are not always completed promptly.</p></li>
<li><p><span class="primary"><strong>Key</strong></span>: An accurate revision log is crucial for researchers building forecasts.</p></li>
</ul>
<p><strong>A forecast that is made today can should rely on information we have access to today.</strong></p>
</section>
<section id="three-types-of-revisions" class="slide level2">
<h2>Three types of revisions</h2>
<ol type="1">
<li><p><span class="primary"><strong>Sources that don’t revise</strong></span> (provisional and final are the same) - Facebook or Google symptoms</p></li>
<li><p><span class="secondary"><strong>Predictable revisions</strong></span> - Claims data (CHNG) and public health reports aligned by test, hospitalization, or death date</p></li>
<li><p><span class="tertiary"><strong>Revisions that are large and erratic to predict</strong></span> - COVID cases and deaths, especially when aligned by report date</p></li>
</ol>
<!-- (which can be highly variable & less predictable compared to test data). -->
</section>
<section id="types-of-revisions---comparison-between-2.-and-3." class="slide level2">
<h2>Types of revisions - Comparison between 2. and 3.</h2>
<ul>
<li>Revision behavior for two indicators in the HRR containing Charlotte, NC.</li>
</ul>
<!-- Each colored line corresponds to the data as reported on a particular date (as of dates varying from 28 September through 19 October). -->
<ul>
<li><p><span class="secondary"><strong>DV-CLI signal (left)</strong></span> was regularly revised throughout the period, although effects fade farther back.</p></li>
<li><p><span class="tertiary"><strong>JHU CSSE cases (right)</strong></span> remain “as reported” on Sept.&nbsp;28, with a spike toward the end of this period, until a major correction is made on Oct.&nbsp;19, which brings this down &amp; affects prior data.</p></li>
</ul>

<!--
<div style="text-align: center;">
<small>[Figure 1 from McDonald et al. (2021)](https://www.pnas.org/doi/pdf/10.1073/pnas.2111453118)</small>
</div>
-->
<img data-src="day1-morning_files/figure-revealjs/fig1-McDonald-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="key-takeaways" class="slide level2">
<h2>Key takeaways</h2>
<ul>
<li><span class="secondary"><strong>Medical claims revisions</strong>:</span> More systematic and predictable.</li>
<li><span class="tertiary"><strong>COVID-19 case report revisions</strong>:</span> Erratic and often unpredictable.</li>
<li>Large spikes or anomalies can occur as
<ul>
<li><span class="tertiary"><strong>Reporting backlogs</strong></span> are cleared.</li>
<li><span class="tertiary"><strong>Changes in case definitions</strong></span> are implemented.</li>
</ul></li>
</ul>
</section>
<section id="reporting-backlogs---example" class="slide level2">
<h2>Reporting backlogs - Example</h2>
<p>In Bexar County, Texas, during the summer of 2020…</p>
<ul>
<li><p><strong>Left</strong>: On July 16, 4,810 <span class="primary">backlogged cases</span> were reported, reflecting a 2-week delay. This caused a prolonged spike due to the 7-day trailing average applied to the counts.</p></li>
<li><p><strong>Right</strong>: CTIS estimates of CLI-in-community showed more stable underlying trends.</p></li>
</ul>

<img data-src="day1-morning_files/figure-revealjs/fig4-Reinhart-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="reporting-backlogs---key-takeaways" class="slide level2">
<h2>Reporting backlogs - Key takeaways</h2>
<ul>
<li><p><span class="primary"><strong>Reporting issues been common across U.S. jurisdictions</strong></span>.</p></li>
<li><p>For example, audits have regularly discovered <span class="primary"><strong>misclassified or unreported cases and deaths</strong></span>.</p></li>
<li><p>This underscores <span class="primary"><strong>cross-checking data with external sources</strong></span> not part of the same reporting systems.</p></li>
</ul>
<!-- This underscores the value of [**cross-checking data with external sources**]{.primary} not part of the same reporting systems. -->
</section></section>
<section>
<section id="epidata-api" class="title-slide slide level1 center" data-number="5">
<h1><span class="header-section-number">5</span> Epidata API</h1>

</section>
<section id="goals-of-delphi-epidata-platform-and-repository" class="slide level2">
<h2>Goals of Delphi Epidata platform and repository</h2>
<ol type="1">
<li><strong>Be the one-stop shop for aggregated epi-surveillance time-series (“epi-signals”)</strong>
<ul>
<li>Includes signals from other sources, especially if they don’t keep data revisions - E.g. CDC’s own NSSP, NWSS</li>
<li>Be the national historical repository of record &amp; preserve the raw data</li>
</ul></li>
</ol>
<!-- Hence: include also signals available elsewhere, especially if they don't keep data revisions - E.g. CDC's own NSSP, NWSS -->
<ol start="2" type="1">
<li><strong>Be the national clearinghouse for epi-signals, including those held elsewhere</strong>
<ul>
<li>The go-to place for signal discovery</li>
</ul></li>
<li><strong>Add value to existing signals and synthesize new ones</strong>
<ul>
<li>Synthesize new: via signal fusion, e.g.&nbsp;nowcasting</li>
</ul></li>
<li><strong>Be the focal point for community-wide efforts to open up privately held data</strong>
<ul>
<li>Better positioned than government or industry</li>
</ul></li>
</ol>
</section>
<section id="the-bigger-goal" class="slide level2">
<h2>The bigger goal</h2>
<p>The goal is to make epi-surveillance <span class="primary">more nimble, complete, standardized, robust, and real-time</span> and</p>
<p><span class="primary">less burdensome on the health system itself</span>.</p>
<p>Epidata is not the solution; but we hope it is a blueprint towards such a solution.</p>
</section>
<section id="what-is-the-epidata-repository" class="slide level2">
<h2>What is the Epidata repository</h2>
<p><span class="primary"><strong>Epidata</strong></span> is a repository of aggregated epi-surveillance time series. We aim to make everything <span class="primary">free and open-source</span>.</p>
<!-- To the full extent we can, we make everything [free and open-source]{.primary}. -->
<ul>
<li>To date, it has accumulated over 5 billion records.</li>
</ul>
<!-- (each record is the value of a signal, at a particular date, and a particular location). -->
<ul>
<li><p>At the peak of the pandemic, we were receiving millions of API queries per day.</p></li>
<li><p><span class="primary"><strong>Data comes from</strong></span>: public health reporting, medical insurance claims, medical device data, Google search queries, wastewater, app-based mobility patterns.</p></li>
<li><p>Many of our data streams simply aren’t available anywhere else.</p></li>
<li><p><span class="primary"><strong>Added value we provide</strong></span>: revision tracking, anomaly detection, trend detection, smoothing, imputation, geo-temporal-demographic disaggregation.</p></li>
</ul>
</section>
<section id="features-of-delphi-epidata" class="slide level2">
<h2>Features of Delphi Epidata</h2>
<ul>
<li><p>Built-in support for:</p>
<ul>
<li>Data revisions (“backfill”), including reporting dates and changes</li>
<li>Geo levels w/ auto-aggregation (e.g.&nbsp;county, state, and nation) and specialized levels (e.g., DMA, sewer sheds)</li>
<li>Demographic breakdown</li>
<li>Representation for missingness and censoring</li>
<li>Population sizes and fine-grained population density</li>
</ul></li>
<li><p>Pre-computed smoothing and normalization (customization planned)</p></li>
<li><p>Access control</p></li>
<li><p>Code is Open Source. Signals are as accessible (w/ API, SDK) as allowed by DUAs</p></li>
</ul>
<!-- MSA, HRR, state, HHS region, nation -->
<!-- 1st point: Concepts of "reporting date" and "as of". -->
<!--     + Backfill projection and alerting to changes in backfill dynamics -->
</section>
<section id="epidata-api-1" class="slide level2">
<h2>Epidata API</h2>
<ul>
<li><p><a href="https://cmu-delphi.github.io/delphi-epidata/">Delphi’s Epidata API</a> provides real-time access to epidemiological surveillance data.</p></li>
<li><p>The <a href="https://cmu-delphi.github.io/delphi-epidata/api/covidcast.html">main endpoint</a> (covidcast) providing daily updates about COVID-19 and influenza in the U.S.</p></li>
<li><p>A <a href="https://cmu-delphi.github.io/delphi-epidata/api/README.html">variety of other endpoints</a>, providing primarily historical data about various diseases including COVID-19, influenza, dengue fever, and norovirus in several countries.</p></li>
<li><p>A <a href="https://cmu-delphi.github.io/delphi-epidata/api/client_libraries.html">full-featured R client</a> is available for quick access to all data.</p></li>
<li><p>A <a href="https://cmu-delphi.github.io/delphi-epidata/api/client_libraries.html#python">Legacy Python client</a> is available, full-featured Python client in development.</p></li>
</ul>
<!-- The main endpoint (covidcast) providing daily updates about current COVID-19 and influenza activity across the United States. -->
</section>
<section id="some-of-our-data-sources" class="slide level2">
<h2>Some of our data sources</h2>
<h3 id="ongoing-sources">Ongoing Sources:</h3>
<ul>
<li><span class="primary"><strong>Insurance claims</strong></span>: %Covid {inpatient, outpatient}, by {county x day}</li>
<li><span class="primary"><strong>Google Symptom searches</strong></span>: 7 symptoms groups, by {county x day}</li>
<li><span class="primary"><strong>Quidel/Ortho antigen tests</strong></span>: %Covid by age group, by {county x day}</li>
<li><span class="primary"><strong>NCHS Deaths</strong></span>: all-cause, pneumonia, flu, Covid, by {state x week}</li>
<li><span class="primary"><strong>NSSP ED visits</strong></span>: %Covid, %flu, %RSV, by {county x week} (new!)</li>
<li><span class="primary"><strong>NWSS Covid</strong></span>, by {sampling-site x day} (in progress)</li>
</ul>
</section>
<section id="some-of-our-data-sources-1" class="slide level2">
<h2>Some of our data sources</h2>
<h3 id="active-during-pandemic-could-be-restarted-for-the-next-phe">Active during pandemic, could be restarted for the next PHE:</h3>
<ul>
<li><span class="primary"><strong>HHS Hosp/ICU beds</strong></span>: Covid, flu, by age-group, by {state x day}, {facility x week}</li>
<li><span class="primary"><strong>CTIS (“Delphi Facebook Survey”)</strong></span>: many dozens of questions, by (county x day)</li>
<li><span class="primary"><strong>STLT-reported</strong></span>: {cases, deaths} via {JHU, USAFacts}, by (country x day)</li>
<li><span class="primary"><strong>Safegraph mobility</strong></span>: misc measures by {county x day},{county x week}</li>
</ul>
</section>
<section id="some-of-our-pre-pandemic-data-sources" class="slide level2">
<h2>Some of our pre-pandemic data sources</h2>
<ul>
<li><p>FluView ILINet, by {state x week}</p></li>
<li><p>FluView Clinical (% positive flu, PH and clinical labs)</p></li>
<li><p>Google Health Trends (GHT), precursor to Google Symptoms</p></li>
<li><p>Google Flu Trends (GFT), precursor to to GHT</p></li>
<li><p>Twitter flu</p></li>
<li><p>Access counts for flu-related CDC pages, by {city x week}</p></li>
<li><p>Access counts for flu-related Wikipedia entries by {day x hour}</p></li>
<li><p>Flu-surv (flu hosp rates, now expanded to RESP-NET)</p></li>
<li><p>Misc signals for dengue, norovirus</p></li>
<li><p>Misc signals for PAHO countries, ECDC, KCDC, Taiwan,…</p></li>
<li><p>Delphi ILI nowcasts, by {state x week}, visualized in “ILI Nearby” website</p></li>
<li><p>Delphi ILI forecasts, by {state x week}</p></li>
</ul>
</section>
<section id="severity-pyramid" class="slide level2">
<h2>Severity pyramid</h2>

<img data-src="gfx/severity-pyramid.svg" class="quarto-figure quarto-figure-center r-stretch"></section></section>
<section>
<section id="epidatr" class="title-slide slide level1 center" data-number="6">
<h1><span class="header-section-number">6</span> Epidatr</h1>

</section>
<section id="installing-epidatr" class="slide level2">
<h2>Installing <code>epidatr</code></h2>
<p>Installing the package in R is straightforward…</p>
<p>Install the CRAN version</p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/install-epidatr-cran_72a247c54ab9e8a1857cca2c0c630e55">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a></a><span class="co"># Install the CRAN version</span></span>
<span id="cb5-2"><a></a>pak<span class="sc">::</span><span class="fu">pkg_install</span>(<span class="st">"epidatr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>or the development version</p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/install-epidatr-dev_33fa97f405ec0c1472fb4dcfa008ba77">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a></a><span class="co"># Install the development version from the GitHub dev branch</span></span>
<span id="cb6-2"><a></a>pak<span class="sc">::</span><span class="fu">pkg_install</span>(<span class="st">"cmu-delphi/epidatr@dev"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The CRAN listing is <a href="https://cran.r-project.org/package=epidatr/index.html">here</a>.</p>
</section>
<section id="python" class="slide level2">
<h2>Python</h2>
<p>In Python, install <a href="https://pypi.org/project/delphi-epidata/"><code>delphi-epidata</code> from PyPI</a> with</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode numberSource sh number-lines code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a></a><span class="ex">pip</span> install delphi-epidata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><br></p>
<p><code>delphi-epidata</code> is soon to be replaced with <code>epidatpy</code>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode numberSource sh number-lines code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a></a><span class="co"># Latest dev version</span></span>
<span id="cb8-2"><a></a><span class="ex">pip</span> install <span class="at">-e</span> <span class="st">"git+https://github.com/cmu-delphi/epidatpy.git#egg=epidatpy"</span></span>
<span id="cb8-3"><a></a></span>
<span id="cb8-4"><a></a><span class="co"># PyPI version (not yet available)</span></span>
<span id="cb8-5"><a></a><span class="ex">pip</span> install epidatpy</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="using-epidatr-and-epidatpy" class="slide level2">
<h2>Using <code>epidatr</code> and <code>epidatpy</code></h2>
<p>The following shows how to import the library and fetch all confirmed influenza hospital admissions occurring each day for North Carolina:</p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/hhs-influenza-pub-covidcast_3f3e40b3d02743a046273c33793dded1">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a></a><span class="fu">library</span>(epidatr)</span>
<span id="cb9-2"><a></a>hhs_flu_nc <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(<span class="st">'hhs'</span>, <span class="st">'confirmed_admissions_influenza_1d'</span>, <span class="st">'state'</span>, <span class="st">'day'</span>, <span class="at">geo_values =</span> <span class="st">'nc'</span>,</span>
<span id="cb9-3"><a></a>                     <span class="at">time_values =</span> <span class="fu">c</span>(<span class="dv">20240401</span>, <span class="dv">20240405</span><span class="sc">:</span><span class="dv">20240414</span>))</span>
<span id="cb9-4"><a></a><span class="fu">head</span>(hhs_flu_nc, <span class="at">n =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/hhs-influenza-pub-covidcast-2_9ae870dc24f6e71f32771c949e939fb2">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 15
  geo_value signal     source geo_type time_type time_value direction issue     
  &lt;chr&gt;     &lt;chr&gt;      &lt;chr&gt;  &lt;fct&gt;    &lt;fct&gt;     &lt;date&gt;         &lt;dbl&gt; &lt;date&gt;    
1 nc        confirmed… hhs    state    day       2024-04-01        NA 2024-04-22
2 nc        confirmed… hhs    state    day       2024-04-05        NA 2024-04-22
3 nc        confirmed… hhs    state    day       2024-04-06        NA 2024-04-22
# ℹ 7 more variables: lag &lt;dbl&gt;, missing_value &lt;dbl&gt;, missing_stderr &lt;dbl&gt;,
#   missing_sample_size &lt;dbl&gt;, value &lt;dbl&gt;, stderr &lt;dbl&gt;, sample_size &lt;dbl&gt;</code></pre>
</div>
</div>
<p><br></p>
<p>Python equivalent:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a></a>res <span class="op">=</span> Epidata.covidcast(<span class="st">'hhs'</span>, <span class="st">'confirmed_admissions_influenza_1d'</span>, <span class="st">'day'</span>, <span class="st">'state'</span>, [<span class="dv">20240401</span>, Epidata.<span class="bu">range</span>(<span class="dv">20240405</span>, <span class="dv">20240414</span>)], <span class="st">'nc'</span>)</span>
<span id="cb11-2"><a></a><span class="bu">print</span>(res[<span class="st">'result'</span>], res[<span class="st">'message'</span>], <span class="bu">len</span>(res[<span class="st">'epidata'</span>]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="api-keys" class="slide level2">
<h2>API keys</h2>
<ul>
<li><p>Anyone may access the Epidata API anonymously without providing any personal data.</p></li>
<li><p>Anonymous API access is subject to the following restrictions:</p>
<ol type="1">
<li>public datasets only; 2) rate limit: 60 requests per hour; 3) only two parameters may have multiple selections</li>
</ol></li>
<li><p>An API key grants priviledged access to the API and can be obtained by <a href="https://api.delphi.cmu.edu/epidata/admin/registration_form">registering with us</a>.</p></li>
<li><p>Privileges of registration:</p>
<ol type="1">
<li>no rate limit; 2) no limit on multiple selections</li>
</ol></li>
</ul>
<!-- rate limited to 60 requests per hour;  -->
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Tip</strong></p>
</div>
<div class="callout-content">
<p>The <code>epidatr</code> client automatically searches for the key in the <code>DELPHI_EPIDATA_KEY</code> environment variable. We recommend storing it in your <code>.Renviron</code> file, which R reads by default. More on setting your API key <a href="https://rdrr.io/cran/epidatr/man/get_api_key.html">here</a>.</p>
</div>
</div>
</div>
</section></section>
<section>
<section id="find-data-sources-signals" class="title-slide slide level1 center" data-number="7">
<h1><span class="header-section-number">7</span> Find Data Sources &amp; Signals</h1>

</section>
<section id="finding-data-sources-and-signals-of-interest" class="slide level2">
<h2>Finding data sources and signals of interest</h2>
<ul>
<li>Diverse Data Streams
<ul>
<li><span class="primary"><strong>Variety of Data</strong></span>: Access to medical claims data, cases and deaths, mobility data, and more.</li>
<li><span class="primary"><strong>Geographic Coverage</strong></span>: Includes multiple regions, making it comprehensive yet complex.</li>
<li><span class="primary"><strong>Challenge</strong></span>: Difficulty in pinpointing the specific data stream of interest.</li>
</ul></li>
<li>Using the Documentation
<ul>
<li><span class="primary"><strong>Comprehensive Listings</strong></span>: Documentation details all available data sources and signals for both <a href="https://cmu-delphi.github.io/delphi-epidata/api/covidcast_signals.html">COVID-19</a> and other <a href="https://cmu-delphi.github.io/delphi-epidata/api/README.html#source-specific-parameters">endpoints</a>.</li>
</ul></li>
<li>Docs are great for a deep dive into the data, while the apps &amp; tools are useful to see what’s available…</li>
</ul>
</section>
<section id="cheatsheet-of-web-based-apps-we-provide" class="slide level2">
<h2>Cheatsheet of web-based apps we provide</h2>
<p>We provide…</p>
<ul>
<li><p>A <a href="https://delphi.cmu.edu/signals/">signal discovery app</a>, to explore what epi-signals are available in Delphi Epidata and elsewhere in the community.</p></li>
<li><p>A <a href="https://delphi.cmu.edu/epivis/">general signal visualization tool</a>.</p></li>
<li><p>A <a href="https://delphi.cmu.edu/covidcast/">signal dashboard</a> and a “classic” <a href="https://delphi.cmu.edu/covidcast/classic/">map-based version</a> to visualize a core set of COVID-19 and flu indicators.</p></li>
<li><p>A <a href="https://delphi.cmu.edu/covidcast/export/">COVID-19 signal export app</a>, a <a href="https://delphi.cmu.edu/covidcast/dashboard/">dashboard builder</a>, and more!</p></li>
</ul>
</section>
<section id="signal-dashboard---for-covid-19-flu-data" class="slide level2">
<h2>Signal dashboard - For COVID-19 &amp; flu data</h2>
<div class="flex">
<div class="w-40">
<ul>
<li>The <a href="https://delphi.cmu.edu/covidcast/">signal dashboard</a> displays a selection of signals for COVID-19 &amp; flu.</li>
</ul>
<p><br></p>
<ul>
<li>Browse by location or indicator to choose which signal you are interested in &amp; then export the data for analysis.</li>
</ul>
<p><br></p>
<ul>
<li><strong>Example</strong>: Google symptom search trends in NC.</li>
</ul>
<!-- 
* Includes "Data Export" to pull a selected signal & download as a CSV. -->
</div>
<div class="w-60">
<p><a href="https://delphi.cmu.edu/covidcast/indicator/?date=20240426&amp;region=NC"> <img src="gfx/google_symptoms_signal_dash.jpg" alt="Signal Discover Screenshot"> </a></p>
</div>
</div>
</section>
<section id="signal-discovery-app---browse-for-more-data" class="slide level2">
<h2>Signal discovery app - Browse for more data</h2>
<ul>
<li><span class="primary"><strong>Signal discovery app</strong></span>: An easy way to find data sources and signals (no programming required).
<ul>
<li>Search tool that is a good to browse &amp; find data.</li>
</ul></li>
<li><a href="https://delphi.cmu.edu/signals/">Let’s try it out together!</a> <a href="https://delphi.cmu.edu/signals/"> <img src="gfx/signal_discover_screenshot.jpg" alt="Signal Discover Screenshot" style="height: auto; width: auto;"> </a></li>
</ul>
</section>
<section id="example---nchs-weekly-flu-mortality-data-in-states" class="slide level2">
<h2>Example - NCHS weekly flu mortality data in states</h2>
<p><a href="https://delphi.cmu.edu/signals/"> <img src="gfx/signal_discover_screenshot_ex.jpg" alt="Signal Discover Screenshot" style="height: auto; width: auto;"> </a></p>
<!-- To reproduce: Pathogen/Syndrom = flu, Severity pyramid = Dead (top of pyramid), Geo-level = ADM1 (e.g. U.S. states) -->
</section>
<section id="interactive-tooling-in-r" class="slide level2">
<h2>Interactive tooling in R</h2>
<ul>
<li><p>Moving from initial web-based browsing to programming, <span class="primary"><strong>how do we find sources and signals in R?</strong></span></p></li>
<li><p>Functions to enhance data discovery in <code>epidatr</code>:</p></li>
</ul>
<p><code>avail_endpoints()</code> Function: + Lists all endpoints with brief descriptions. + Highlights specific endpoints that cover non-US locations, facilitating targeted searches.</p>
<p><span class="primary"><strong>Output Format</strong></span>: Returns a tibble for easy viewing and analysis of available data sources.</p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/avail-endpoints-fun_a0d2bbe64a1d412c400965de2ffe3c52">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a></a><span class="fu">avail_endpoints</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="using-the-covidcast_epidata-function" class="slide level2">
<h2>Using the <code>covidcast_epidata()</code> function</h2>
<ul>
<li><strong>Function Overview</strong>: <code>covidcast_epidata()</code> offers detailed insights from the COVIDcast endpoint.</li>
</ul>
<!-- detailed insights into data sources from the COVIDcast endpoint. -->
<ul>
<li><p><strong>Source List</strong>: Each data source is listed in <code>covid_sources$sources</code>, with tibbles describing the included signals.</p></li>
<li><p><strong>Editor Support</strong>: In RStudio or similar editors, use tab completion to explore:</p>
<ul>
<li><p><strong>Data Sources</strong>: Type <code>covid_sources$source$</code> to view available data sources.</p></li>
<li><p><strong>Signals</strong>: Type <code>covid_sources$signals$</code> to see signal options with autocomplete assistance.</p></li>
</ul></li>
<li><p><strong>Filtering Convenience</strong>: Signal names are prefixed with their respective data source for easier navigation.</p></li>
</ul>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/covidcast-epidata-fun_24a6873628ca78d81851132e34413bdd">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a></a>covid_sources <span class="ot">&lt;-</span> <span class="fu">covidcast_epidata</span>()</span>
<span id="cb13-2"><a></a><span class="fu">head</span>(covid_sources<span class="sc">$</span>sources, <span class="at">n =</span> <span class="dv">2</span>) <span class="co"># head(list, n = 2) will print the first two elements of the list</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fetching-data---covidcast-main-endpoint" class="slide level2">
<h2>Fetching data - COVIDcast main endpoint</h2>
<ul>
<li><p>Fetching data from the Delphi Epidata API is simple.</p></li>
<li><p>The <code>pub_covidcast()</code> function lets us access the <code>covidcast</code> endpoint.</p></li>
<li><p>We need to specify the following six arguments…</p>
<ol type="1">
<li><code>source</code>: Data source name</li>
<li><code>signals</code>: Signal name</li>
<li><code>geo_type</code>: Geographic level</li>
<li><code>time_type</code>: Time resolution</li>
<li><code>geo_values</code>: Location(s)</li>
<li><code>time_values</code>: times of interest</li>
</ol></li>
<li><p>Let’s give this a try!</p></li>
</ul>
</section>
<section id="fetching-data---covidcast-main-endpoint-1" class="slide level2">
<h2>Fetching data - COVIDcast main endpoint</h2>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/us-jhu-pub-covidcast_53d509c560666401353437a2858b2d55">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a></a><span class="fu">library</span>(epidatr)</span>
<span id="cb14-2"><a></a><span class="fu">library</span>(dplyr)</span>
<span id="cb14-3"><a></a></span>
<span id="cb14-4"><a></a>jhu_us_cases <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb14-5"><a></a>  <span class="at">source =</span> <span class="st">"jhu-csse"</span>,</span>
<span id="cb14-6"><a></a>  <span class="at">signals =</span> <span class="st">"confirmed_7dav_incidence_prop"</span>, </span>
<span id="cb14-7"><a></a>  <span class="at">geo_type =</span> <span class="st">"nation"</span>,</span>
<span id="cb14-8"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb14-9"><a></a>  <span class="at">geo_values =</span> <span class="st">"us"</span>,</span>
<span id="cb14-10"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">20210101</span>, <span class="dv">20210401</span>)</span>
<span id="cb14-11"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/head-us-jhu-pub-covidcast_6863459436d0c740bf6875eddf05d459">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 8
  geo_value signal             source geo_type time_value issue        lag value
  &lt;chr&gt;     &lt;chr&gt;              &lt;chr&gt;  &lt;fct&gt;    &lt;date&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt;
1 us        confirmed_7dav_in… jhu-c… nation   2021-01-01 2023-03-10   798  61.9
2 us        confirmed_7dav_in… jhu-c… nation   2021-01-02 2023-03-10   797  64.2
3 us        confirmed_7dav_in… jhu-c… nation   2021-01-03 2023-03-10   796  67.1</code></pre>
</div>
</div>
<p>Here <code>value</code> is the requested signal – in this case, the number of daily new confirmed COVID-19 cases per 100,000 population from January to April 2021. <!-- (and standard error if it is applicable to the metric). --></p>
</section>
<section id="returned-data---covidcast-main-endpoint" class="slide level2">
<h2>Returned data - COVIDcast main endpoint</h2>
<ul>
<li><p><code>pub_covidcast()</code> outputs a tibble, where each row represents one observation.</p></li>
<li><p>Each observation covers a set of events aggregated by time and by geographic region is a record in our database. Each such record includes:</p>
<ul>
<li><code>time_value</code>: time period when the events occurred.</li>
<li><code>geo_value</code>: geographic region where the events occurred.</li>
<li><code>value</code>: estimated value.</li>
<li><code>stderr</code>: standard error of the estimate, usually referring to the sampling error.</li>
<li><code>sample_size</code>: number of events used in the estimation.</li>
</ul></li>
</ul>
<!--For example, a number of COVID-19 antigen tests were performed in the state of New York on August 1. The `time_value` would be August 1, with `geo_value` indicating the state of New York, while the remaining fields would give the estimated test positivity rate (the percentage of tests that were positive for COVID-19), its standard error, and the number of tests used to calculate the estimate.-->
</section>
<section id="returned-data---covidcast-main-endpoint-1" class="slide level2">
<h2>Returned data - COVIDcast main endpoint</h2>
<p>Crucially—and unlike most other COVID-19 data sources—our API reports two additional fields with each record:</p>
<ul>
<li><p><code>issue</code>: The time period when this observation was published.</p></li>
<li><p><code>lag</code>: The time delay between when the events occurred and when this observation was published.</p></li>
<li><p>Meaning that it tracks the complete revision history of the signal.</p></li>
<li><p>This allows for historical reconstructions of what information was available at specific times. <span class="primary"><strong>More on this soon!</strong></span></p></li>
</ul>
<!-- * Meaning that unlike most other sources of COVID data, it tracks the complete revision history of the signal. -->
</section>
<section id="geographic-levels" class="slide level2">
<h2>Geographic levels</h2>
<ul>
<li><p>The Epidata API makes signals available at different geographic levels, depending on the endpoint.</p></li>
<li><p>For the <code>confirmed_7dav_incidence_prop</code> signal, we can obtain values for each state.</p></li>
<li><p>Simply change <code>geo_type</code> and <code>geo_values</code> in the previous example to get…</p></li>
</ul>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/state-jhu-pub-covidcast_e2335c96b370b6bcc4a4a747b1b82ff8">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a></a><span class="co"># Obtain the most up-to-date version of the smoothed covid-like illness (CLI)</span></span>
<span id="cb16-2"><a></a><span class="co"># signal from the COVID-19 Trends and Impact survey for all states</span></span>
<span id="cb16-3"><a></a>jhu_state_cases <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb16-4"><a></a>  <span class="at">source =</span> <span class="st">"jhu-csse"</span>,</span>
<span id="cb16-5"><a></a>  <span class="at">signals =</span> <span class="st">"confirmed_7dav_incidence_prop"</span>,</span>
<span id="cb16-6"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb16-7"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb16-8"><a></a>  <span class="at">geo_values =</span> <span class="st">"*"</span>,</span>
<span id="cb16-9"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">20210101</span>, <span class="dv">20210401</span>)</span>
<span id="cb16-10"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/head-state-jhu-pub-covidcast_1ffe83212b7e3b0eb5a1f729c52b26bb">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  geo_value signal             source geo_type time_value issue        lag value
  &lt;chr&gt;     &lt;chr&gt;              &lt;chr&gt;  &lt;fct&gt;    &lt;date&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt;
1 ak        confirmed_7dav_in… jhu-c… state    2021-01-01 2023-03-03   791  35.9
2 al        confirmed_7dav_in… jhu-c… state    2021-01-01 2023-03-03   791  67.7
3 ar        confirmed_7dav_in… jhu-c… state    2021-01-01 2023-03-03   791  76.2
4 as        confirmed_7dav_in… jhu-c… state    2021-01-01 2023-03-03   791   0  
5 az        confirmed_7dav_in… jhu-c… state    2021-01-01 2023-03-03   791  83.4
6 ca        confirmed_7dav_in… jhu-c… state    2021-01-01 2023-03-10   798 104. </code></pre>
</div>
</div>
</section>
<section id="covidcast-main-endpoint---example-query" class="slide level2">
<h2>COVIDcast main endpoint - Example query</h2>
<p>County <code>geo_values</code> are <a href="https://en.wikipedia.org/wiki/List_of_United_States_FIPS_codes_by_county">FIPS codes</a> and are discussed in the API docs <a href="https://cmu-delphi.github.io/delphi-epidata/api/covidcast_geography.html">here</a>. The example below is for Orange County, California.</p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/county-jhu-pub-covid_04f08f2948439ccdd5f6831c14522aad">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a></a>jhu_county_cases <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb18-2"><a></a>  <span class="at">source =</span> <span class="st">"jhu-csse"</span>,</span>
<span id="cb18-3"><a></a>  <span class="at">signals =</span> <span class="st">"confirmed_7dav_incidence_prop"</span>,</span>
<span id="cb18-4"><a></a>  <span class="at">geo_type =</span> <span class="st">"county"</span>,</span>
<span id="cb18-5"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb18-6"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">20210101</span>, <span class="dv">20210401</span>),</span>
<span id="cb18-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"06059"</span></span>
<span id="cb18-8"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/head-county-jhu-pub-covid_89bd498070735785aa695ca40b4ba213">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  geo_value signal             source geo_type time_value issue        lag value
  &lt;chr&gt;     &lt;chr&gt;              &lt;chr&gt;  &lt;fct&gt;    &lt;date&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt;
1 06059     confirmed_7dav_in… jhu-c… county   2021-01-01 2023-03-03   791  105.
2 06059     confirmed_7dav_in… jhu-c… county   2021-01-02 2023-03-03   790  107.
3 06059     confirmed_7dav_in… jhu-c… county   2021-01-03 2023-03-03   789  108.
4 06059     confirmed_7dav_in… jhu-c… county   2021-01-04 2023-03-03   788  107.
5 06059     confirmed_7dav_in… jhu-c… county   2021-01-05 2023-03-03   787  105.
6 06059     confirmed_7dav_in… jhu-c… county   2021-01-06 2023-03-03   786  104.</code></pre>
</div>
</div>
<div class="callout callout-important no-icon callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>The <code>covidcast</code> endpoint supports <code>*</code> in its time and geo fields. Try to obtain the signal values for all available counties by replacing <code>geo_values = "06059"</code> with <code>geo_values = "*"</code>.</p>
</div>
</div>
</div>
</section>
<section id="example-queries---other-endpoints-hospitalizations" class="slide level2">
<h2>Example queries - Other endpoints: Hospitalizations</h2>
<p><span class="primary"><strong>COVID-19 Hospitalization: Facility Lookup</strong></span></p>
<p><small> API docs: <a href="https://cmu-delphi.github.io/delphi-epidata/api/covid_hosp_facility_lookup.html" class="uri">https://cmu-delphi.github.io/delphi-epidata/api/covid_hosp_facility_lookup.html</a> </small></p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/hosp-facility-lookup_f8c34dd3f5640cfe56f366a103ca4927">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a></a><span class="fu">pub_covid_hosp_facility_lookup</span>(<span class="at">city =</span> <span class="st">"southlake"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 10
  hospital_pk state ccn    hospital_name    address city  zip   hospital_subtype
  &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;            &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;           
1 450888      TX    450888 TEXAS HEALTH HA… 1545 E… SOUT… 76092 Short Term      
2 670132      TX    670132 METHODIST SOUTH… 421 E … SOUT… 76092 Short Term      
# ℹ 2 more variables: fips_code &lt;chr&gt;, is_metro_micro &lt;dbl&gt;</code></pre>
</div>
</div>
<p><br></p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/hosp-facility-lookup-wy_528fa452b45a6a360961bad7e6db6787">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a></a><span class="fu">pub_covid_hosp_facility_lookup</span>(<span class="at">state =</span> <span class="st">"WY"</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 10
  hospital_pk     state ccn   hospital_name address city  zip   hospital_subtype
  &lt;chr&gt;           &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;         &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;           
1 100 LANCASTER … WY    2020… 42091         &lt;NA&gt;    [C39… MAIN  390195          
2 2333 BIDDLE AVE WY    2020… 26163         POINT … [C23… HENRY 230146          
3 2333 BIDDLE AV… WY    2020… 26163         POINT … [C23… SELEC 232031          
4 2752 CENTURY B… WY    2020… 42011         POINT … [C39… SURGI 390316          
5 310 SOUTH FALL… WY    2020… 05037         POINT … [C04… CROSS 041307          
6 5200 FAIRVIEW … WY    2020… 27025         POINT … [C24… FAIRV 240050          
# ℹ 2 more variables: fips_code &lt;chr&gt;, is_metro_micro &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a></a><span class="co"># A non-example (there is no city called New York in Wyoming)</span></span>
<span id="cb24-2"><a></a><span class="co"># pub_covid_hosp_facility_lookup(state = "WY", city = "New York")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="example-queries---other-endpoints-hospitalizations-1" class="slide level2">
<h2>Example queries - Other endpoints: Hospitalizations</h2>
<p><span class="primary"><strong>COVID-19 Hospitalization by Facility</strong></span></p>
<p><small> API docs: <a href="https://cmu-delphi.github.io/delphi-epidata/api/covid_hosp_facility.html" class="uri">https://cmu-delphi.github.io/delphi-epidata/api/covid_hosp_facility.html</a> </small></p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/hosp-by-facility_4067152acf661f167d7814d5563bbf2f">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a></a><span class="fu">pub_covid_hosp_facility</span>(</span>
<span id="cb25-2"><a></a>  <span class="at">hospital_pks =</span> <span class="st">"100075"</span>,</span>
<span id="cb25-3"><a></a>  <span class="at">collection_weeks =</span> <span class="fu">epirange</span>(<span class="dv">20200101</span>, <span class="dv">20200501</span>)</span>
<span id="cb25-4"><a></a>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 113
  hospital_pk state ccn    hospital_name    address city  zip   hospital_subtype
  &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;            &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;           
1 100075      FL    100075 ST JOSEPHS HOSP… 3001 W… TAMPA 33677 Short Term      
2 100075      FL    100075 ST JOSEPHS HOSP… 3001 W… TAMPA 33677 Short Term      
3 100075      FL    100075 ST JOSEPHS HOSP… 3001 W… TAMPA 33677 Short Term      
4 100075      FL    100075 ST JOSEPHS HOSP… 3001 W… TAMPA 33677 Short Term      
5 100075      FL    100075 ST JOSEPHS HOSP… 3001 W… TAMPA 33677 Short Term      
6 100075      FL    100075 ST JOSEPHS HOSP… 3001 W… TAMPA 33677 Short Term      
# ℹ 105 more variables: fips_code &lt;chr&gt;, geocoded_hospital_address &lt;chr&gt;,
#   hhs_ids &lt;chr&gt;, publication_date &lt;date&gt;, collection_week &lt;date&gt;,
#   is_metro_micro &lt;lgl&gt;, total_beds_7_day_sum &lt;dbl&gt;,
#   all_adult_hospital_beds_7_day_sum &lt;dbl&gt;,
#   all_adult_hospital_inpatient_beds_7_day_sum &lt;dbl&gt;,
#   inpatient_beds_used_7_day_sum &lt;dbl&gt;,
#   all_adult_hospital_inpatient_bed_occupied_7_day_sum &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p><span class="primary"><strong>COVID-19 Hospitalization by State</strong></span></p>
<p><small> API docs: <a href="https://cmu-delphi.github.io/delphi-epidata/api/covid_hosp.html" class="uri">https://cmu-delphi.github.io/delphi-epidata/api/covid_hosp.html</a> </small></p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/hosp-by-state_5bdad26c96af76731a6643450c9cb9de">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a></a><span class="fu">pub_covid_hosp_state_timeseries</span>(<span class="at">states =</span> <span class="st">"MA"</span>, <span class="at">dates =</span> <span class="st">"20200510"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="example-queries---other-endpoints-flu-endpoints" class="slide level2">
<h2>Example queries - Other endpoints: Flu endpoints</h2>
<p><span class="primary"><strong>FluSurv hospitalization data</strong></span> – Data ends around 2020</p>
<p><small> API docs: <a href="https://cmu-delphi.github.io/delphi-epidata/api/flusurv.html" class="uri">https://cmu-delphi.github.io/delphi-epidata/api/flusurv.html</a> </small></p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/flusurv_cf8e6c8484467f1d2666dbc08bcd26d8">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a></a><span class="fu">pub_flusurv</span>(<span class="at">locations =</span> <span class="st">"ca"</span>, <span class="at">epiweeks =</span> <span class="dv">202001</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><span class="primary"><strong>Fluview data</strong></span> – Remains active</p>
<p><small> API docs: <a href="https://cmu-delphi.github.io/delphi-epidata/api/fluview.html" class="uri">https://cmu-delphi.github.io/delphi-epidata/api/fluview.html</a> </small></p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/fluview_f5939aa9b463e4a24bfa4ef8485104ab">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a></a><span class="fu">pub_fluview</span>(<span class="at">regions =</span> <span class="st">"nat"</span>, <span class="at">epiweeks =</span> <span class="fu">epirange</span>(<span class="dv">201201</span>, <span class="dv">202001</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Tip - public vs private methods</strong></p>
</div>
<div class="callout-content">
<p>Aside from these public methods (starting with <code>pub_</code>), there are private methods (starting with <code>pvt_</code>) that require private access keys. To run these locally, store the keys in your <code>.Reviron</code> file, or set them as environmental variables. See <a href="https://cmu-delphi.github.io/epidatr/articles/signal-discovery.html">Private methods</a> for examples of using private endpoints.</p>
</div>
</div>
</div>
<!-- Aside from these public methods (starting with `pub_`),
there are private methods (starting with `pvt_` when you type `avail_endpoints()`) that require private access keys. To run these locally, store the keys in your `.Reviron` file, or set them as environmental variables.
See [Private methods](https://cmu-delphi.github.io/epidatr/articles/signal-discovery.html) for examples of using private endpoints. -->
</section>
<section id="signal-metadata" class="slide level2">
<h2>Signal metadata</h2>
<ul>
<li>Some endpoints provide additional metadata for signals.
<ul>
<li><span class="primary"><strong>Time Information</strong></span>: Details on available time frames and last update times.</li>
<li><span class="primary"><strong>Geography Information</strong></span>: Details on available geography types.</li>
</ul></li>
<li>Key Endpoints for Metadata
<ul>
<li><code>pub_covidcast_meta()</code>: Access metadata for the COVIDcast endpoint.</li>
<li><code>pub_fluview_meta()</code>: Get metadata for the FluView endpoint.</li>
<li><code>pub_meta()</code>: General metadata for the Delphi Epidata API.</li>
</ul></li>
</ul>
</section></section>
<section>
<section id="versioning-in-epidatr" class="title-slide slide level1 center" data-number="8">
<h1><span class="header-section-number">8</span> Versioning in <code>epidatr</code></h1>

</section>
<section id="versioned-data-in-epidatr" class="slide level2">
<h2>Versioned data in <code>epidatr</code></h2>
<ul>
<li><p>Epidata API contains comprehensive data record, capturing each signal’s estimate, location, date, and update timeline.</p></li>
<li><p>Requesting Specific Data Versions:</p>
<ul>
<li>Use <code>as_of</code> or <code>issues</code> arguments to specify data availability.</li>
<li>Use <code>as_of</code> to fetch one version and <code>issues</code> to fetch multiple.</li>
<li>Only one argument can be used at a time; not all endpoints support both</li>
</ul></li>
<li><p>Let’s start with fetching one version…</p></li>
</ul>
</section>
<section id="obtaining-data-as-of-a-specific-date" class="slide level2">
<h2>Obtaining data “as of” a specific date</h2>
<ul>
<li><strong>Example</strong>: <a href="https://cmu-delphi.github.io/delphi-epidata/api/covidcast-signals/doctor-visits.html">Doctor Visits Signal</a> (from the <code>covidcast</code> endpoint)
<ul>
<li>Estimates the percentage of outpatient doctor visits that are COVID-related. To give a specific example, let’s consider the estimate for PA on May 1, 2020:</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/pa-may-1st-as-of_ad2bc391a3cb44d2744802b30a2be303">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a></a>dv_pa_as_of <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb30-2"><a></a>  <span class="at">source =</span> <span class="st">"doctor-visits"</span>,</span>
<span id="cb30-3"><a></a>  <span class="at">signals =</span> <span class="st">"smoothed_adj_cli"</span>,</span>
<span id="cb30-4"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb30-5"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="st">"2020-05-01"</span>, <span class="st">"2020-05-01"</span>),</span>
<span id="cb30-6"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb30-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"pa"</span>,</span>
<span id="cb30-8"><a></a>  <span class="at">as_of =</span> <span class="st">"2020-05-07"</span></span>
<span id="cb30-9"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/head-pa-may-1st-as-of_72a3c6929f774c83558ccb534d875437">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 7
  geo_value signal           source        time_value issue        lag value
  &lt;chr&gt;     &lt;chr&gt;            &lt;chr&gt;         &lt;date&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt;
1 pa        smoothed_adj_cli doctor-visits 2020-05-01 2020-05-07     6  2.58</code></pre>
</div>
</div>
<ul>
<li>Initial estimate was <em>issued</em> on May 7, 2020 (due to delay from aggregation and ingestion by the API).</li>
</ul>
</section>
<section id="obtaining-data-as-of-a-specific-date-1" class="slide level2">
<h2>Obtaining data “as of” a specific date</h2>
<p><span class="primary"><strong>Default behaviour:</strong></span> If we don’t specify <code>as_of</code>, we get the most recent estimate:</p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/pa-may-1st-most-recent_8966f1265d2bed51323965e9f5c854b8">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a></a>dv_pa_final <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb32-2"><a></a>  <span class="at">source =</span> <span class="st">"doctor-visits"</span>,</span>
<span id="cb32-3"><a></a>  <span class="at">signals =</span> <span class="st">"smoothed_adj_cli"</span>,</span>
<span id="cb32-4"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb32-5"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="st">"2020-05-01"</span>, <span class="st">"2020-05-01"</span>),</span>
<span id="cb32-6"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb32-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"pa"</span></span>
<span id="cb32-8"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/head-pa-may-1st-most-recent_f8c45b908a6d3d8c4646f1ba3f084699">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 8
  geo_value signal           source     time_value issue        lag value stderr
  &lt;chr&gt;     &lt;chr&gt;            &lt;chr&gt;      &lt;date&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-07-04    64  5.97     NA</code></pre>
</div>
</div>
<ul>
<li><span class="primary"><strong>Substantial Estimate Change</strong></span>:
<ul>
<li>Estimate increased from &lt;3% to almost 6% after May 7, reflecting new data on visits from May 1.</li>
</ul></li>
<li><span class="primary"><strong>Versioning is Important in Forecasting</strong></span>:
<ul>
<li>Accurate backtesting requires using data available at the time of model fitting, not later updates, for accurate forecasting results.</li>
</ul></li>
</ul>
</section>
<section id="obtaining-multiple-specific-issues-for-one-state" class="slide level2">
<h2>Obtaining multiple specific issues for one state</h2>
<p>By using the <code>issues</code> argument, we can request all issues in a certain time period:</p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/pa-multiple-issues_df500ca64f10fb1046af2d3895f2b51e">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a></a>dv_pa_issues <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb34-2"><a></a>  <span class="at">source =</span> <span class="st">"doctor-visits"</span>,</span>
<span id="cb34-3"><a></a>  <span class="at">signals =</span> <span class="st">"smoothed_adj_cli"</span>,</span>
<span id="cb34-4"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb34-5"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="st">"2020-05-01"</span>, <span class="st">"2020-05-01"</span>),</span>
<span id="cb34-6"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb34-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"pa"</span>,</span>
<span id="cb34-8"><a></a>  <span class="at">issues =</span> <span class="fu">epirange</span>(<span class="st">"2020-05-01"</span>, <span class="st">"2020-05-15"</span>)</span>
<span id="cb34-9"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/head-pa-multiple-issues_b8a16afb7061c9f28561f65c651d2671">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  geo_value signal           source     time_value issue        lag value stderr
  &lt;chr&gt;     &lt;chr&gt;            &lt;chr&gt;      &lt;date&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-07     6  2.58     NA
2 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-08     7  3.28     NA
3 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-09     8  3.32     NA
4 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-12    11  3.59     NA
5 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-13    12  3.63     NA
6 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-14    13  3.66     NA</code></pre>
</div>
</div>
</section>
<section id="obtaining-multiple-issues-for-one-state" class="slide level2">
<h2>Obtaining multiple issues for one state</h2>
<p>To ensure that you’ve captured all issues up to a specific date (e.g., “2020-05-15”), set an extreme lower bound: <code>issues = epirange("1900-01-01", "2020-05-15")</code>.</p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/extreme-lb-multiple-issues_e9b5e792616b86de83dce7cfa48bf8ef">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a></a>dv_pa_issues_sub <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb36-2"><a></a>  <span class="at">source =</span> <span class="st">"doctor-visits"</span>,</span>
<span id="cb36-3"><a></a>  <span class="at">signals =</span> <span class="st">"smoothed_adj_cli"</span>,</span>
<span id="cb36-4"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb36-5"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="st">"2020-05-01"</span>, <span class="st">"2020-05-01"</span>),</span>
<span id="cb36-6"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb36-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"pa"</span>,</span>
<span id="cb36-8"><a></a>  <span class="at">issues =</span> <span class="fu">epirange</span>(<span class="st">"1900-01-01"</span>, <span class="st">"2020-05-15"</span>)</span>
<span id="cb36-9"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/head-extreme-lb-multiple-issues_a5de7665760c3f0b35aeec1d2c846f5f">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  geo_value signal           source     time_value issue        lag value stderr
  &lt;chr&gt;     &lt;chr&gt;            &lt;chr&gt;      &lt;date&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-07     6  2.58     NA
2 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-08     7  3.28     NA
3 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-09     8  3.32     NA
4 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-12    11  3.59     NA
5 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-13    12  3.63     NA
6 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-14    13  3.66     NA</code></pre>
</div>
</div>
<ul>
<li><p>This doesn’t change anything here, but it may matter for other types of data where you don’t know the latency or reporting lag.</p></li>
<li><p>When in doubt, refer to the <a href="https://cmu-delphi.github.io/delphi-epidata/api/covidcast-signals/doctor-visits.html">signal’s API documentation</a> to find the earliest date available.</p></li>
</ul>
</section>
<section id="obtaining-multiple-issues-for-one-state-1" class="slide level2">
<h2>Obtaining multiple issues for one state</h2>
<ul>
<li><p>Once the maximum issue is reached and there are no more issues to come, the data is considered finalized, regardless of the maximum issue you requested.</p></li>
<li><p><strong>Ex:</strong> The last issue is from July 2020 and so the data is finalized as of that date (not the last requested).</p></li>
</ul>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/extreme-lb-ub-multiple-issues_bad41df4b9695caf36781f435864ceb7">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a></a>dv_pa_issues_all <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb38-2"><a></a>  <span class="at">source =</span> <span class="st">"doctor-visits"</span>,</span>
<span id="cb38-3"><a></a>  <span class="at">signals =</span> <span class="st">"smoothed_adj_cli"</span>,</span>
<span id="cb38-4"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb38-5"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="st">"2020-05-01"</span>, <span class="st">"2020-05-01"</span>),</span>
<span id="cb38-6"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb38-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"pa"</span>,</span>
<span id="cb38-8"><a></a>  <span class="at">issues =</span> <span class="fu">epirange</span>(<span class="st">"1900-01-01"</span>, <span class="st">"2024-12-11"</span>) <span class="co"># From the 1900s to today</span></span>
<span id="cb38-9"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/head-extreme-lb-ub-multiple-issues_c6b8c5a9926b510e14fb9a6cf60439a6">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  geo_value signal           source     time_value issue        lag value stderr
  &lt;chr&gt;     &lt;chr&gt;            &lt;chr&gt;      &lt;date&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-06-29    59  5.99     NA
2 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-06-30    60  5.99     NA
3 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-07-01    61  5.95     NA
4 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-07-02    62  5.97     NA
5 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-07-03    63  5.97     NA
6 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-07-04    64  5.97     NA</code></pre>
</div>
</div>
<ul>
<li>Avoid queries with too-late minimum issue or too-early maximum issue, as it could lead to misleading results. You’re safest bet to capture all issues is on the next slide…</li>
</ul>
<!-- * We caution against starting queries with too late a minimum issue or too early maximum issue, as it could lead to incorrect or misleading results. You're safest bet to capture all issues is on the next slide... -->
</section>
<section id="obtaining-all-issues-for-one-state" class="slide level2">
<h2>Obtaining all issues for one state</h2>
<p>The fast way to obtain all available issues for PA, set <code>issues</code> to be <code>*</code>:</p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/pa-all-issues_89925c0bc63e666c7c9addf110babda7">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a></a>dv_pa_issues_star <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb40-2"><a></a>  <span class="at">source =</span> <span class="st">"doctor-visits"</span>,</span>
<span id="cb40-3"><a></a>  <span class="at">signals =</span> <span class="st">"smoothed_adj_cli"</span>,</span>
<span id="cb40-4"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb40-5"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="st">"2020-05-01"</span>, <span class="st">"2020-05-07"</span>),</span>
<span id="cb40-6"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb40-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"pa"</span>,</span>
<span id="cb40-8"><a></a>  <span class="at">issues =</span> <span class="st">"*"</span></span>
<span id="cb40-9"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/head-pa-all-issues_265965b71ef902e912bba6bbacd1707f">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 × 8
  geo_value signal           source     time_value issue        lag value stderr
  &lt;chr&gt;     &lt;chr&gt;            &lt;chr&gt;      &lt;date&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-07     6  2.58     NA
2 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-08     7  3.28     NA
3 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-09     8  3.32     NA
4 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-12    11  3.59     NA
5 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-13    12  3.63     NA
6 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-14    13  3.66     NA
7 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-15    14  3.66     NA
8 pa        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-16    15  3.61     NA</code></pre>
</div>
</div>
</section>
<section id="obtaining-all-issues-for-all-states" class="slide level2">
<h2>Obtaining all issues for all states</h2>
<p>A very useful feature is the ability to extract all <code>geo_values</code> and <code>issues</code> by using <code>*</code> in both.</p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/all-the-states-and-issues_43b39bbeaba474d6579f3999b345efee">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a></a>dv_state_issues_star <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb42-2"><a></a>  <span class="at">source =</span> <span class="st">"doctor-visits"</span>,</span>
<span id="cb42-3"><a></a>  <span class="at">signals =</span> <span class="st">"smoothed_adj_cli"</span>,</span>
<span id="cb42-4"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb42-5"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="st">"2020-05-01"</span>, <span class="st">"2020-05-07"</span>),</span>
<span id="cb42-6"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb42-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"*"</span>,</span>
<span id="cb42-8"><a></a>  <span class="at">issues =</span> <span class="st">"*"</span></span>
<span id="cb42-9"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/head-all-the-states-and-issues_5e26bd72f983503547177c5ccee958d4">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  geo_value signal           source     time_value issue        lag value stderr
  &lt;chr&gt;     &lt;chr&gt;            &lt;chr&gt;      &lt;date&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 ak        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-07     6  1.61     NA
2 ak        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-08     7  2.40     NA
3 ak        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-09     8  2.38     NA
4 ak        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-12    11  2.38     NA
5 ak        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-13    12  2.36     NA
6 ak        smoothed_adj_cli doctor-vi… 2020-05-01 2020-05-14    13  2.36     NA</code></pre>
</div>
</div>
<p>So, now we obtain all available issues for all states (not just PA) for the specified date range. Neat!</p>
<!--

## Last but not least... The do nothing approach 

![](gfx/do-nothing.jpg){height="550px"}

-->
</section>
<section id="last-but-not-least-the-do-nothing-approach" class="slide level2">
<h2>Last but not least… The do nothing approach</h2>
<p><strong>Final question:</strong> What do you think happens when we adopt a “do nothing” approach to <code>geo_values</code> and <code>issues</code> (take both of them out) for this example?</p>
<p><small> <strong>Hint</strong>: Remember a couple slides ago, when we removed <code>as_of</code>, we got the most recent estimate for PA. </small></p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/do-nothing_8140b318a9e14465e96f4ec5433d6e3a">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a></a>dv_state_default <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb44-2"><a></a>  <span class="at">source =</span> <span class="st">"doctor-visits"</span>,</span>
<span id="cb44-3"><a></a>  <span class="at">signals =</span> <span class="st">"smoothed_adj_cli"</span>,</span>
<span id="cb44-4"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb44-5"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="st">"2020-05-01"</span>, <span class="st">"2020-05-07"</span>),</span>
<span id="cb44-6"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span></span>
<span id="cb44-7"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="obtaining-one-issue-for-all-states" class="slide level2">
<h2>Obtaining one issue for all states</h2>
<p><span class="primary"><strong>Final question:</strong></span> What do you think happens when we adopt a “do nothing” approach to <code>geo_values</code> and <code>issues</code> (take both of them out)?</p>
<p><small> <span class="primary"><strong>Hint</strong></span>: remember a couple slides ago, when we removed <code>as_of</code>, we got the most recent estimate for PA. </small></p>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/one-issue-all-states_a5acd089689a36ae3fa621c13eab4cac">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a></a>dv_state_default <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb45-2"><a></a>  <span class="at">source =</span> <span class="st">"doctor-visits"</span>,</span>
<span id="cb45-3"><a></a>  <span class="at">signals =</span> <span class="st">"smoothed_adj_cli"</span>,</span>
<span id="cb45-4"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb45-5"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="st">"2020-05-01"</span>, <span class="st">"2020-05-07"</span>),</span>
<span id="cb45-6"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span></span>
<span id="cb45-7"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="day1-morning_cache/revealjs/head-one-issue-all-states_767858e7181e00c7c7a68c1dc0c765c4">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  geo_value signal           source     time_value issue        lag value stderr
  &lt;chr&gt;     &lt;chr&gt;            &lt;chr&gt;      &lt;date&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 ak        smoothed_adj_cli doctor-vi… 2020-05-01 2020-07-04    64  5.72     NA
2 al        smoothed_adj_cli doctor-vi… 2020-05-01 2020-07-04    64  2.74     NA
3 ar        smoothed_adj_cli doctor-vi… 2020-05-01 2020-07-04    64  4.23     NA
4 az        smoothed_adj_cli doctor-vi… 2020-05-01 2020-07-04    64  2.78     NA
5 ca        smoothed_adj_cli doctor-vi… 2020-05-01 2020-07-04    64  4.25     NA
6 co        smoothed_adj_cli doctor-vi… 2020-05-01 2020-07-04    64  8.77     NA</code></pre>
</div>
</div>
<p>We get the <span class="primary"><strong>most recent issue</strong></span> for <span class="primary"><strong>all states</strong></span>!</p>
<!-- ## Observations issued with a specific lag
<div style="font-size: 0.8em;">
* We can use the `lag` argument to request only data reported with a certain lag. 

* **Example**: Request  a lag of 7 days fetches only data issued exactly 7 days after the corresponding `time_value`:


::: {.cell layout-align="center" hash='day1-morning_cache/revealjs/specific lag_02e6698e68e4b54a4e01ec7e7b0b520f'}

:::

::: {.cell layout-align="center" hash='day1-morning_cache/revealjs/unnamed-chunk-2_4e944eb2996463a1b13c962aec363504'}

:::

</div>

## Query results exclusion
<div style="font-size: 0.8em;">
* Although the query we ran on the previous slide requested values from May 1 to May 7, May 3 and May 4 were not included due to a 7-day lag.

* Results for those dates appear only if updates are issued on the corresponding lag day (e.g., May 10).


::: {.cell layout-align="center" hash='day1-morning_cache/revealjs/query results exclusion_e2a9a198f959384e6526ee86b8ff27f0'}

:::

::: {.cell layout-align="center" hash='day1-morning_cache/revealjs/unnamed-chunk-3_6b2fcc2e75a59a8f5735c3400b23d071'}

:::

</div>
-->
</section>
<section id="main-takeaways" class="slide level2">
<h2>Main takeaways</h2>
<ul>
<li><span class="primary"><strong>Delphi Epidata:</strong></span> A platform for real-time epidemic data, providing aggregated signals for disease tracking and forecasting from diverse sources like <span class="secondary"><strong>health records</strong></span>, <span class="tertiary"><strong>mobility patterns</strong></span>, and <span class="fourth-colour"><strong>more</strong></span>.</li>
</ul>
<!-- * [**Delphi Epidata:**]{.primary} A one-stop platform for real-time epidemic data, providing aggregated signals for disease tracking and forecasting from diverse sources like health records, mobility patterns, and more. -->
<ul>
<li><span class="primary"><strong>Epidata API:</strong></span> Open-access API delivering up-to-date, granular epidemiological data + historical versions.</li>
</ul>
<!-- Open-access API delivering up-to-date, granular epidemiological data + makes all historical versions available. -->
<ul>
<li><span class="primary"><strong>Epidatr:</strong></span> Enables you to access Delphi’s epidemiological data through R, offering powerful API functions.</li>
</ul>
<!-- and interactive tools for discovering and analyzing health signals. -->
<ul>
<li><p><span class="primary"><strong>Versioned Data and Latency:</strong></span> A standout feature of this API is its inclusion of two critical fields…</p>
<ol type="1">
<li><p><code>as_of</code>: One version of the data, and referring to the specific date when the data was last updated</p></li>
<li><p><code>issues</code>: Multiple versions of the data, each corresponding to different <code>as_of</code> dates.</p></li>
</ol></li>
</ul>
<p>Their purpose is to manage the record of revisions for transparency and accuracy in data analysis.</p>
<!-- Panel data captures time-series trends, which are often subject to revision.  A standout feature of this API is its inclusion of two critical fields... -->
<!-- 1. as_of:  One version of the data, and referring to the specific date when the data was last updated (i.e. the data was updated `as_of` this date) -->
<!-- 2. `issues`: Multiple versions of the data, each corresponding to different `as_of` dates, capturing revisions over time. -->


</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<p><img src="gfx/delphi.jpg" class="slide-logo"></p>
<div class="footer footer-default">
<p>Panel Data — cmu-delphi/insightnet-workshop-2024</p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: false,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        function fireSlideChanged(previousSlide, currentSlide) {

          // dispatch for htmlwidgets
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for reveal
        if (window.Reveal) {
          window.Reveal.addEventListener("slidechanged", function(event) {
            fireSlideChanged(event.previousSlide, event.currentSlide);
          });
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp("https:\/\/cmu-delphi\.github\.io\/insightnet-workshop-2024\/");
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
              // target, if specified
              link.setAttribute("target", "_blank");
              if (link.getAttribute("rel") === null) {
                link.setAttribute("rel", "noopener");
              }
              // default icon
              link.classList.add("external");
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>