<!DOCTYPE html>
<html lang="en-CA"><head>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.6.39">

  <meta name="author" content="Alice Cima, Rachel Lobay, Daniel McDonald, Ryan Tibshirani">
  <title>InsightNet EpiData Workshop 2024 – day1-afternoon</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto-09d4a5b6094fc3e7d9a2f1c4dd5542e0.css">
  <link rel="stylesheet" href="tachyons-minimal.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">


<section class="slide level2">

<div class="flex">
<div class="w-20">

</div>
<section id="section" class="slide level2 w-80" data-background-image="gfx/cover-art-1.svg" data-background-position="bottom">
<h2>Data Cleaning, Versioning, Nowcasting With <code>{epiprocess}</code></h2>
<h3 id="section-1">InsightNet Forecasting Workshop 2024</h3>
<p><br></p>
<h4 id="section-2">Alice Cima, Rachel Lobay, Daniel McDonald, Ryan Tibshirani</h4>
<p><span class="fstyle">with huge thanks to Logan Brooks, Xueda Shen, and also to Nat DeFries, Dmitry Shemetov, and David Weber</span></p>
<p>11 December – Afternoon</p>
</section>
</div>
</section>
<section id="outline" class="slide level2">
<h2>Outline</h2>
<ol type="1">
<li><p>Essentials of <code>{dplyr}</code> and <code>{tidyr}</code></p></li>
<li><p>Epiverse Software ecosystem</p></li>
<li><p>Panel and Versioned Data in the epiverse</p></li>
<li><p>Basic Nowcasting using <code>{epiprocess}</code></p></li>
<li><p>Nowcasting with One Variable</p></li>
<li><p>Nowcasting with Two Variables</p></li>
<li><p>Case Study - Nowcasting Cases Using %CLI</p></li>
</ol>
</section>
<section>
<section id="essentials-of-dplyr-and-tidyr" class="title-slide slide level1 center" data-number="1">
<h1><span class="header-section-number">1</span> Essentials of <code>{dplyr}</code> and <code>{tidyr}</code></h1>

</section>
<section id="down-with-spreadsheets-for-data-manipulation" class="slide level2">
<h2>Down with spreadsheets for data manipulation</h2>
<ul>
<li>Spreadsheets make it difficult to rerun analyses consistently.</li>
<li>Using R (and <code>{dplyr}</code>) allows for:
<ul>
<li>Reproducibility</li>
<li>Ease of modification</li>
</ul></li>
<li><span class="primary"><strong>Recommendation</strong></span>: Avoid manual edits; instead, use code for transformations.</li>
<li>Let’s see what we mean by this…</li>
</ul>
</section>
<section id="introduction-to-dplyr" class="slide level2">
<h2>Introduction to <code>dplyr</code></h2>
<ul>
<li><code>dplyr</code> is a powerful package in R for data manipulation.</li>
<li>It is part of the <code>tidyverse</code>, which includes a collection of packages designed to work together… Here’s some of its greatest hits:</li>
</ul>
<div style="text-align: center;">
<p><img data-src="gfx/tidyverse_packages.png" style="width: 30%; display: block; margin-left: auto; margin-right: auto;"> <br> <small><a href="https://laddem.github.io/courses/tidyverse/">Source</a></small></p>
</div>
</section>
<section id="introduction-to-dplyr-1" class="slide level2">
<h2>Introduction to <code>dplyr</code></h2>
<p>To load <code>dplyr</code>, you may simply load the <code>tidyverse</code> package:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a></a><span class="co"># install.packages("tidyverse")</span></span>
<span id="cb1-2"><a></a><span class="fu">library</span>(tidyverse)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Our focus will be on basic operations like selecting and filtering data.</p>
<img data-src="gfx/dplyr_and_fun.png" style="width: 60%;">
<div style="text-align: center;">
<p><small><a href="https://towardsdatascience.com/data-manipulation-in-r-with-dplyr-3095e0867f75">Source</a></small></p>
</div>
</section>
<section id="downloading-jhu-csse-covid-19-case-data" class="slide level2">
<h2>Downloading JHU CSSE COVID-19 case data</h2>
<ul>
<li>Let’s start with something familiar… Here’s a task for you:</li>
<li>Use <code>pub_covidcast()</code> to download <span class="primary"><strong>JHU CSSE COVID-19 confirmed case data</strong></span> (<code>confirmed_incidence_num</code>) for CA, NC, and NY from March 1, 2022 to March 31, 2022 as of January 1, 2024.</li>
<li>Try this for yourself. Then click the dropdown on the next slide to check your work…</li>
</ul>
</section>
<section id="downloading-jhu-csse-covid-19-case-data-1" class="slide level2">
<h2>Downloading JHU CSSE COVID-19 case data</h2>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a></a><span class="fu">library</span>(epidatr)</span>
<span id="cb2-2"><a></a></span>
<span id="cb2-3"><a></a>cases_df_api <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb2-4"><a></a>  <span class="at">source =</span> <span class="st">"jhu-csse"</span>,</span>
<span id="cb2-5"><a></a>  <span class="at">signals =</span> <span class="st">"confirmed_incidence_num"</span>,</span>
<span id="cb2-6"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb2-7"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb2-8"><a></a>  <span class="at">geo_values =</span> <span class="st">"ca,nc,ny"</span>,</span>
<span id="cb2-9"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">20220301</span>, <span class="dv">20220331</span>),</span>
<span id="cb2-10"><a></a>  <span class="at">as_of =</span> <span class="fu">as.Date</span>(<span class="st">"2024-01-01"</span>)</span>
<span id="cb2-11"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now we only really need a few columns here…</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a></a>cases_df <span class="ot">&lt;-</span> cases_df_api <span class="sc">|&gt;</span></span>
<span id="cb3-2"><a></a>  <span class="fu">select</span>(geo_value, time_value, <span class="at">raw_cases =</span> value) <span class="co"># We'll talk more about this soon :)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ways-to-inspect-the-dataset" class="slide level2">
<h2>Ways to inspect the dataset</h2>
<p>Use <code>head()</code> to view the first six row of the data</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a></a><span class="fu">head</span>(cases_df)  <span class="co"># First 6 rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  geo_value time_value raw_cases
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;
1 ca        2022-03-01      4310
2 nc        2022-03-01      1231
3 ny        2022-03-01      1487
4 ca        2022-03-02      7044
5 nc        2022-03-02      2243
6 ny        2022-03-02      1889</code></pre>
</div>
</div>
<p>and tail to view the last six</p>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  geo_value time_value raw_cases
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;
1 ca        2022-03-30      3785
2 nc        2022-03-30      1067
3 ny        2022-03-30      3127
4 ca        2022-03-31      4533
5 nc        2022-03-31      1075
6 ny        2022-03-31      4763</code></pre>
</div>
</div>
</section>
<section id="ways-to-inspect-the-dataset-1" class="slide level2">
<h2>Ways to inspect the dataset</h2>
<p>Now, for our first foray into the <code>tidyverse</code>…</p>
<p>Use <code>glimpse()</code> to get a compact overview of the dataset.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a></a><span class="fu">glimpse</span>(cases_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 93
Columns: 3
$ geo_value  &lt;chr&gt; "ca", "nc", "ny", "ca", "nc", "ny", "ca", "nc", "ny", "ca",…
$ time_value &lt;date&gt; 2022-03-01, 2022-03-01, 2022-03-01, 2022-03-02, 2022-03-02…
$ raw_cases  &lt;dbl&gt; 4310, 1231, 1487, 7044, 2243, 1889, 7509, 2377, 2390, 3586,…</code></pre>
</div>
</div>
<!-- ## Creating tibbles

* [**Tibbles**]{.primary}: Modern data frames with enhanced features.
* Rows represent [**observations**]{.primary} (or cases).
* Columns represent [**variables**]{.primary} (or features).
* You can create tibbles manually using the `tibble()` function.




::: {.cell layout-align="center"}

```{.r .cell-code}
tibble(x = letters, y = 1:26)
```

::: {.cell-output .cell-output-stdout}

```
# A tibble: 26 × 2
   x         y
   <chr> <int>
 1 a         1
 2 b         2
 3 c         3
 4 d         4
 5 e         5
 6 f         6
 7 g         7
 8 h         8
 9 i         9
10 j        10
# ℹ 16 more rows
```


:::
:::



-->
</section>
<section id="selecting-columns-with-select" class="slide level2">
<h2>Selecting columns with <code>select()</code></h2>
<p>The <code>select()</code> function is used to pick specific columns from your dataset.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a></a><span class="fu">select</span>(cases_df, geo_value, time_value)  <span class="co"># Select the 'geo_value' and 'time_value' columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 93 × 2
   geo_value time_value
   &lt;chr&gt;     &lt;date&gt;    
 1 ca        2022-03-01
 2 nc        2022-03-01
 3 ny        2022-03-01
 4 ca        2022-03-02
 5 nc        2022-03-02
 6 ny        2022-03-02
 7 ca        2022-03-03
 8 nc        2022-03-03
 9 ny        2022-03-03
10 ca        2022-03-04
# ℹ 83 more rows</code></pre>
</div>
</div>
</section>
<section id="selecting-columns-with-select-1" class="slide level2">
<h2>Selecting columns with <code>select()</code></h2>
<p>You can exclude columns by prefixing the column names with a minus sign <code>-</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a></a><span class="fu">select</span>(cases_df, <span class="sc">-</span>raw_cases)  <span class="co"># Exclude the 'raw_cases' column from the dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 93 × 2
   geo_value time_value
   &lt;chr&gt;     &lt;date&gt;    
 1 ca        2022-03-01
 2 nc        2022-03-01
 3 ny        2022-03-01
 4 ca        2022-03-02
 5 nc        2022-03-02
 6 ny        2022-03-02
 7 ca        2022-03-03
 8 nc        2022-03-03
 9 ny        2022-03-03
10 ca        2022-03-04
# ℹ 83 more rows</code></pre>
</div>
</div>
<!-- ## Extracting columns with `pull()`

* `pull()`: Extract a column as a vector.
* Let's try this with the `cases` column...




::: {.cell layout-align="center"}

```{.r .cell-code}
pull(cases_df, raw_cases) 
```

::: {.cell-output .cell-output-stdout}

```
 [1] 4310 1231 1487 7044 2243 1889 7509 2377 2390 3586 2646  350 1438    0 3372
[16] 6465    0 2343 6690 4230 1033 3424  894 1025 4591 1833 1691 5359 1783 1747
[31] 2713 1849 2229 1623    0 1396 5151    0 2202 4826 3130  982 1831  649 3128
[46] 3706    0 2039 6143 2742 2356 4204 1740 2052 3256    0 2188 4659    0 2667
[61] 5499 2508 1177 3004  819 1603 3943 1602  551 3550 1288 6596 1960 1224 3542
[76] 1035    0    0 3384    0 5908 2811 2291 2286 1846  624 2394 3785 1067 3127
[91] 4533 1075 4763
```


:::
:::



-->
</section>
<section id="subsetting-rows-with-filter" class="slide level2">
<h2>Subsetting rows with <code>filter()</code></h2>
<div style="text-align: center;">
<p><img data-src="gfx/dplyr_filter.png" style="width: 65%; display: block; margin-left: auto; margin-right: auto;"> <br> <small><a href="https://x.com/allison_horst">Artwork by <span class="citation" data-cites="allison_horst">@allison_horst</span></a></small></p>
</div>
</section>
<section id="subsetting-rows-with-filter-1" class="slide level2">
<h2>Subsetting rows with <code>filter()</code></h2>
<ul>
<li>The <code>filter()</code> function allows you to subset rows that meet specific conditions.</li>
<li>Conditions regard column values, such as filtering for only NC or cases higher than some threshold.</li>
<li>This enables you to narrow down your dataset to focus on relevant data.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a></a><span class="fu">filter</span>(cases_df, geo_value <span class="sc">==</span> <span class="st">"nc"</span>, raw_cases <span class="sc">&gt;</span> <span class="dv">500</span>)  <span class="co"># Filter for NC with raw daily cases &gt; 500</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 22 × 3
   geo_value time_value raw_cases
   &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;
 1 nc        2022-03-01      1231
 2 nc        2022-03-02      2243
 3 nc        2022-03-03      2377
 4 nc        2022-03-04      2646
 5 nc        2022-03-07      4230
 6 nc        2022-03-08       894
 7 nc        2022-03-09      1833
 8 nc        2022-03-10      1783
 9 nc        2022-03-11      1849
10 nc        2022-03-14      3130
# ℹ 12 more rows</code></pre>
</div>
</div>
</section>
<section id="combining-select-and-filter-functions" class="slide level2">
<h2>Combining <code>select()</code> and <code>filter()</code> functions</h2>
<ul>
<li>You can further combine <code>select()</code> and <code>filter()</code> to further refine the dataset.</li>
<li>Use <code>select()</code> to choose columns and <code>filter()</code> to narrow down rows.</li>
<li>This helps in extracting the exact data needed for analysis.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a></a><span class="fu">select</span>(<span class="fu">filter</span>(cases_df, geo_value <span class="sc">==</span> <span class="st">"nc"</span>, raw_cases <span class="sc">&gt;</span> <span class="dv">1000</span>), time_value, raw_cases) <span class="sc">|&gt;</span> </span>
<span id="cb15-2"><a></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  time_value raw_cases
  &lt;date&gt;         &lt;dbl&gt;
1 2022-03-01      1231
2 2022-03-02      2243
3 2022-03-03      2377
4 2022-03-04      2646
5 2022-03-07      4230
6 2022-03-09      1833</code></pre>
</div>
</div>
</section>
<section id="using-the-pipe-operator" class="slide level2">
<h2>Using the pipe operator <code>|&gt;</code></h2>
<ul>
<li>The pipe operator (<code>|&gt;</code>) makes code more readable by chaining multiple operations together.</li>
<li>The output of one function is automatically passed to the next function.</li>
<li>This allows you to perform multiple steps (e.g., <code>filter()</code> followed by <code>select()</code>) in a clear and concise manner.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a></a><span class="co"># This code reads more like poetry!</span></span>
<span id="cb17-2"><a></a>cases_df <span class="sc">|&gt;</span> </span>
<span id="cb17-3"><a></a>  <span class="fu">filter</span>(geo_value <span class="sc">==</span> <span class="st">"nc"</span>, raw_cases <span class="sc">&gt;</span> <span class="dv">1000</span>) <span class="sc">|&gt;</span> </span>
<span id="cb17-4"><a></a>  <span class="fu">select</span>(time_value, raw_cases) <span class="sc">|&gt;</span> </span>
<span id="cb17-5"><a></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  time_value raw_cases
  &lt;date&gt;         &lt;dbl&gt;
1 2022-03-01      1231
2 2022-03-02      2243
3 2022-03-03      2377
4 2022-03-04      2646
5 2022-03-07      4230
6 2022-03-09      1833</code></pre>
</div>
</div>
</section>
<section id="key-practices-in-dplyr" class="slide level2">
<h2>Key practices in <code>dplyr</code></h2>
<ul>
<li>Use <span class="primary"><strong>tibbles</strong></span> for easier data handling.</li>
<li>Use <code>head()</code>, <code>tail()</code>, and <code>glimpse()</code> for quick data inspection.</li>
<li>Use <code>select()</code> and <code>filter()</code> for data manipulation.</li>
<li>Chain functions with <code>|&gt;</code> for cleaner code.</li>
</ul>
<!-- * Use `pull()` to extract columns as vectors. -->
</section>
<section id="grouping-data-with-group_by" class="slide level2">
<h2>Grouping data with <code>group_by()</code></h2>
<ul>
<li>Use <code>group_by()</code> to group data by one or more columns.</li>
<li>Allows performing operations on specific groups of data.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a></a>cases_df <span class="sc">|&gt;</span></span>
<span id="cb19-2"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">|&gt;</span></span>
<span id="cb19-3"><a></a>  <span class="fu">filter</span>(raw_cases <span class="sc">==</span> <span class="fu">max</span>(raw_cases, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 3
# Groups:   geo_value [3]
  geo_value time_value raw_cases
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;
1 ca        2022-03-03      7509
2 nc        2022-03-07      4230
3 ny        2022-03-24      6596</code></pre>
</div>
</div>
</section>
<section id="creating-new-columns-with-mutate" class="slide level2">
<h2>Creating new columns with <code>mutate()</code></h2>
<div style="text-align: center;">
<p><img data-src="gfx/dplyr_mutate.jpg" style="width: 45%; display: block; margin-left: auto; margin-right: auto;"> <br> <small><a href="https://x.com/allison_horst">Artwork by <span class="citation" data-cites="allison_horst">@allison_horst</span></a></small></p>
</div>
</section>
<section id="creating-new-columns-with-mutate-1" class="slide level2">
<h2>Creating new columns with <code>mutate()</code></h2>
<ul>
<li><code>mutate()</code> is used to create new columns.</li>
<li>Perform calculations using existing columns and assign to new columns.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a></a>ny_subset <span class="ot">=</span> cases_df <span class="sc">|&gt;</span></span>
<span id="cb21-2"><a></a>  <span class="fu">filter</span>(geo_value <span class="sc">==</span> <span class="st">"ny"</span>)</span>
<span id="cb21-3"><a></a></span>
<span id="cb21-4"><a></a>ny_subset <span class="sc">|&gt;</span> </span>
<span id="cb21-5"><a></a>  <span class="fu">mutate</span>(<span class="at">cumulative_cases =</span> <span class="fu">cumsum</span>(raw_cases)) <span class="sc">|&gt;</span> </span>
<span id="cb21-6"><a></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  geo_value time_value raw_cases cumulative_cases
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;            &lt;dbl&gt;
1 ny        2022-03-01      1487             1487
2 ny        2022-03-02      1889             3376
3 ny        2022-03-03      2390             5766
4 ny        2022-03-04       350             6116
5 ny        2022-03-05      3372             9488
6 ny        2022-03-06      2343            11831</code></pre>
</div>
</div>
<!-- ## Creating new columns with `mutate()`

* `mutate()` can create multiple new columns in one step.
* Logical comparisons (e.g., `over_5000 = raw_cases > 5000`) can be used within `mutate()`.




::: {.cell layout-align="center"}

```{.r .cell-code}
ny_subset |> 
  mutate(over_5000 = raw_cases > 5000,
         cumulative_cases = cumsum(raw_cases)) |> 
  head()
```

::: {.cell-output .cell-output-stdout}

```
# A tibble: 6 × 5
  geo_value time_value raw_cases over_5000 cumulative_cases
  <chr>     <date>         <dbl> <lgl>                <dbl>
1 ny        2022-03-01      1487 FALSE                 1487
2 ny        2022-03-02      1889 FALSE                 3376
3 ny        2022-03-03      2390 FALSE                 5766
4 ny        2022-03-04       350 FALSE                 6116
5 ny        2022-03-05      3372 FALSE                 9488
6 ny        2022-03-06      2343 FALSE                11831
```


:::
:::



-->
</section>
<section id="combining-group_by-and-mutate" class="slide level2">
<h2>Combining <code>group_by()</code> and <code>mutate()</code></h2>
<ul>
<li>First, group data using <code>group_by()</code>.</li>
<li>Then, use <code>mutate</code> to perform the calculations for each group.</li>
<li>Finally, use <code>arrange</code> to display the output by <code>geo_value</code>.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a></a>cases_df <span class="sc">|&gt;</span></span>
<span id="cb23-2"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">|&gt;</span></span>
<span id="cb23-3"><a></a>  <span class="fu">mutate</span>(<span class="at">cumulative_cases =</span> <span class="fu">cumsum</span>(raw_cases)) <span class="sc">|&gt;</span> </span>
<span id="cb23-4"><a></a>  <span class="fu">arrange</span>(geo_value) <span class="sc">|&gt;</span> </span>
<span id="cb23-5"><a></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
# Groups:   geo_value [1]
  geo_value time_value raw_cases cumulative_cases
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;            &lt;dbl&gt;
1 ca        2022-03-01      4310             4310
2 ca        2022-03-02      7044            11354
3 ca        2022-03-03      7509            18863
4 ca        2022-03-04      3586            22449
5 ca        2022-03-05      1438            23887
6 ca        2022-03-06      6465            30352</code></pre>
</div>
</div>
<!-- ## Conditional calculations with `if_else()`
* `if_else()` allows conditional logic within `mutate()`.
* Perform different operations depending on conditions, like "high" or "low."




::: {.cell layout-align="center"}

```{.r .cell-code}
t <- 5000

cases_df |>
  mutate(high_low_cases = if_else(raw_cases > t, "high", "low")) |> 
  head()
```

::: {.cell-output .cell-output-stdout}

```
# A tibble: 6 × 4
  geo_value time_value raw_cases high_low_cases
  <chr>     <date>         <dbl> <chr>         
1 ca        2022-03-01      4310 low           
2 nc        2022-03-01      1231 low           
3 ny        2022-03-01      1487 low           
4 ca        2022-03-02      7044 high          
5 nc        2022-03-02      2243 low           
6 ny        2022-03-02      1889 low           
```


:::
:::



-->
</section>
<section id="summarizing-data-with-summarise" class="slide level2">
<h2>Summarizing data with <code>summarise()</code></h2>
<ul>
<li><code>summarise()</code> reduces data to summary statistics (e.g., mean, median).</li>
<li>Typically used after <code>group_by()</code> to summarize each group.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a></a>cases_df <span class="sc">|&gt;</span></span>
<span id="cb25-2"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">|&gt;</span></span>
<span id="cb25-3"><a></a>  <span class="fu">summarise</span>(<span class="at">median_cases =</span> <span class="fu">median</span>(raw_cases))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  geo_value median_cases
  &lt;chr&gt;            &lt;dbl&gt;
1 ca                3785
2 nc                1224
3 ny                2188</code></pre>
</div>
</div>
</section>
<section id="using-count-to-aggregate-data" class="slide level2">
<h2>Using <code>count()</code> to aggregate data</h2>
<p><code>count()</code> is a shortcut for grouping and summarizing the data.</p>
<p>For example, if we want to get the total number of complete rows for each state, then</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a></a>cases_count <span class="ot">&lt;-</span> cases_df <span class="sc">|&gt;</span></span>
<span id="cb27-2"><a></a>  <span class="fu">drop_na</span>() <span class="sc">|&gt;</span> <span class="co"># Removes rows where any value is missing (from tidyr)</span></span>
<span id="cb27-3"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">|&gt;</span></span>
<span id="cb27-4"><a></a>  <span class="fu">summarize</span>(<span class="at">count =</span> <span class="fu">n</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br> is equivalent to</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a></a>cases_count <span class="ot">&lt;-</span> cases_df <span class="sc">|&gt;</span></span>
<span id="cb28-2"><a></a>  <span class="fu">drop_na</span>() <span class="sc">|&gt;</span> </span>
<span id="cb28-3"><a></a>  <span class="fu">count</span>(geo_value)</span>
<span id="cb28-4"><a></a></span>
<span id="cb28-5"><a></a>cases_count <span class="co"># Let's see what the counts are.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  geo_value     n
  &lt;chr&gt;     &lt;int&gt;
1 ca           31
2 nc           31
3 ny           31</code></pre>
</div>
</div>
</section>
<section id="key-practices-in-dplyr-round-2" class="slide level2">
<h2>Key practices in <code>dplyr</code>: Round 2</h2>
<ul>
<li>Use <code>group_by()</code> to group data by one or more variables before applying functions.</li>
<li>Use <code>mutate</code> to create new columns or modify existing ones by applying functions to existing data.</li>
<li>Use <code>summarise</code> to reduce data to summary statistics (e.g., mean, median).</li>
<li><code>count()</code> is a convenient shortcut for counting rows by group without needing <code>group_by()</code> and <code>summarise()</code>.</li>
</ul>
</section>
<section id="tidy-data-and-tolstoy" class="slide level2">
<h2>Tidy data and Tolstoy</h2>
<blockquote>
<p>“Happy families are all alike; every unhappy family is unhappy in its own way.” — Leo Tolstoy</p>
</blockquote>
<ul>
<li><span class="primary"><strong>Tidy datasets</strong></span> are like happy families: consistent, standardized, and easy to work with.<br>
</li>
<li><span class="primary"><strong>Messy datasets</strong></span> are like unhappy families: each one messy in its own unique way.<br>
In this section:</li>
<li>We’ll define what makes data <em>tidy</em> and how to transform between the tidy and messy formats.</li>
</ul>
</section>
<section id="tidy-data-and-tolstoy-1" class="slide level2">
<h2>Tidy data and Tolstoy</h2>

<img data-src="gfx/tidy_messy_data.jpg" style="width: 60%;" class="r-stretch"><p><small><a href="https://x.com/allison_horst">Artwork by <span class="citation" data-cites="allison_horst">@allison_horst</span></a></small></p>
</section>
<section id="what-is-tidy-data" class="slide level2">
<h2>What is tidy data?</h2>
<ul>
<li>Tidy data follows a consistent structure: <span class="primary"><strong>each row represents one observation, and each column represents one variable.</strong></span></li>
<li><code>cases_df</code> is one classic example of tidy data.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  geo_value time_value raw_cases
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;
1 ca        2022-03-01      4310
2 nc        2022-03-01      1231
3 ny        2022-03-01      1487
4 ca        2022-03-02      7044
5 nc        2022-03-02      2243
6 ny        2022-03-02      1889</code></pre>
</div>
</div>
<ul>
<li>To convert between tidy and messy data, we can use the <code>tidyr</code> package in the tidyverse.</li>
</ul>
</section>
<section id="pivot_wider-and-pivot_longer" class="slide level2">
<h2><code>pivot_wider()</code> and <code>pivot_longer()</code></h2>
<div style="text-align: center;">
<p><img data-src="gfx/pivot_wider_longer.jpg" style="width: 40%; display: block; margin-left: auto; margin-right: auto;"> <br> <small><a href="https://x.com/allison_horst">Artwork by <span class="citation" data-cites="allison_horst">@allison_horst</span></a></small></p>
</div>
</section>
<section id="making-data-wider-with-pivot_wider" class="slide level2">
<h2>Making data wider with <code>pivot_wider()</code></h2>
<ul>
<li>To convert data from long format to wide/messy format use <code>pivot_wider()</code>.</li>
<li>For example, let’s try creating a column for each time value in <code>cases_df</code>:</li>
</ul>
<!-- Example. Spreadsheet from hell -->
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a></a>messy_cases_df <span class="ot">&lt;-</span> cases_df <span class="sc">|&gt;</span></span>
<span id="cb31-2"><a></a>  <span class="fu">pivot_wider</span>(</span>
<span id="cb31-3"><a></a>    <span class="at">names_from =</span> time_value,   <span class="co"># Create new columns for each unique date</span></span>
<span id="cb31-4"><a></a>    <span class="at">values_from =</span> raw_cases    <span class="co"># Fill those columns with the raw_case values</span></span>
<span id="cb31-5"><a></a>  )</span>
<span id="cb31-6"><a></a></span>
<span id="cb31-7"><a></a><span class="co"># View the result</span></span>
<span id="cb31-8"><a></a>messy_cases_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 32
  geo_value `2022-03-01` `2022-03-02` `2022-03-03` `2022-03-04` `2022-03-05`
  &lt;chr&gt;            &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;
1 ca                4310         7044         7509         3586         1438
2 nc                1231         2243         2377         2646            0
3 ny                1487         1889         2390          350         3372
# ℹ 26 more variables: `2022-03-06` &lt;dbl&gt;, `2022-03-07` &lt;dbl&gt;,
#   `2022-03-08` &lt;dbl&gt;, `2022-03-09` &lt;dbl&gt;, `2022-03-10` &lt;dbl&gt;,
#   `2022-03-11` &lt;dbl&gt;, `2022-03-12` &lt;dbl&gt;, `2022-03-13` &lt;dbl&gt;,
#   `2022-03-14` &lt;dbl&gt;, `2022-03-15` &lt;dbl&gt;, `2022-03-16` &lt;dbl&gt;,
#   `2022-03-17` &lt;dbl&gt;, `2022-03-18` &lt;dbl&gt;, `2022-03-19` &lt;dbl&gt;,
#   `2022-03-20` &lt;dbl&gt;, `2022-03-21` &lt;dbl&gt;, `2022-03-22` &lt;dbl&gt;,
#   `2022-03-23` &lt;dbl&gt;, `2022-03-24` &lt;dbl&gt;, `2022-03-25` &lt;dbl&gt;, …</code></pre>
</div>
</div>
</section>
<section id="tidying-messy-data-with-pivot_longer" class="slide level2">
<h2>Tidying messy data with <code>pivot_longer()</code></h2>
<ul>
<li>Use <code>pivot_longer()</code> to convert data from <span class="primary"><strong>wide format</strong></span> (multiple columns for the same variable) to <span class="primary"><strong>long format</strong></span> (one column per variable).</li>
<li>Let’s try turning <code>messy_cases_df</code> back into the original tidy <code>cases_df</code>!</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a></a>tidy_cases_df <span class="ot">&lt;-</span> messy_cases_df <span class="sc">|&gt;</span></span>
<span id="cb33-2"><a></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb33-3"><a></a>    <span class="at">cols =</span> <span class="sc">-</span>geo_value,          <span class="co"># Keep the 'geo_value' column as it is</span></span>
<span id="cb33-4"><a></a>    <span class="at">names_to =</span> <span class="st">"time_value"</span>,    <span class="co"># Create a new 'time_value' column from the column names</span></span>
<span id="cb33-5"><a></a>    <span class="at">values_to =</span> <span class="st">"raw_cases"</span>     <span class="co"># Values from the wide columns should go into 'raw_cases'</span></span>
<span id="cb33-6"><a></a>  )</span>
<span id="cb33-7"><a></a></span>
<span id="cb33-8"><a></a><span class="co"># View the result</span></span>
<span id="cb33-9"><a></a><span class="fu">head</span>(tidy_cases_df, <span class="at">n =</span> <span class="dv">3</span>) <span class="co"># Notice the class of time_value here</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 3
  geo_value time_value raw_cases
  &lt;chr&gt;     &lt;chr&gt;          &lt;dbl&gt;
1 ca        2022-03-01      4310
2 ca        2022-03-02      7044
3 ca        2022-03-03      7509</code></pre>
</div>
</div>
</section>
<section id="tidying-messy-data-with-pivot_longer-1" class="slide level2">
<h2>Tidying messy data with <code>pivot_longer()</code></h2>
<ul>
<li>When we used <code>pivot_longer()</code>, the <code>time_value</code> column is converted to a character class because the column names are treated as strings.</li>
<li>So, to truly get the original <code>cases_df</code> we need to convert <code>time_value</code> back to the <code>Date</code> class.</li>
<li>Then, we can use <code>identical()</code> to check if the two data frames are exactly the same.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a></a>tidy_cases_df <span class="ot">=</span> tidy_cases_df <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">time_value =</span> <span class="fu">as.Date</span>(time_value))</span>
<span id="cb35-2"><a></a></span>
<span id="cb35-3"><a></a><span class="fu">identical</span>(tidy_cases_df <span class="sc">|&gt;</span> <span class="fu">arrange</span>(time_value), cases_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>Great. That was a success!</p>
</section>
<section id="introduction-to-joins-in-dplyr" class="slide level2">
<h2>Introduction to joins in <code>dplyr</code></h2>
<ul>
<li>Joining datasets is a powerful tool for combining info. from multiple sources.</li>
<li>In R, <code>dplyr</code> provides several functions to perform different types of joins.</li>
<li>We’ll demonstrate joining a subset of <code>cases_df</code> (our case counts dataset) with <code>state_census</code>.</li>
<li><span class="primary"><strong>Motivation</strong></span>: We can scale the case counts by population to make them comparable across regions of different sizes.</li>
</ul>
</section>
<section id="subset-cases_df" class="slide level2">
<h2>Subset <code>cases_df</code></h2>
<p>To simplify things, let’s use <code>filter()</code> to only grab one date of <code>cases_df</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a></a>cases_df_sub <span class="ot">&lt;-</span> cases_df <span class="sc">|&gt;</span> <span class="fu">filter</span>(time_value <span class="sc">==</span> <span class="st">"2022-03-01"</span>)</span>
<span id="cb37-2"><a></a>cases_df_sub</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 3
  geo_value time_value raw_cases
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;
1 ca        2022-03-01      4310
2 nc        2022-03-01      1231
3 ny        2022-03-01      1487</code></pre>
</div>
</div>
<p>Though note that what we’re going to do can be applied to the entirety of <code>cases_df</code>.</p>
</section>
<section id="load-state-census-data" class="slide level2">
<h2>Load state census data</h2>
<p>The <code>state_census</code> dataset from <code>epidatasets</code> contains state populations from the 2019 census.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a></a><span class="co"># State census dataset from epidatasets</span></span>
<span id="cb39-2"><a></a><span class="fu">library</span>(epidatasets)</span>
<span id="cb39-3"><a></a>state_census <span class="ot">=</span> state_census <span class="sc">|&gt;</span> <span class="fu">select</span>(abbr, pop) <span class="sc">|&gt;</span> <span class="fu">filter</span>(abbr <span class="sc">!=</span> <span class="st">"us"</span>)</span>
<span id="cb39-4"><a></a></span>
<span id="cb39-5"><a></a>state_census <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  abbr       pop
  &lt;chr&gt;    &lt;dbl&gt;
1 al     4903185
2 ak      731545
3 az     7278717
4 ar     3017804
5 ca    39512223
6 co     5758736</code></pre>
</div>
</div>
<p>Notice that this includes many states that are not in <code>cases_df_sub</code>.</p>
</section>
<section id="left-join-keep-all-rows-from-the-first-dataset" class="slide level2">
<h2>Left Join: Keep all rows from the first dataset</h2>
<ul>
<li>A <span class="primary"><strong>left join</strong></span> keeps all rows from the <span class="primary"><strong>first dataset</strong></span> (<code>cases_df_sub</code>), and adds matching data from the second dataset (<code>state_census</code>).</li>
<li>So <span class="primary"><strong>all rows from the first dataset</strong></span> (<code>cases_df_sub</code>) will be preserved.</li>
<li>The datasets are joined by matching the <code>geo_value</code> column, specified by the by argument.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a></a><span class="co"># Left join: combining March 1, 2022 state case data with the census data</span></span>
<span id="cb41-2"><a></a>cases_sub_left_join <span class="ot">&lt;-</span> cases_df_sub <span class="sc">|&gt;</span></span>
<span id="cb41-3"><a></a>  <span class="fu">left_join</span>(state_census, <span class="fu">join_by</span>(geo_value <span class="sc">==</span> abbr))</span>
<span id="cb41-4"><a></a></span>
<span id="cb41-5"><a></a>cases_sub_left_join</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  geo_value time_value raw_cases      pop
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;    &lt;dbl&gt;
1 ca        2022-03-01      4310 39512223
2 nc        2022-03-01      1231 10488084
3 ny        2022-03-01      1487 19453561</code></pre>
</div>
</div>
</section>
<section id="right-join-keep-all-rows-from-the-second-dataset" class="slide level2">
<h2>Right Join: Keep all rows from the second dataset</h2>
<ul>
<li>A <span class="primary"><strong>right join</strong></span> keeps all rows from the <span class="primary"><strong>second dataset</strong></span> (<code>state_census</code>), and adds matching data from the first dataset (<code>cases_df_sub</code>).</li>
<li>If a row in the second dataset doesn’t have a match in the first, then the columns from the first will be filled with NA.</li>
<li>For example, can see this for the <code>al</code> row from <code>state_census</code>…</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a></a><span class="co"># Right join: keep all rows from state_census</span></span>
<span id="cb43-2"><a></a>cases_sub_right_join <span class="ot">&lt;-</span> cases_df_sub <span class="sc">|&gt;</span></span>
<span id="cb43-3"><a></a>  <span class="fu">right_join</span>(state_census, <span class="fu">join_by</span>(geo_value <span class="sc">==</span> abbr))</span>
<span id="cb43-4"><a></a></span>
<span id="cb43-5"><a></a><span class="fu">head</span>(cases_sub_right_join)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  geo_value time_value raw_cases      pop
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;    &lt;dbl&gt;
1 ca        2022-03-01      4310 39512223
2 nc        2022-03-01      1231 10488084
3 ny        2022-03-01      1487 19453561
4 al        NA                NA  4903185
5 ak        NA                NA   731545
6 az        NA                NA  7278717</code></pre>
</div>
</div>
</section>
<section id="inner-join-only-keep-matching-rows" class="slide level2">
<h2>Inner Join: Only keep matching rows</h2>
<ul>
<li>An <span class="primary"><strong>inner join</strong></span> will only keep rows where there is a match in both datasets.</li>
<li>So, if a state in <code>state_census</code> does not have a corresponding entry in <code>cases_df_sub</code>, then that row will be excluded.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a></a><span class="co"># Inner join: only matching rows are kept</span></span>
<span id="cb45-2"><a></a>cases_sub_inner_join <span class="ot">&lt;-</span> cases_df_sub <span class="sc">|&gt;</span></span>
<span id="cb45-3"><a></a>  <span class="fu">inner_join</span>(state_census, <span class="fu">join_by</span>(geo_value <span class="sc">==</span> abbr))</span>
<span id="cb45-4"><a></a></span>
<span id="cb45-5"><a></a>cases_sub_inner_join</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  geo_value time_value raw_cases      pop
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;    &lt;dbl&gt;
1 ca        2022-03-01      4310 39512223
2 nc        2022-03-01      1231 10488084
3 ny        2022-03-01      1487 19453561</code></pre>
</div>
</div>
</section>
<section id="full-join-keep-all-rows-from-both-datasets" class="slide level2">
<h2>Full Join: Keep all rows from both datasets</h2>
<ul>
<li>A <span class="primary"><strong>full join</strong></span> will keep all rows from both datasets.</li>
<li>If a state in either dataset has no match in the other, the missing values will be filled with NA.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a></a><span class="co"># Full join: keep all rows from both datasets</span></span>
<span id="cb47-2"><a></a>cases_sub_full_join <span class="ot">&lt;-</span> cases_df_sub <span class="sc">|&gt;</span></span>
<span id="cb47-3"><a></a>  <span class="fu">full_join</span>(state_census, <span class="fu">join_by</span>(geo_value <span class="sc">==</span> abbr))</span>
<span id="cb47-4"><a></a></span>
<span id="cb47-5"><a></a><span class="fu">head</span>(cases_sub_full_join)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  geo_value time_value raw_cases      pop
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;    &lt;dbl&gt;
1 ca        2022-03-01      4310 39512223
2 nc        2022-03-01      1231 10488084
3 ny        2022-03-01      1487 19453561
4 al        NA                NA  4903185
5 ak        NA                NA   731545
6 az        NA                NA  7278717</code></pre>
</div>
</div>
</section>
<section id="pictorial-summary-of-the-four-join-functions" class="slide level2">
<h2>Pictorial summary of the four join functions</h2>
<!--* **Left join:** All rows from the left dataset and matching rows from the right dataset.
* **Right join:** All rows from the right dataset and matching rows from the left dataset.
* **Inner join:** Only matching rows from both datasets.
* **Full join:** All rows from both datasets, with NA where no match exists.-->
<img data-src="gfx/join_funs_cheatsheet.png" style="width: 40%; display: block; margin-left: auto; margin-right: auto;">
<div style="text-align: center;">
<p><small><a href="https://ohi-science.org/data-science-training/dplyr.html">Source</a></small></p>
</div>
<!-- ## Final thoughts on joins
* Joins are an essential part of data wrangling in R.
* The choice of join depends on the analysis you need to perform:
    + Use [**left joins**]{.primary} when you want to keep all data from the first dataset.
    + Use [**right joins**]{.primary} when you want to keep all data from the second dataset.
    + Use [**inner joins**]{.primary} when you're only interested in matching rows.
    + Use [**full joins**]{.primary} when you want to preserve all information from both datasets.
-->
</section>
<section id="two-review-questions" class="slide level2">
<h2>Two review questions</h2>
<p><strong>Q1)</strong>: What join function should you use if your goal is to scale the cases by population in <code>cases_df</code>?</p>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a></a><span class="co"># Either left_join</span></span>
<span id="cb49-2"><a></a>cases_left_join <span class="ot">&lt;-</span> cases_df <span class="sc">|&gt;</span></span>
<span id="cb49-3"><a></a>  <span class="fu">left_join</span>(state_census, <span class="fu">join_by</span>(geo_value <span class="sc">==</span> abbr))</span>
<span id="cb49-4"><a></a></span>
<span id="cb49-5"><a></a>cases_left_join</span>
<span id="cb49-6"><a></a></span>
<span id="cb49-7"><a></a><span class="co"># Or inner_join</span></span>
<span id="cb49-8"><a></a>cases_inner_join <span class="ot">&lt;-</span> cases_df <span class="sc">|&gt;</span></span>
<span id="cb49-9"><a></a>  <span class="fu">inner_join</span>(state_census, <span class="fu">join_by</span>(geo_value <span class="sc">==</span> abbr))</span>
<span id="cb49-10"><a></a></span>
<span id="cb49-11"><a></a>cases_inner_join</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p><strong>Q2)</strong>: Lastly, please create a new column in <code>cases_df</code> where you scale the cases by population and multiply by <code>1e5</code> to get cases / 100k.</p>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a></a>case_rates_df <span class="ot">&lt;-</span> cases_inner_join <span class="sc">|&gt;</span></span>
<span id="cb50-2"><a></a>  <span class="fu">mutate</span>(<span class="at">scaled_cases =</span> raw_cases <span class="sc">/</span> pop <span class="sc">*</span> <span class="fl">1e5</span>) <span class="co"># cases / 100K</span></span>
<span id="cb50-3"><a></a><span class="fu">head</span>(case_rates_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Congratulations for making it through this crash course! That’s all for this <code>glimpse()</code> into the tidyverse.</p>
</section>
<section id="epi.-data-processing-with-epiprocess" class="slide level2">
<h2>Epi. data processing with <code>epiprocess</code></h2>
<ul>
<li><code>epiprocess</code> is a package that offers additional functionality to pre-process such epidemiological data.</li>
<li>You can work with an <code>epi_df</code> like you can with a tibble by using dplyr verbs.</li>
<li>For example, on <code>cases_df</code>, we can easily use <code>epi_slide_mean()</code> to calculate trailing 14 day averages of cases:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a></a>case_rates_df <span class="ot">&lt;-</span> case_rates_df <span class="sc">|&gt;</span></span>
<span id="cb51-2"><a></a>  <span class="fu">as_epi_df</span>(<span class="at">as_of =</span> <span class="fu">as.Date</span>(<span class="st">"2024-01-01"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb51-3"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">|&gt;</span></span>
<span id="cb51-4"><a></a>  <span class="fu">epi_slide_mean</span>(scaled_cases, <span class="at">.window_size =</span> <span class="dv">14</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb51-5"><a></a>  <span class="fu">rename</span>(<span class="at">smoothed_scaled_cases =</span> slide_value_scaled_cases)</span>
<span id="cb51-6"><a></a><span class="fu">head</span>(case_rates_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An `epi_df` object, 6 x 6 with metadata:
* geo_type  = state
* time_type = day
* as_of     = 2024-01-01

# A tibble: 6 × 6
# Groups:   geo_value [1]
  geo_value time_value raw_cases      pop scaled_cases smoothed_scaled_cases
* &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;    &lt;dbl&gt;        &lt;dbl&gt;                 &lt;dbl&gt;
1 ca        2022-03-01      4310 39512223        10.9                   10.9
2 ca        2022-03-02      7044 39512223        17.8                   14.4
3 ca        2022-03-03      7509 39512223        19.0                   15.9
4 ca        2022-03-04      3586 39512223         9.08                  14.2
5 ca        2022-03-05      1438 39512223         3.64                  12.1
6 ca        2022-03-06      6465 39512223        16.4                   12.8</code></pre>
</div>
</div>
</section>
<section id="epi.-data-processing-with-epiprocess-1" class="slide level2">
<h2>Epi. data processing with <code>epiprocess</code></h2>
<p>It is easy to produce an autoplot the smoothed confirmed daily cases for each <code>geo_value</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a></a>case_rates_df <span class="sc">|&gt;</span></span>
<span id="cb53-2"><a></a>  <span class="fu">autoplot</span>(smoothed_scaled_cases)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="day1-afternoon_files/figure-revealjs/autoplot-ex-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="epi.-data-processing-with-epiprocess-2" class="slide level2">
<h2>Epi. data processing with <code>epiprocess</code></h2>
<p>Alternatively, we can display both the smoothed and the original daily case rates:</p>

<img data-src="day1-afternoon_files/figure-revealjs/smoothed-original-plot-1.svg" class="quarto-figure quarto-figure-center r-stretch"><p>Now, before exploring some more features of <code>epiprocess</code>, let’s have a look at the epiverse software ecosystem it’s part of…</p>
</section></section>
<section>
<section id="epiverse-software-ecosystem" class="title-slide slide level1 center" data-number="2">
<h1><span class="header-section-number">2</span> Epiverse Software Ecosystem</h1>

</section>
<section id="the-epiverse-ecosystem" class="slide level2">
<h2>The epiverse ecosystem</h2>
<p>Interworking, community-driven, packages for epi tracking &amp; forecasting.</p>

<!-- 1. Fetch data: epidatr, epidatpy, and other sources, 2. Explore, clean, transform & backtest 3. Pre-built forecasters, modular forecasting framework: epipredict -->
<img data-src="gfx/epiverse_packages_flow.jpg" style="width: 60%; display: block; margin-left: auto; margin-right: auto;" class="r-stretch"></section></section>
<section>
<section id="panel-and-versioned-data-in-the-epiverse" class="title-slide slide level1 center" data-number="3">
<h1><span class="header-section-number">3</span> Panel and Versioned Data in the Epiverse</h1>

</section>
<section id="what-is-panel-data" class="slide level2">
<h2>What is panel data?</h2>
<ul>
<li>Recall that <a href="https://en.wikipedia.org/wiki/Panel_data">panel data</a>, or longitudinal data, contain cross-sectional measurements of subjects over time.</li>
<li>Built-in example: <a href="https://cmu-delphi.github.io/epidatasets/reference/covid_case_death_rates.html"><code>covid_case_death_rates</code></a> dataset, which is a snapshot <span class="primary"><strong>as of</strong></span> May 31, 2022 that contains daily state-wise measures of <code>case_rate</code> and <code>death_rate</code> for COVID-19 over 2021:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  geo_value time_value case_rate death_rate
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 ak        2020-12-31      35.9      0.158
2 al        2020-12-31      65.1      0.438
3 ar        2020-12-31      66.0      1.27 
4 az        2020-12-31      76.8      1.10 
5 ca        2020-12-31      96.0      0.751
6 co        2020-12-31      35.8      0.649</code></pre>
</div>
</div>
<ul>
<li>How do we store &amp; work with such snapshots in the epiverse software ecosystem?</li>
</ul>
</section>
<section id="epi_df-snapshot-of-a-dataset" class="slide level2">
<h2><code>epi_df</code>: Snapshot of a dataset</h2>
<ul>
<li>You can convert panel data into an <code>epi_df</code> with the required <code>geo_value</code> and <code>time_value</code> columns</li>
</ul>
<p>Therefore, an <code>epi_df</code> is…</p>
<ul>
<li><p>a tibble that requires columns <code>geo_value</code> and <code>time_value</code>.</p></li>
<li><p>arbitrary additional columns containing <span class="primary">measured values</span></p></li>
<li><p>additional <span class="primary">keys</span> to index (<code>age_group</code>, <code>ethnicity</code>, etc.)</p></li>
</ul>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong><code>epi_df</code></strong></p>
</div>
<div class="callout-content">
<p>Represents a <span class="primary">snapshot</span> that contains the most <span class="primary">up-to-date values</span> of the signal variables, <span class="primary">as of</span> a given time.</p>
</div>
</div>
</div>
</section>
<section id="epi_df-snapshot-of-a-dataset-1" class="slide level2">
<h2><code>epi_df</code>: Snapshot of a dataset</h2>
<ul>
<li><p>Consider the same dataset we just encountered on JHU daily COVID-19 cases and deaths rates from all states <span class="primary">as of</span> May 31, 2022.</p></li>
<li><p>We can see that it meets the criteria <code>epi_df</code> (has <code>geo_value</code> and <code>time_value</code> columns) and that it contains additional metadata (i.e.&nbsp;<code>geo_type</code>, <code>time_type</code>, <code>as_of</code>, and <code>other_keys</code>).</p></li>
</ul>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a></a>edf <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An `epi_df` object, 6 x 4 with metadata:
* geo_type  = state
* time_type = day
* as_of     = 2022-05-31

# A tibble: 6 × 4
  geo_value time_value case_rate death_rate
* &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 ak        2020-12-31      35.9      0.158
2 al        2020-12-31      65.1      0.438
3 ar        2020-12-31      66.0      1.27 
4 as        2020-12-31       0        0    
5 az        2020-12-31      76.8      1.10 
6 ca        2020-12-31      96.0      0.751</code></pre>
</div>
</div>
</div><div class="column" style="width:40%;">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a></a><span class="fu">attr</span>(edf, <span class="st">"metadata"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$geo_type
[1] "state"

$time_type
[1] "day"

$as_of
[1] "2022-05-31"

$other_keys
character(0)</code></pre>
</div>
</div>
</div></div>
</section>
<section id="examples-of-preprocessing" class="slide level2">
<h2>Examples of preprocessing</h2>
<h3 id="eda-features">EDA features</h3>
<ol type="1">
<li>Making locations commensurate (per capita scaling)</li>
<li>Correlating signals across location or time</li>
<li>Computing growth rates</li>
<li>Detecting and removing outliers</li>
<li>Calculating summaries with rolling windows</li>
<li>Dealing with revisions</li>
</ol>
</section>
<section id="features---correlations-at-different-lags" class="slide level2">
<h2>Features - Correlations at different lags</h2>
<!-- * There are always at least two ways to compute correlations in an `epi_df`: grouping by `time_value`, and by `geo_value`. 

* The latter is obtained by setting `cor_by = geo_value`. -->
<ul>
<li><p>The below plot addresses the question: “For each state, are case and death rates linearly associated across all days?”</p></li>
<li><p>To explore <strong>lagged correlations</strong> and how case rates associate with future death rates, we can use the <code>dt1</code> parameter in <code>epi_cor()</code> to shift case rates by a specified number of days.</p></li>
</ul>
<!--  * For example, setting `dt1 = -14` means that case rates on June 1st will be correlated with death rates on June 15th, assessing how past case rates influence future death rates. -->
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a></a>cor0 <span class="ot">&lt;-</span> <span class="fu">epi_cor</span>(edf, case_rate, death_rate, <span class="at">cor_by =</span> geo_value)</span>
<span id="cb59-2"><a></a>cor14 <span class="ot">&lt;-</span> <span class="fu">epi_cor</span>(edf, case_rate, death_rate, <span class="at">cor_by =</span> geo_value, <span class="at">dt1 =</span> <span class="sc">-</span><span class="dv">14</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<img data-src="day1-afternoon_files/figure-revealjs/plot-corr-lags-ex-1.svg" class="quarto-figure quarto-figure-center r-stretch"><ul>
<li>We can see that, in general, lagging the case rates back by 14 days improves the correlations.</li>
</ul>
</section>
<section id="features---systematic-lag-analysis" class="slide level2">
<h2>Features - Systematic lag analysis</h2>
<p>The analysis helps identify the lag at which case rates from the past have the strongest correlation with future death rates.</p>

<img data-src="day1-afternoon_files/figure-revealjs/plot-sys-lag-ex-1.svg" class="quarto-figure quarto-figure-center r-stretch"><p>The strongest correlation occurs at a lag of about 23 days, indicating that case rates are best correlated with death rates 23 days from now.</p>
</section>
<section id="features---compute-growth-rates" class="slide level2">
<h2>Features - Compute growth rates</h2>
<ul>
<li><p>Growth rate measures the relative change in a signal over time. <!-- indicating how quickly a quantity (like case rates) is increasing or decreasing. --></p></li>
<li><p>We can compute time-varying growth rates for the two states, and see how this cases evolves over time.</p></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a></a>edfg <span class="ot">&lt;-</span> <span class="fu">filter</span>(edf, geo_value <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"ut"</span>, <span class="st">"ca"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb60-2"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">|&gt;</span></span>
<span id="cb60-3"><a></a>  <span class="fu">mutate</span>(<span class="at">gr_cases =</span> <span class="fu">growth_rate</span>(time_value, case_rate, <span class="at">method =</span> <span class="st">"trend_filter"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb60-4"><a></a>  <span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<img data-src="day1-afternoon_files/figure-revealjs/plot-growth-rates-ex-1.svg" class="quarto-figure quarto-figure-center r-stretch"><ul>
<li>As expected, the peak growth rates for both states occurred during the January 2022 Omicron wave, reflecting the sharp rise in cases over that period.</li>
</ul>
</section>
<section id="features---outlier-detection" class="slide level2">
<h2>Features - Outlier detection</h2>
<!-- * There are multiple outliers in these data that a modeler may want to detect and correct. -->
<ul>
<li><p>The <code>detect_outlr()</code> function offers multiple outlier detection methods on a signal.</p></li>
<li><p>The simplest is <code>detect_outlr_rm()</code>, which works by calculating an outlier threshold using the rolling median and the rolling Interquartile Range (IQR) for each time point:</p></li>
</ul>
<p><strong>Threshold = Rolling Median ± (Detection Multiplier × Rolling IQR)</strong></p>
<ul>
<li>Note that the default number of time steps to use in the rolling window by default is 21 and is centrally aligned.</li>
<li>The detection multiplier default is 2 and controls how far away a data point must be from the median to be considered an outlier.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a></a>edfo <span class="ot">&lt;-</span> <span class="fu">filter</span>(edf, geo_value <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"ca"</span>, <span class="st">"ut"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb61-2"><a></a>  <span class="fu">select</span>(geo_value, time_value, case_rate) <span class="sc">|&gt;</span></span>
<span id="cb61-3"><a></a>  <span class="fu">as_epi_df</span>() <span class="sc">|&gt;</span></span>
<span id="cb61-4"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">|&gt;</span></span>
<span id="cb61-5"><a></a>  <span class="fu">mutate</span>(<span class="at">outlier_info =</span> <span class="fu">detect_outlr_rm</span>(</span>
<span id="cb61-6"><a></a>    <span class="at">x =</span> time_value, <span class="at">y =</span> case_rate</span>
<span id="cb61-7"><a></a>  )) <span class="sc">|&gt;</span></span>
<span id="cb61-8"><a></a>  <span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="features---outlier-detection-1" class="slide level2">
<h2>Features - Outlier detection</h2>
<ul>
<li><p>Several data points that deviate from the expected case cadence have been flagged as outliers, and may require further investigation.</p></li>
<li><p>However, the peak in Jan.&nbsp;2022 has also been flagged as an outlier. This highlights the importance of manual inspection before correcting the data, as these may represent valid events (e.g., a genuine surge in cases).</p></li>
</ul>

<img data-src="day1-afternoon_files/figure-revealjs/plot-outlier-ex-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="features-sliding-a-computation-on-an-epi_df" class="slide level2">
<h2>Features – sliding a computation on an <code>epi_df</code></h2>
<ul>
<li><p>It is often useful to compute rolling summaries of signals.</p></li>
<li><p>These depend on the reference time, and are computed separately over geographies (and other groups).</p></li>
<li><p>For example, a trailing average can smooth out daily variation.</p></li>
<li><p>In <code>epiprocess</code>, this is achieved by <code>epi_slide()</code>.</p></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a></a><span class="fu">epi_slide</span>(</span>
<span id="cb62-2"><a></a>  .x,</span>
<span id="cb62-3"><a></a>  .f,</span>
<span id="cb62-4"><a></a>  ...,</span>
<span id="cb62-5"><a></a>  <span class="at">.window_size =</span> <span class="cn">NULL</span>,</span>
<span id="cb62-6"><a></a>  <span class="at">.align =</span> <span class="fu">c</span>(<span class="st">"right"</span>, <span class="st">"center"</span>, <span class="st">"left"</span>),</span>
<span id="cb62-7"><a></a>  <span class="at">.ref_time_values =</span> <span class="cn">NULL</span>,</span>
<span id="cb62-8"><a></a>  <span class="at">.new_col_name =</span> <span class="cn">NULL</span>,</span>
<span id="cb62-9"><a></a>  <span class="at">.all_rows =</span> <span class="cn">FALSE</span></span>
<span id="cb62-10"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For example, we can use <code>epi_slide()</code> to compute a trailing 7-day average.</p>
</section>
<section id="features-sliding-a-computation-on-an-epi_df-1" class="slide level2">
<h2>Features – sliding a computation on an <code>epi_df</code></h2>
<ul>
<li><p>The simplest way to use <code>epi_slide</code> is tidy evaluation.</p></li>
<li><p>For a grouped <code>epi_df</code>, <code>epi_slide()</code> applies the computation to groups <span class="primary">separately</span>.</p></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a></a>cases_7dav <span class="ot">&lt;-</span> <span class="fu">epi_slide</span>(</span>
<span id="cb63-2"><a></a>  <span class="at">.x =</span> cases_edf,</span>
<span id="cb63-3"><a></a>  <span class="at">cases_7dav =</span> <span class="fu">mean</span>(raw_cases, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb63-4"><a></a>  <span class="at">.window_size =</span> <span class="dv">7</span>,</span>
<span id="cb63-5"><a></a>  <span class="at">.align =</span> <span class="st">"right"</span></span>
<span id="cb63-6"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  geo_value time_value raw_cases cases_7dav
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 ca        2022-03-01      4310       4310
2 ca        2022-03-02      7044       5677
3 nc        2022-03-01      1231       1231
4 nc        2022-03-02      2243       1737
5 ny        2022-03-01      1487       1487
6 ny        2022-03-02      1889       1688</code></pre>
</div>
</div>
</section>
<section id="features-sliding-a-computation-on-an-epi_df-2" class="slide level2">
<h2>Features – sliding a computation on an <code>epi_df</code></h2>
<p><code>epi_slide</code> also accepts custom functions of a certain form.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a></a>custom_function <span class="ot">&lt;-</span> <span class="cf">function</span>(x, g, t, ...) {</span>
<span id="cb65-2"><a></a>  </span>
<span id="cb65-3"><a></a>  <span class="co"># Function body</span></span>
<span id="cb65-4"><a></a>  </span>
<span id="cb65-5"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><code>x</code>: the data frame with all the columns with original object <span class="primary">except</span> groupping vars.</li>
<li><code>g</code>: the one-row tibble with values of gropping vars of the given group.</li>
<li><code>t</code>: the <code>.ref_time_value</code> of the current window.</li>
<li><code>...</code>: additional arguments.</li>
</ul>
</section>
<section id="features-sliding-a-computation-on-an-epi_df-3" class="slide level2">
<h2>Features – sliding a computation on an <code>epi_df</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb66" data-code-line-numbers="|9-17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a></a>mean_by_hand <span class="ot">&lt;-</span> <span class="cf">function</span>(x, g, t, ...) {</span>
<span id="cb66-2"><a></a>  <span class="fu">data.frame</span>(<span class="at">cases_7dav =</span> <span class="fu">mean</span>(x<span class="sc">$</span>raw_cases, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb66-3"><a></a>}</span>
<span id="cb66-4"><a></a></span>
<span id="cb66-5"><a></a>cases_mean_custom_f <span class="ot">=</span> <span class="fu">epi_slide</span>(</span>
<span id="cb66-6"><a></a>    <span class="at">.x =</span> cases_edf,</span>
<span id="cb66-7"><a></a>    <span class="at">.f =</span> mean_by_hand,</span>
<span id="cb66-8"><a></a>    <span class="at">.window_size =</span> <span class="dv">7</span>,</span>
<span id="cb66-9"><a></a>    <span class="at">.align =</span> <span class="st">"right"</span></span>
<span id="cb66-10"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  geo_value time_value raw_cases cases_7dav
  &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 ca        2022-03-01      4310       4310
2 nc        2022-03-01      1231       1231
3 ny        2022-03-01      1487       1487</code></pre>
</div>
</div>
</section>
<section id="epi_archive-collection-of-epi_dfs" class="slide level2">
<h2><code>epi_archive</code>: Collection of <code>epi_df</code>s</h2>
<ul>
<li>full version history of a data set</li>
<li>acts like a bunch of <code>epi_df</code>s — but stored <span class="primary">compactly</span></li>
<li>allows similar functionality as <code>epi_df</code> but using only <span class="primary">data that would have been available at the time</span></li>
</ul>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Revisions</strong></p>
</div>
<div class="callout-content">
<p>Epidemiology data gets revised frequently.</p>
<ul>
<li>We may want to use the data [as it looked in the past].{.primary}</li>
<li>or we may want to examine <span class="primary">the history of revisions</span>.</li>
</ul>
</div>
</div>
</div>
</section>
<section id="epi_archive-collection-of-epi_dfs-1" class="slide level2">
<h2><code>epi_archive</code>: Collection of <code>epi_df</code>s</h2>
<p>Subset of daily COVID-19 doctor visits (Optum) and cases (JHU CSSE) from all U.S. states in <code>archive</code> format:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a></a>archive_cases_dv_subset_all_states <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$DT
Key: &lt;geo_value, time_value, version&gt;
         geo_value time_value    version percent_cli case_rate_7d_av
            &lt;char&gt;     &lt;Date&gt;     &lt;Date&gt;       &lt;num&gt;           &lt;num&gt;
      1:        ak 2020-06-01 2020-06-02          NA        1.145652
      2:        ak 2020-06-01 2020-06-06    0.136815        1.145652
      3:        ak 2020-06-01 2020-06-08    0.136249        1.145652
      4:        ak 2020-06-01 2020-06-09    0.106744        1.145652
      5:        ak 2020-06-01 2020-06-10    0.106676        1.145652
     ---                                                            
1514485:        wy 2021-11-26 2021-11-29    3.739819       23.207343
1514486:        wy 2021-11-27 2021-11-28          NA       23.207343
1514487:        wy 2021-11-28 2021-11-29          NA       23.207343
1514488:        wy 2021-11-29 2021-11-30          NA       25.071781
1514489:        wy 2021-11-30 2021-12-01          NA       25.464294

$geo_type
[1] "state"

$time_type
[1] "day"

$other_keys
character(0)

$clobberable_versions_start
[1] NA

$versions_end
[1] "2021-12-01"</code></pre>
</div>
</div>
</section>
<section id="features-sliding-computation-over-epi_dfs" class="slide level2">
<h2>Features – sliding computation over <code>epi_df</code>s</h2>
<ul>
<li>We can apply a computation over different snapshots in an <code>epi_archive</code>.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a></a><span class="fu">epix_slide</span>(</span>
<span id="cb70-2"><a></a>  .x,</span>
<span id="cb70-3"><a></a>  .f,</span>
<span id="cb70-4"><a></a>  ...,</span>
<span id="cb70-5"><a></a>  <span class="at">.before =</span> <span class="cn">Inf</span>,</span>
<span id="cb70-6"><a></a>  <span class="at">.versions =</span> <span class="cn">NULL</span>,</span>
<span id="cb70-7"><a></a>  <span class="at">.new_col_name =</span> <span class="cn">NULL</span>,</span>
<span id="cb70-8"><a></a>  <span class="at">.all_versions =</span> <span class="cn">FALSE</span></span>
<span id="cb70-9"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This functionality is very helpful in version aware forecasting. We will return with a concrete example.</p>
</section>
<section id="features-summarize-revision-behavior" class="slide level2">
<h2>Features – summarize revision behavior</h2>
<ul>
<li><code>revision_summary()</code> is a helper function that summarizes revision behavior of an <code>epix_archive</code>.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a></a>revision_data <span class="ot">&lt;-</span> <span class="fu">revision_summary</span>(</span>
<span id="cb71-2"><a></a>  archive_cases_dv_subset,</span>
<span id="cb71-3"><a></a>  case_rate_7d_av,</span>
<span id="cb71-4"><a></a>  <span class="at">drop_nas =</span> <span class="cn">TRUE</span>,</span>
<span id="cb71-5"><a></a>  <span class="at">print_inform =</span> <span class="cn">FALSE</span>, <span class="co"># NOT the default, to save space</span></span>
<span id="cb71-6"><a></a>  <span class="at">min_waiting_period =</span> <span class="fu">as.difftime</span>(<span class="dv">60</span>, <span class="at">units =</span> <span class="st">"days"</span>),</span>
<span id="cb71-7"><a></a>  <span class="at">within_latest =</span> <span class="fl">0.2</span>,</span>
<span id="cb71-8"><a></a>  <span class="at">quick_revision =</span> <span class="fu">as.difftime</span>(<span class="dv">3</span>, <span class="at">units =</span> <span class="st">"days"</span>),</span>
<span id="cb71-9"><a></a>  <span class="at">few_revisions =</span> <span class="dv">3</span>,</span>
<span id="cb71-10"><a></a>  <span class="at">abs_spread_threshold =</span> <span class="cn">NULL</span>,</span>
<span id="cb71-11"><a></a>  <span class="at">rel_spread_threshold =</span> <span class="fl">0.1</span>,</span>
<span id="cb71-12"><a></a>  <span class="at">compactify_tol =</span> .Machine<span class="sc">$</span>double.eps<span class="sc">^</span><span class="fl">0.5</span>,</span>
<span id="cb71-13"><a></a>  <span class="at">should_compactify =</span> <span class="cn">TRUE</span></span>
<span id="cb71-14"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="features-summarize-revision-behavior-1" class="slide level2">
<h2>Features – summarize revision behavior</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a></a><span class="fu">head</span>(revision_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 11
  time_value geo_value n_revisions min_lag max_lag  time_near_latest spread
  &lt;date&gt;     &lt;chr&gt;           &lt;dbl&gt; &lt;drtn&gt;  &lt;drtn&gt;   &lt;drtn&gt;            &lt;dbl&gt;
1 2020-06-01 ca                 12 1 days  546 days 1 days           0.248 
2 2020-06-02 ca                 12 1 days  545 days 1 days           0.416 
3 2020-06-03 ca                 11 1 days  544 days 1 days           0.115 
4 2020-06-04 ca                 11 1 days  543 days 1 days           0.342 
5 2020-06-05 ca                  7 1 days  520 days 1 days           0.0982
6 2020-06-06 ca                  8 1 days  519 days 1 days           0.188 
# ℹ 4 more variables: rel_spread &lt;dbl&gt;, min_value &lt;dbl&gt;, max_value &lt;dbl&gt;,
#   median_value &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="visualize-revision-patterns" class="slide level2">
<h2>Visualize revision patterns</h2>

<img data-src="day1-afternoon_files/figure-revealjs/plot-revision-patterns-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="finalized-data" class="slide level2">
<h2>Finalized data</h2>
<ul>
<li><p>Counts are revised as time proceeds</p></li>
<li><p>Want to know the <span class="primary">final</span> value</p></li>
<li><p>Often not available until weeks/months later</p>
<dl>
<dt>Forecasting</dt>
<dd>
At time <span class="math inline">\(t\)</span>, predict the final value for time <span class="math inline">\(t+h\)</span>, <span class="math inline">\(h &gt; 0\)</span>
</dd>
</dl>
<p><br></p>
<dl>
<dt>Backcasting</dt>
<dd>
At time <span class="math inline">\(t\)</span>, predict the final value for time <span class="math inline">\(t-h\)</span>, <span class="math inline">\(h &lt; 0\)</span>
</dd>
</dl>
<p><br></p>
<dl>
<dt>Nowcasting</dt>
<dd>
At time <span class="math inline">\(t\)</span>, predict the final value for time <span class="math inline">\(t\)</span>
</dd>
</dl></li>
</ul>
</section></section>
<section>
<section id="basic-nowcasting-in-the-epiverse" class="title-slide slide level1 center" data-number="4">
<h1><span class="header-section-number">4</span> Basic Nowcasting in the Epiverse</h1>
<!-- predicting a finalized value from a provisional value and making predictions. -->
</section>
<section id="backfill-canadian-edition" class="slide level2">
<h2>Backfill Canadian edition</h2>
<ul>
<li><p>Every week the <a href="http://www.bccdc.ca/health-info/diseases-conditions/covid-19/archived-b-c-covid-19-data">BC CDC releases COVID-19 hospitalization data</a>.</p></li>
<li><p>Following week they revise the number upward (by ~25%) due to lagged reports.</p></li>
</ul>
<p><img data-src="gfx/bc_hosp_admissions_ex.jpg" style="width: 45%; display: block; margin-left: auto; margin-right: auto;"> <!-- Newest iteration of "backfill”, Canada edition. Every week the BC CDC releases hospitalization data. The following week they revise the number upward (by about 25%) due to lagging reports. Every single week, the newspaper says “hospitalizations have declined”. This week the BC CDC’s own report said “hospitalizations have declined”. The takeaway in the news is that hospitalizations ALWAYS fall from the previous week, but once backfilled, they’re rarely down --></p>
<ul>
<li><span class="primary"><strong>Takeaway</strong></span>: Once the data is backfilled, hospitalizations rarely show a decline, challenging the common media narrative.</li>
</ul>
</section>
<section id="backfill-american-edition" class="slide level2">
<h2>Backfill American edition</h2>
<ul>
<li><p>Again, we can see a similar systematic underestimation problem for COVID-19 morality rates in CA. <!-- 2023-2024 --></p></li>
<li><p>This plot also illustrates the <span class="primary"><strong>revision process</strong></span> - how the reported mortality changes &amp; increases across multiple updates until it stabilizes at the final value (black line).</p></li>
</ul>

<img data-src="gfx/am_mortality_revisions_ex.jpg" style="width: 45%; display: block; margin-left: auto; margin-right: auto;" class="r-stretch"><ul>
<li>These two examples show the problem and now we need a solution…</li>
</ul>
</section>
<section id="nowcasting-and-its-mathematical-setup" class="slide level2">
<h2>Nowcasting and its mathematical setup</h2>
<ul>
<li><p><strong>Nowcasting</strong>: Predict a finalized value from a provisional value.</p></li>
<li><p>Suppose today is time <span class="math inline">\(t\)</span></p></li>
<li><p>Let <span class="math inline">\(y_i\)</span> denote a series of interest observed at times <span class="math inline">\(i=1,\ldots, t\)</span>.</p></li>
</ul>
<div class="callout callout-important no-icon callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<p><strong>Our goal</strong></p>
</div>
<div class="callout-content">
<ul>
<li>Produce a <span class="primary"><strong>point nowcast</strong></span> for the finalized values of <span class="math inline">\(y_t\)</span>.</li>
<li>Accompany with time-varying prediction intervals</li>
</ul>
</div>
</div>
</div>
<ul>
<li>We may also have access to <span class="math inline">\(p\)</span> other time series <span class="math inline">\(x_{ij},\; i=1,\ldots,t, \; j = 1,\ldots, p\)</span> which may be subject to revisions.</li>
</ul>
</section>
<section id="case-study-nchs-mortality" class="slide level2">
<h2>Case study: NCHS mortality</h2>
<ul>
<li>In this example, we’ll demonstrate the concept of nowcasting using <span class="primary"><strong>NHCS mortality data</strong></span>. (the number of weekly new deaths with confirmed or presumed COVID-19, per 100,000 population).</li>
<li>We will work with <span class="primary"><strong>provisional</strong></span> data (real-time reports) and compare them to <strong>finalized</strong> data (final reports).</li>
<li>The goal is to estimate or <span class="primary"><strong>nowcast the mortality rate</strong></span> for weeks when only provisional data is available.</li>
</ul>
</section>
<section id="fetch-versioned-data" class="slide level2">
<h2>Fetch versioned data</h2>
<p>Let’s fetch versioned mortality data from the API (<code>pub_covidcast</code>) for CA (<code>geo_values = "ca"</code>) and the signal of interest (<code>deaths_covid_incidence_num</code>) over early 2024.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a></a><span class="co"># Fetch the versioned NCHS mortality data (weekly)</span></span>
<span id="cb74-2"><a></a>nchs_archive <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb74-3"><a></a>  <span class="at">source =</span> <span class="st">"nchs-mortality"</span>,</span>
<span id="cb74-4"><a></a>  <span class="at">signals =</span> <span class="st">"deaths_covid_incidence_num"</span>,</span>
<span id="cb74-5"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb74-6"><a></a>  <span class="at">time_type =</span> <span class="st">"week"</span>,</span>
<span id="cb74-7"><a></a>  <span class="at">geo_values =</span> <span class="fu">c</span>(<span class="st">"ca"</span>, <span class="st">"ut"</span>),  </span>
<span id="cb74-8"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">202001</span>, <span class="dv">202440</span>),  </span>
<span id="cb74-9"><a></a>  <span class="at">issues =</span> <span class="st">"*"</span></span>
<span id="cb74-10"><a></a>) <span class="sc">|&gt;</span> </span>
<span id="cb74-11"><a></a>  <span class="fu">select</span>(geo_value, time_value, <span class="at">version =</span> issue, <span class="at">mortality =</span> value) <span class="sc">|&gt;</span> </span>
<span id="cb74-12"><a></a>  <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="analysis-of-versioning-behavior" class="slide level2">
<h2>Analysis of versioning behavior</h2>
<p>Recall, we need to watch out for:</p>
<ul>
<li><span class="primary"><strong>Latency</strong></span> the time difference between date of reference and date of the initial report</li>
<li><span class="primary"><strong>Backfill</strong></span> how data for a given date is updated after initial report.</li>
</ul>
<p><code>revision_summary()</code> provides a sumamry to both aspects.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a></a>revision_data <span class="ot">=</span> <span class="fu">revision_summary</span>(nchs_archive, mortality, <span class="at">print_inform =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="versioning-analysis-latency" class="slide level2">
<h2>Versioning analysis – latency</h2>
<ul>
<li><span class="primary"><strong>Question:</strong></span> What is the latency of NCHS data?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a></a>revision_data <span class="sc">|&gt;</span> <span class="fu">select</span>(geo_value, time_value, min_lag) <span class="sc">|&gt;</span> <span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 3
   geo_value time_value min_lag
   &lt;chr&gt;     &lt;date&gt;     &lt;drtn&gt; 
 1 ca        2022-08-07  7 days
 2 ca        2023-10-01  7 days
 3 ca        2024-04-21 14 days
 4 ut        2022-12-25 14 days
 5 ut        2020-09-27 70 days
 6 ut        2022-07-31 21 days
 7 ut        2023-07-09  7 days
 8 ca        2023-05-07  7 days
 9 ca        2020-09-27 70 days
10 ut        2022-04-17  7 days</code></pre>
</div>
</div>
<ul>
<li>We randomly sampled some dates to check if there is a consistent latency pattern.</li>
<li>Understanding latency prevents us from using data that we shouldn’t have access to.</li>
</ul>
</section>
<section id="versioning-analysis-backfill" class="slide level2">
<h2>Versioning analysis – backfill</h2>
<ul>
<li><span class="primary"><strong>Question:</strong></span> How long does it take for the reported value to be close to the finalized value?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a></a>revision_data <span class="sc">|&gt;</span> <span class="fu">select</span>(geo_value, time_value, time_near_latest) <span class="sc">|&gt;</span> <span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 3
   geo_value time_value time_near_latest
   &lt;chr&gt;     &lt;date&gt;     &lt;drtn&gt;          
 1 ut        2021-11-14  28 days        
 2 ut        2021-03-28  28 days        
 3 ut        2021-02-21  28 days        
 4 ca        2022-07-17  28 days        
 5 ca        2022-10-30  28 days        
 6 ut        2020-08-30  98 days        
 7 ca        2021-08-29  28 days        
 8 ut        2022-03-27   7 days        
 9 ca        2020-12-06  70 days        
10 ca        2020-05-03 217 days        </code></pre>
</div>
</div>
<ul>
<li>It generally takes at least 4 weeks for reported value to be within 20% (default in <code>revision_summary()</code>) of the finalized value.</li>
<li>We can change the threshold of percentage difference by specifying the <code>within_latest</code> argument of <code>revision_summary()</code>.</li>
</ul>
</section>
<section id="versioning-analysis---backfill" class="slide level2">
<h2>Versioning analysis - backfill</h2>
<ul>
<li><span class="primary"><strong>Question:</strong></span> When is the <span class="primary"><strong>finalized value</strong></span> first attained for each date? Would we have access to any in real-time?</li>
<li>How fast are the final values attained &amp; what’s the pattern for these times, if any?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  geo_value time_value min_version diff    
  &lt;chr&gt;     &lt;date&gt;     &lt;date&gt;      &lt;drtn&gt;  
1 ca        2020-01-26 2020-12-06  315 days
2 ca        2020-02-09 2020-12-06  301 days
3 ca        2020-02-23 2020-12-06  287 days
4 ca        2020-03-15 2020-12-06  266 days
5 ca        2020-03-22 2021-05-16  420 days
6 ca        2020-03-29 2021-04-04  371 days</code></pre>
</div>
</div>
<ul>
<li><span class="primary"><strong>Conclusion</strong></span>: The revision behavior is pretty long-tailed. Value reported 4 weeks later is reasonably close to the finalized value.</li>
</ul>
</section>
<section id="revision-pattern-visualization" class="slide level2">
<h2>Revision pattern visualization</h2>
<p>This shows the finalized rates in comparison to <span class="primary"><strong>multiple revisions</strong></span> to see how the data changes over time:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="day1-afternoon_files/figure-revealjs/final-vs-revisions-plot-1.svg" class="quarto-figure quarto-figure-center" height="500"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="revision-pattern-visualization-1" class="slide level2">
<h2>Revision pattern visualization</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="day1-afternoon_files/figure-revealjs/nchs-plot-val-different-ver-1.svg" class="quarto-figure quarto-figure-center" height="500"></p>
</figure>
</div>
</div>
</div>
<!--
## Aside: Do we need a burn-in/training set? 

* Typical stat-ML practise suggests a train, validation, test split.
* But our exploratory analysis covered all available data, is that fine?


* Generally, for exploratory analysis, it is fine to not do train/test split. 
  + These analyses do not involve model fitting, we have little risk of an overly optimistic performance evaluation (no overfitting on test data).
* However, for a [**psuedo-prospective analysis**]{.primary}, the best practise is to do a train/test split.
  + In such analysis, one would be fitting and validating many models, a train/test split provides a more rigorous control on overfitting to test set. 
-->
</section>
<section id="ratio-nowcaster-jumping-from-provisional-to-finalized-value" class="slide level2">
<h2>Ratio nowcaster: jumping from provisional to finalized value</h2>
<ul>
<li>Recall, the goal of nowcast at date <span class="math inline">\(t\)</span> is to use project the <span class="primary"><em>finalized value</em></span> of <span class="math inline">\(y_t,\)</span> given the information available on date <span class="math inline">\(t\)</span>.</li>
<li>A very simple nowcaster is the ratio between finalized and provisional value.</li>
</ul>
<p>How can we sensibly estimate this quantity?</p>
</section>
<section id="ratio-nowcaster-building-training-samples" class="slide level2">
<h2>Ratio nowcaster: building training samples</h2>
<ul>
<li>At nowcast date <span class="math inline">\(t,\)</span> would have received reports with versions up to and including <span class="math inline">\(t.\)</span></li>
<li>We need to build training samples, which
<ul>
<li>correctly aligns finalized value against provisional value</li>
<li>uses features that would have been available at test time</li>
<li>have enough samples to ensure sensible estimation results</li>
</ul></li>
</ul>
<div class="fragment fade-in">
<ul>
<li>Build training samples by treating dates prior to date <span class="math inline">\(t\)</span> as actual nowcast dates.
<ul>
<li>What is the provisional data on that date?</li>
<li>Have we received finalized value for that date?</li>
</ul></li>
</ul>
</div>
</section>
<section id="ratio-nowcaster-building-training-samples-1" class="slide level2">
<h2>Ratio nowcaster: building training samples</h2>
<ul>
<li>At an earlier nowcast date <span class="math inline">\(t_0,\)</span> we define
<ul>
<li><span class="primary"><strong>Provisional value</strong></span> as the reported value of <span class="math inline">\(Y_s\)</span> with version <span class="math inline">\(t_0.\)</span> Here <span class="math inline">\(s_0\)</span> is the largest occurence date among all values reported up until <span class="math inline">\(t_0.\)</span></li>
<li><span class="primary"><strong>Finalized value</strong></span> as the (potentially unobserved) finalized value of <span class="math inline">\(Y_{s_0}.\)</span>
<ul>
<li>We only know in <em>hindsight</em> when reported value of <span class="math inline">\(Y_{s_0}\)</span> is finalized – need an approximation.</li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="revisiting-revision_summary" class="slide level2">
<h2>Revisiting <code>revision_summary()</code></h2>
<p>Recall, <code>revision_summary()</code> reports the number of days to be within 20% (default value) of finalized value</p>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  geo_value time_value time_near_latest
  &lt;chr&gt;     &lt;date&gt;     &lt;drtn&gt;          
1 ca        2022-08-07 21 days         
2 ca        2023-10-01 28 days         
3 ca        2024-04-21 42 days         
4 ut        2022-12-25 21 days         
5 ut        2020-09-27 70 days         </code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Time differences in days
  0%  25%  50%  75% 100% 
   7   21   28   49  315 </code></pre>
</div>
</div>
<div class="fragment fade-in">
<p>Let’s say data reported 49 days after reference date is good enough to be considered finalized.</p>
</div>
</section>
<section id="ratio-nowcaster-test-time-feature" class="slide level2">
<h2>Ratio nowcaster: test time feature</h2>
<ul>
<li>Due to latency, provisional values may not be available at lag 0</li>
<li>We use last-observation-carried-forward (LOCF) to impute missing values at test time</li>
<li>Precisely, at test time <span class="math inline">\(t,\)</span> we use last observed data point (among all those reported up through time <span class="math inline">\(t\)</span>)</li>
</ul>
</section>
<section id="nowcasting-at-a-single-date-building-training-samples" class="slide level2">
<h2>Nowcasting at a single date: building training samples</h2>
<div class="fragment fade-in">
<ul>
<li>Searching for provisional values, at previous hypothetical nowcast dates.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a></a>nowcast_date <span class="ot">&lt;-</span> <span class="fu">as.Date</span>(<span class="st">"2022-01-02"</span>); window_length <span class="ot">=</span> <span class="dv">180</span></span>
<span id="cb83-2"><a></a></span>
<span id="cb83-3"><a></a>initial_data <span class="ot">&lt;-</span> nchs_archive<span class="sc">$</span>DT <span class="sc">|&gt;</span></span>
<span id="cb83-4"><a></a>  <span class="fu">group_by</span>(geo_value, time_value) <span class="sc">|&gt;</span></span>
<span id="cb83-5"><a></a>  <span class="fu">filter</span>(version <span class="sc">==</span> <span class="fu">min</span>(version)) <span class="sc">|&gt;</span></span>
<span id="cb83-6"><a></a>  <span class="fu">rename</span>(<span class="at">initial_val =</span> mortality) <span class="sc">|&gt;</span></span>
<span id="cb83-7"><a></a>  <span class="fu">select</span>(geo_value, time_value, initial_val)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment fade-in">
<ul>
<li>Searching for finalized values, at previous hypothetical nowcast dates.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a></a>finalized_data <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(nchs_archive, nowcast_date) <span class="sc">|&gt;</span></span>
<span id="cb84-2"><a></a>  <span class="fu">filter</span>(time_value <span class="sc">&gt;=</span> nowcast_date <span class="sc">-</span> approx_final_lag <span class="sc">-</span> window_length <span class="sc">&amp;</span> time_value <span class="sc">&lt;=</span> nowcast_date <span class="sc">-</span> approx_final_lag) <span class="sc">|&gt;</span></span>
<span id="cb84-3"><a></a>  <span class="fu">rename</span>(<span class="at">finalized_val =</span> mortality) <span class="sc">|&gt;</span></span>
<span id="cb84-4"><a></a>  <span class="fu">select</span>(geo_value, time_value, finalized_val)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="nowcasting-at-a-single-date-estimating-ratio-model" class="slide level2">
<h2>Nowcasting at a single date: estimating ratio model</h2>
<ul>
<li>After searching for both provisional and finalized values, we merge them together and estimate the ratio.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a></a>ratio <span class="ot">&lt;-</span> finalized_data <span class="sc">|&gt;</span></span>
<span id="cb85-2"><a></a>  <span class="fu">inner_join</span>(initial_data, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"geo_value"</span>, <span class="st">"time_value"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb85-3"><a></a>  <span class="fu">mutate</span>(<span class="at">ratio =</span> finalized_val <span class="sc">/</span> initial_val) <span class="sc">|&gt;</span></span>
<span id="cb85-4"><a></a>  <span class="fu">pull</span>(ratio) <span class="sc">|&gt;</span></span>
<span id="cb85-5"><a></a>  <span class="fu">median</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="nowcasting-at-a-single-date-test-feature-construction" class="slide level2">
<h2>Nowcasting at a single date: test feature construction</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a></a>last_avail <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(nchs_archive, nowcast_date) <span class="sc">|&gt;</span></span>
<span id="cb86-2"><a></a>  <span class="fu">slice_max</span>(time_value) <span class="sc">|&gt;</span></span>
<span id="cb86-3"><a></a>  <span class="fu">pull</span>(mortality) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="nowcasting-at-a-single-date-producing-the-nowcast" class="slide level2">
<h2>Nowcasting at a single date: producing the nowcast</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a></a>nowcast <span class="ot">&lt;-</span> last_avail <span class="sc">*</span> ratio</span>
<span id="cb87-2"><a></a>finalized_val <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(nchs_archive, nchs_archive<span class="sc">$</span>versions_end) <span class="sc">|&gt;</span></span>
<span id="cb87-3"><a></a>  <span class="fu">filter</span>(time_value <span class="sc">==</span> nowcast_date) <span class="sc">|&gt;</span></span>
<span id="cb87-4"><a></a><span class="fu">pull</span>(mortality)</span>
<span id="cb87-5"><a></a></span>
<span id="cb87-6"><a></a>nowcast_final <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">Nowcast =</span> nowcast, <span class="st">`</span><span class="at">Finalized value</span><span class="st">`</span> <span class="ot">=</span> finalized_val, <span class="at">check.names=</span><span class="cn">FALSE</span>)</span>
<span id="cb87-7"><a></a>knitr<span class="sc">::</span><span class="fu">kable</span>(nowcast_final)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top">
<thead>
<tr class="header">
<th style="text-align: right;">Nowcast</th>
<th style="text-align: right;">Finalized value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">305.1608</td>
<td style="text-align: right;">946</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="nowcasting-for-multiple-dates" class="slide level2">
<h2>Nowcasting for multiple dates</h2>
<ul>
<li>All previous manipulations should really be seen as a template for all nowcast dates.</li>
<li>The template computation sould be applied over all nowcast dates, <span class="primary"><strong>but we must respect data versioning</strong></span>!</li>
<li><code>epix_slide()</code> is designed just for this! It behaves similarly to <code>epi_slide</code>.</li>
<li>Key exception: <code>epix_slide()</code> is version aware: the sliding computation at any reference time <span class="math inline">\(t\)</span> is performed on <span class="primary"><strong>data that would have been available as of t</strong></span>.</li>
</ul>
</section>
<section id="nowcasting-for-multiple-dates-via-epix_slide" class="slide level2">
<h2>Nowcasting for multiple dates via <code>epix_slide()</code></h2>
<p>We begin by templatizing our previous operations.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a></a>nowcaster <span class="ot">=</span> <span class="cf">function</span>(x, g, t, <span class="at">wl=</span><span class="dv">180</span>, <span class="at">appx=</span>approx_final_lag) {</span>
<span id="cb88-2"><a></a>  </span>
<span id="cb88-3"><a></a></span>
<span id="cb88-4"><a></a>  initial_data <span class="ot">=</span> x<span class="sc">$</span>DT <span class="sc">|&gt;</span></span>
<span id="cb88-5"><a></a>    <span class="fu">group_by</span>(geo_value, time_value) <span class="sc">|&gt;</span></span>
<span id="cb88-6"><a></a>    <span class="fu">filter</span>(version <span class="sc">==</span>  <span class="fu">min</span>(version)) <span class="sc">|&gt;</span></span>
<span id="cb88-7"><a></a>    <span class="fu">filter</span>(time_value <span class="sc">&gt;=</span> t <span class="sc">-</span> wl <span class="sc">-</span> appx <span class="sc">&amp;</span> time_value <span class="sc">&lt;=</span> t <span class="sc">-</span> appx) <span class="sc">|&gt;</span></span>
<span id="cb88-8"><a></a>    <span class="fu">rename</span>(<span class="at">initial_val =</span> mortality) <span class="sc">|&gt;</span></span>
<span id="cb88-9"><a></a>    <span class="fu">select</span>(geo_value, time_value, initial_val)</span>
<span id="cb88-10"><a></a></span>
<span id="cb88-11"><a></a>  finalized_data <span class="ot">=</span> x<span class="sc">$</span>DT <span class="sc">|&gt;</span></span>
<span id="cb88-12"><a></a>    <span class="fu">group_by</span>(geo_value, time_value) <span class="sc">|&gt;</span></span>
<span id="cb88-13"><a></a>    <span class="fu">filter</span>(version <span class="sc">==</span>  <span class="fu">max</span>(version)) <span class="sc">|&gt;</span></span>
<span id="cb88-14"><a></a>    <span class="fu">filter</span>(time_value <span class="sc">&gt;=</span> t <span class="sc">-</span> wl <span class="sc">-</span> appx <span class="sc">&amp;</span> time_value <span class="sc">&lt;=</span> t <span class="sc">-</span> appx) <span class="sc">|&gt;</span></span>
<span id="cb88-15"><a></a>    <span class="fu">rename</span>(<span class="at">finalized_val =</span> mortality) <span class="sc">|&gt;</span></span>
<span id="cb88-16"><a></a>    <span class="fu">select</span>(geo_value, time_value, finalized_val)</span>
<span id="cb88-17"><a></a>  </span>
<span id="cb88-18"><a></a>  ratio <span class="ot">=</span> finalized_data <span class="sc">|&gt;</span></span>
<span id="cb88-19"><a></a>    <span class="fu">inner_join</span>(initial_data, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"geo_value"</span>, <span class="st">"time_value"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb88-20"><a></a>    <span class="fu">mutate</span>(<span class="at">ratio =</span> finalized_val <span class="sc">/</span> initial_val) <span class="sc">|&gt;</span></span>
<span id="cb88-21"><a></a>    <span class="fu">pull</span>(ratio) <span class="sc">|&gt;</span></span>
<span id="cb88-22"><a></a>    <span class="fu">median</span>(<span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb88-23"><a></a></span>
<span id="cb88-24"><a></a>  last_avail <span class="ot">=</span> <span class="fu">epix_as_of</span>(x, t) <span class="sc">|&gt;</span></span>
<span id="cb88-25"><a></a>    <span class="fu">slice_max</span>(time_value) <span class="sc">|&gt;</span></span>
<span id="cb88-26"><a></a>    <span class="fu">pull</span>(mortality) </span>
<span id="cb88-27"><a></a></span>
<span id="cb88-28"><a></a></span>
<span id="cb88-29"><a></a>  res <span class="ot">=</span> <span class="fu">tibble</span>(<span class="at">geo_value =</span> x<span class="sc">$</span>geo_value, <span class="at">target_date =</span> t, <span class="at">nowcast =</span> last_avail <span class="sc">*</span> ratio)</span>
<span id="cb88-30"><a></a>  </span>
<span id="cb88-31"><a></a>  <span class="fu">return</span>(res)</span>
<span id="cb88-32"><a></a>  </span>
<span id="cb88-33"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sanity-check-of-epix_slide" class="slide level2">
<h2>Sanity check of <code>epix_slide()</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a></a>slided_nowcast_1d <span class="ot">=</span> <span class="fu">epix_slide</span>(</span>
<span id="cb89-2"><a></a>  <span class="at">.x =</span> nchs_archive,</span>
<span id="cb89-3"><a></a>  <span class="at">.f =</span> nowcaster,</span>
<span id="cb89-4"><a></a>  <span class="at">.before =</span> <span class="cn">Inf</span>,</span>
<span id="cb89-5"><a></a>  <span class="at">.versions =</span> nowcast_date,</span>
<span id="cb89-6"><a></a>  <span class="at">.all_versions =</span> <span class="cn">TRUE</span></span>
<span id="cb89-7"><a></a>)</span>
<span id="cb89-8"><a></a></span>
<span id="cb89-9"><a></a>nowcast_check <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="st">`</span><span class="at">Manual nowcast</span><span class="st">`</span> <span class="ot">=</span> nowcast, <span class="st">`</span><span class="at">Slided nowcast</span><span class="st">`</span> <span class="ot">=</span> slided_nowcast_1d<span class="sc">$</span>nowcast, <span class="at">check.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb89-10"><a></a>knitr<span class="sc">::</span><span class="fu">kable</span>(nowcast_check)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top">
<thead>
<tr class="header">
<th style="text-align: right;">Manual nowcast</th>
<th style="text-align: right;">Slided nowcast</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">305.1608</td>
<td style="text-align: right;">305.1608</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="nowcasting-for-multiple-dates-via-epix_slide-1" class="slide level2">
<h2>Nowcasting for multiple dates via <code>epix_slide()</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a></a>nowcasts <span class="ot">=</span> nchs_archive <span class="sc">|&gt;</span></span>
<span id="cb90-2"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">|&gt;</span></span>
<span id="cb90-3"><a></a>  <span class="fu">epix_slide</span>(</span>
<span id="cb90-4"><a></a>    nowcaster,</span>
<span id="cb90-5"><a></a>    <span class="at">.before=</span><span class="cn">Inf</span>,</span>
<span id="cb90-6"><a></a>    <span class="at">.versions =</span> all_nowcast_dates,</span>
<span id="cb90-7"><a></a>    <span class="at">.all_versions =</span> <span class="cn">TRUE</span></span>
<span id="cb90-8"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="details-of-epix_slide" class="slide level2">
<h2>Details of <code>epix_slide()</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a></a><span class="fu">epix_slide</span>(</span>
<span id="cb91-2"><a></a>  .x,</span>
<span id="cb91-3"><a></a>  .f,</span>
<span id="cb91-4"><a></a>  ...,</span>
<span id="cb91-5"><a></a>  <span class="at">.before =</span> <span class="cn">Inf</span>,</span>
<span id="cb91-6"><a></a>  <span class="at">.versions =</span> <span class="cn">NULL</span>,</span>
<span id="cb91-7"><a></a>  <span class="at">.new_col_name =</span> <span class="cn">NULL</span>,</span>
<span id="cb91-8"><a></a>  <span class="at">.all_versions =</span> <span class="cn">FALSE</span></span>
<span id="cb91-9"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><code>.f</code> in <code>epix_slide()</code> can be specified with the same form of custom function as <code>epi_slide()</code>.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a></a><span class="cf">function</span>(x, g, t) {</span>
<span id="cb92-2"><a></a>  <span class="co"># function body</span></span>
<span id="cb92-3"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Mandatory variables of <code>.f</code> would have different forms depending on the value of <code>.all_versions</code>.</li>
</ul>
</section>
<section id="details-of-epix_slide-1" class="slide level2">
<h2>Details of <code>epix_slide()</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb93" data-code-line-numbers="|8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a></a><span class="fu">epix_slide</span>(</span>
<span id="cb93-2"><a></a>  .x,                          </span>
<span id="cb93-3"><a></a>  .f,                         </span>
<span id="cb93-4"><a></a>  ...,                        </span>
<span id="cb93-5"><a></a>  <span class="at">.before =</span> <span class="cn">Inf</span>,             </span>
<span id="cb93-6"><a></a>  <span class="at">.versions =</span> <span class="cn">NULL</span>,           </span>
<span id="cb93-7"><a></a>  <span class="at">.new_col_name =</span> <span class="cn">NULL</span>,      </span>
<span id="cb93-8"><a></a>  <span class="at">.all_versions =</span> <span class="cn">FALSE</span></span>
<span id="cb93-9"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment fade-in">
<ul>
<li>When <code>.all_versions = FALSE</code>, <code>epix_slide()</code> essentially iterates the templatized computation over snapshots.</li>
<li>Said differently, when <code>.all_versions = FALSE</code>, data accessed at any sliding iteration <span class="primary"><strong>only involves a single version</strong></span>.</li>
</ul>
</div>
<div class="fragment fade-in">
<ul>
<li>Hence:
<ul>
<li><code>x</code>: an <code>epi_df</code> with same column names as archive’s <code>DT</code>, minus the <code>version</code> column.</li>
<li><code>g</code>: a one-row tibble containing the values of groupping variables of the associated group.</li>
<li><code>t</code>: the <code>ref_time_value</code> of the current window.</li>
<li><code>...</code>: additional arguments.</li>
</ul></li>
</ul>
</div>
</section>
<section id="details-of-epix_slide-2" class="slide level2">
<h2>Details of <code>epix_slide()</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb94" data-code-line-numbers="|8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a></a><span class="fu">epix_slide</span>(</span>
<span id="cb94-2"><a></a>  .x,                          </span>
<span id="cb94-3"><a></a>  .f,                         </span>
<span id="cb94-4"><a></a>  ...,                        </span>
<span id="cb94-5"><a></a>  <span class="at">.before =</span> <span class="cn">Inf</span>,             </span>
<span id="cb94-6"><a></a>  <span class="at">.versions =</span> <span class="cn">NULL</span>,           </span>
<span id="cb94-7"><a></a>  <span class="at">.new_col_name =</span> <span class="cn">NULL</span>,      </span>
<span id="cb94-8"><a></a>  <span class="at">.all_versions =</span> <span class="cn">TRUE</span></span>
<span id="cb94-9"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment fade-in">
<ul>
<li>When <code>.all_versions = FALSE</code>, data accessed at any sliding iteration involves versions <span class="primary"><strong>up to and including .ref_time_value</strong></span>.</li>
</ul>
</div>
<div class="fragment fade-in">
<ul>
<li>Hence:
<ul>
<li><code>x</code>: an <code>epi_archive</code>, with version up to and including <code>.ref_time_value</code>.</li>
<li><code>g</code>: a one-row tibble containing the values of groupping variables of the associated group.</li>
<li><code>t</code>: the <code>ref_time_value</code> of the current window.</li>
<li><code>...</code>: additional arguments.</li>
</ul></li>
</ul>
</div>
</section>
<section id="details-of-epix_slide-3" class="slide level2">
<h2>Details of <code>epix_slide()</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb95" data-code-line-numbers="|7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a></a>nowcasts <span class="ot">&lt;-</span> nchs_archive <span class="sc">|&gt;</span></span>
<span id="cb95-2"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">|&gt;</span></span>
<span id="cb95-3"><a></a>  <span class="fu">epix_slide</span>(</span>
<span id="cb95-4"><a></a>    nowcaster,</span>
<span id="cb95-5"><a></a>    <span class="at">.before=</span><span class="cn">Inf</span>,</span>
<span id="cb95-6"><a></a>    <span class="at">.versions =</span> all_nowcast_dates,</span>
<span id="cb95-7"><a></a>    <span class="at">.all_versions =</span> <span class="cn">TRUE</span></span>
<span id="cb95-8"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="details-of-epix_slide-4" class="slide level2">
<h2>Details of <code>epix_slide()</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb96" data-code-line-numbers="|4,11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a></a>nowcaster <span class="ot">=</span> <span class="cf">function</span>(x, g, t, <span class="at">wl=</span><span class="dv">180</span>, <span class="at">appx=</span>approx_final_lag) {</span>
<span id="cb96-2"><a></a></span>
<span id="cb96-3"><a></a>  initial_data <span class="ot">=</span> x<span class="sc">$</span>DT <span class="sc">|&gt;</span></span>
<span id="cb96-4"><a></a>    <span class="fu">group_by</span>(geo_value, time_value) <span class="sc">|&gt;</span></span>
<span id="cb96-5"><a></a>    <span class="fu">filter</span>(version <span class="sc">==</span>  <span class="fu">min</span>(version)) <span class="sc">|&gt;</span></span>
<span id="cb96-6"><a></a>    <span class="fu">filter</span>(time_value <span class="sc">&gt;=</span> t <span class="sc">-</span> wl <span class="sc">-</span> appx <span class="sc">&amp;</span> time_value <span class="sc">&lt;=</span> t <span class="sc">-</span> appx) <span class="sc">|&gt;</span></span>
<span id="cb96-7"><a></a>    <span class="fu">rename</span>(<span class="at">initial_val =</span> mortality) <span class="sc">|&gt;</span></span>
<span id="cb96-8"><a></a>    <span class="fu">select</span>(geo_value, time_value, initial_val)</span>
<span id="cb96-9"><a></a></span>
<span id="cb96-10"><a></a>  finalized_data <span class="ot">=</span> x<span class="sc">$</span>DT <span class="sc">|&gt;</span></span>
<span id="cb96-11"><a></a>    <span class="fu">group_by</span>(geo_value, time_value) <span class="sc">|&gt;</span></span>
<span id="cb96-12"><a></a>    <span class="fu">filter</span>(version <span class="sc">==</span>  <span class="fu">max</span>(version)) <span class="sc">|&gt;</span></span>
<span id="cb96-13"><a></a>    <span class="fu">filter</span>(time_value <span class="sc">&gt;=</span> t <span class="sc">-</span> wl <span class="sc">-</span> appx <span class="sc">&amp;</span> time_value <span class="sc">&lt;=</span> t <span class="sc">-</span> appx) <span class="sc">|&gt;</span></span>
<span id="cb96-14"><a></a>    <span class="fu">rename</span>(<span class="at">finalized_val =</span> mortality) <span class="sc">|&gt;</span></span>
<span id="cb96-15"><a></a>    <span class="fu">select</span>(geo_value, time_value, finalized_val)</span>
<span id="cb96-16"><a></a>  </span>
<span id="cb96-17"><a></a>  ratio <span class="ot">=</span> finalized_data <span class="sc">|&gt;</span></span>
<span id="cb96-18"><a></a>    <span class="fu">inner_join</span>(initial_data, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"geo_value"</span>, <span class="st">"time_value"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb96-19"><a></a>    <span class="fu">mutate</span>(<span class="at">ratio =</span> finalized_val <span class="sc">/</span> initial_val) <span class="sc">|&gt;</span></span>
<span id="cb96-20"><a></a>    <span class="fu">pull</span>(ratio) <span class="sc">|&gt;</span></span>
<span id="cb96-21"><a></a>    <span class="fu">median</span>(<span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb96-22"><a></a></span>
<span id="cb96-23"><a></a>  last_avail <span class="ot">=</span> <span class="fu">epix_as_of</span>(x, t) <span class="sc">|&gt;</span></span>
<span id="cb96-24"><a></a>    <span class="fu">slice_max</span>(time_value) <span class="sc">|&gt;</span></span>
<span id="cb96-25"><a></a>    <span class="fu">pull</span>(mortality) </span>
<span id="cb96-26"><a></a>  </span>
<span id="cb96-27"><a></a>  res <span class="ot">=</span> <span class="fu">tibble</span>(<span class="at">geo_value =</span> x<span class="sc">$</span>geo_value, <span class="at">target_date =</span> t, <span class="at">nowcast =</span> last_avail <span class="sc">*</span> ratio)</span>
<span id="cb96-28"><a></a>  </span>
<span id="cb96-29"><a></a>  <span class="fu">return</span>(res)</span>
<span id="cb96-30"><a></a>  </span>
<span id="cb96-31"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="visualize-nowcasts" class="slide level2">
<h2>Visualize nowcasts</h2>
<p>We are now finally able to compare nowcasts against first available reports:</p>

<img data-src="day1-afternoon_files/figure-revealjs/nowcast-fun-plot-results-1.svg" class="quarto-figure quarto-figure-center r-stretch"><ul>
<li>The real-time counts tend to be biased below the finalized counts. Nowcasted values tend to provide a much better approximation of the truth (at least for these dates).</li>
</ul>
</section>
<section id="smoothing-nowcasts" class="slide level2">
<h2>Smoothing nowcasts</h2>
<ul>
<li>Nowcasts are quite volatile, reflecting the provisional counts are far from complete.</li>
<li>We can use a trailing average to smooth them.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a></a>smoothed_nowcasts <span class="ot">&lt;-</span> <span class="fu">epi_slide</span>(</span>
<span id="cb97-2"><a></a>  nowcasts <span class="sc">|&gt;</span> <span class="fu">as_epi_df</span>(),</span>
<span id="cb97-3"><a></a>  <span class="at">smoothed_nowcasts =</span> <span class="fu">mean</span>(nowcast, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb97-4"><a></a>  <span class="at">.window_size =</span> <span class="fu">as.difftime</span>(<span class="dv">3</span>, <span class="at">units =</span> <span class="st">"weeks"</span>)</span>
<span id="cb97-5"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<img data-src="day1-afternoon_files/figure-revealjs/nowcast-smoothed-vis-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="evaluation-using-mae" class="slide level2">
<h2>Evaluation using MAE</h2>
<ul>
<li><p>Assume we have prediction <span class="math inline">\(\hat y_{t}\)</span> for the provisional value at time <span class="math inline">\(t\)</span>.</p></li>
<li><p>Then for <span class="math inline">\(y_{t}\)</span> over times <span class="math inline">\(t = 1, \dots, N\)</span>, then we may compute error metrics like mean absolute error (MAE).</p></li>
<li><p>MAE measures the average absolute difference between the nowcast and finalized values.</p></li>
</ul>
<p><span class="math display">\[MAE = \frac{1}{N} \sum_{t=1}^N |y_{t}- \hat y_{t}|\]</span></p>
<ul>
<li>Note that it’s scale-dependent, meaning it can vary depending on the units of the data (e.g., cases, deaths, etc.).</li>
</ul>
</section>
<section id="evaluation-using-mae-1" class="slide level2">
<h2>Evaluation using MAE</h2>
<p>Let’s numerically evaluate our point nowcasts for the provisional values of a time series (e.g., COVID-19 mortality) using MAE.</p>
<!-- Accuracy of nowcast is assessed by how close provisional estimates are to the finalized values to gauge the model's performance. -->
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="caption-top">
<thead>
<tr class="header">
<th style="text-align: right;">Smoothed MAE</th>
<th style="text-align: right;">Unsmoothed nowcast MAE</th>
<th style="text-align: right;">Provisional value MAE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">161.6591</td>
<td style="text-align: right;">172.4131</td>
<td style="text-align: right;">199.6061</td>
</tr>
</tbody>
</table>
</div>
</div>
</section></section>
<section>
<section id="nowcasting-with-two-variables" class="title-slide slide level1 center" data-number="5">
<h1><span class="header-section-number">5</span> Nowcasting with Two Variables</h1>

</section>
<section id="nowcasting-moving-from-one-signal-to-two" class="slide level2">
<h2>Nowcasting: Moving from one signal to two</h2>
<ul>
<li>Recall that in nowcasting the goal is to predict a finalized value from a provisional value.</li>
<li>Now, we’ll move from one signal to two, creating a simple linear model to nowcast.</li>
<li>Exogenous features (predictors) could include relevant signals, such as Google symptom search trends.</li>
<li>We will use these signals to nowcast hospital admissions related to influenza.</li>
</ul>
</section>
<section id="data-sources-google-searches-hospital-admissions" class="slide level2">
<h2>Data Sources: Google searches &amp; hospital admissions</h2>
<ul>
<li><p><span class="primary"><strong>Google Search Trends</strong></span>: Symptoms like cough, fever, and shortness of breath.</p>
<ul>
<li><span class="primary"><strong>s01</strong></span>: Cough, Phlegm, Sputum, Upper respiratory tract infection<br>
</li>
<li><span class="primary"><strong>s02</strong></span>: Nasal congestion, Post nasal drip, Sinusitis, Common cold</li>
</ul></li>
<li><p><span class="primary"><strong>Hospital Admissions</strong></span>: Data from the Department of Health &amp; Human Services on confirmed influenza admissions.</p></li>
<li><p>Using these, we will <span class="primary"><strong>nowcast</strong></span> hospital admissions by using Google symptom search trends for GA from April to June 2023.</p></li>
<li><p>The first step is to fetch this data…</p></li>
</ul>
</section>
<section id="data-sources-google-searches-hospital-admissions-1" class="slide level2">
<h2>Data Sources: Google searches &amp; hospital admissions</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a></a><span class="co"># Fetch Google symptom data for s01 and s02</span></span>
<span id="cb98-2"><a></a>x1 <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb98-3"><a></a>  <span class="at">source =</span> <span class="st">"google-symptoms"</span>,</span>
<span id="cb98-4"><a></a>  <span class="at">signals =</span> <span class="st">"s01_smoothed_search"</span>, </span>
<span id="cb98-5"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb98-6"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb98-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"ga"</span>,</span>
<span id="cb98-8"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">20230401</span>, <span class="dv">20230701</span>),</span>
<span id="cb98-9"><a></a>  <span class="at">issues =</span> <span class="st">"*"</span></span>
<span id="cb98-10"><a></a>) <span class="sc">|&gt;</span></span>
<span id="cb98-11"><a></a>  <span class="fu">select</span>(geo_value, time_value, <span class="at">version =</span> issue, <span class="at">avg_search_vol_s01 =</span> value) <span class="sc">|&gt;</span></span>
<span id="cb98-12"><a></a>  <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">FALSE</span>)</span>
<span id="cb98-13"><a></a></span>
<span id="cb98-14"><a></a>x2 <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb98-15"><a></a>  <span class="at">source =</span> <span class="st">"google-symptoms"</span>,</span>
<span id="cb98-16"><a></a>  <span class="at">signals =</span> <span class="st">"s02_smoothed_search"</span>,</span>
<span id="cb98-17"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb98-18"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb98-19"><a></a>  <span class="at">geo_values =</span> <span class="st">"ga"</span>,</span>
<span id="cb98-20"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">20230401</span>, <span class="dv">20230701</span>),</span>
<span id="cb98-21"><a></a>  <span class="at">issues =</span> <span class="st">"*"</span></span>
<span id="cb98-22"><a></a>) <span class="sc">|&gt;</span></span>
<span id="cb98-23"><a></a>  <span class="fu">select</span>(geo_value, time_value, <span class="at">version =</span> issue, <span class="at">avg_search_vol_s02 =</span> value) <span class="sc">|&gt;</span></span>
<span id="cb98-24"><a></a>  <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">FALSE</span>)</span>
<span id="cb98-25"><a></a></span>
<span id="cb98-26"><a></a><span class="co"># Fetch hospital admissions data</span></span>
<span id="cb98-27"><a></a>y1 <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb98-28"><a></a>  <span class="at">source =</span> <span class="st">"hhs"</span>,</span>
<span id="cb98-29"><a></a>  <span class="at">signals =</span> <span class="st">"confirmed_admissions_influenza_1d"</span>,</span>
<span id="cb98-30"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb98-31"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb98-32"><a></a>  <span class="at">geo_values =</span> <span class="st">"ga"</span>,</span>
<span id="cb98-33"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">20230401</span>, <span class="dv">20230701</span>),</span>
<span id="cb98-34"><a></a>  <span class="at">issues =</span> <span class="st">"*"</span></span>
<span id="cb98-35"><a></a>) <span class="sc">|&gt;</span></span>
<span id="cb98-36"><a></a>  <span class="fu">select</span>(geo_value, time_value, <span class="at">version =</span> issue, <span class="at">admissions =</span> value) <span class="sc">|&gt;</span></span>
<span id="cb98-37"><a></a>  <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="merging-the-archives" class="slide level2">
<h2>Merging the archives</h2>
<ul>
<li>We’ll merge the symptom search trends (<code>x1</code>, <code>x2</code>) with hospital admissions data (<code>y</code>) using <code>epix_merge()</code> from <code>epiprocess</code>.</li>
<li>This allows us to match data by time and geography, &amp; fill any missing values with the most recent observation (LOCF).</li>
</ul>
</section>
<section id="linear-model-a-simple-approach-for-nowcasting" class="slide level2">
<h2>Linear Model: A simple approach for nowcasting</h2>
<ul>
<li>Aside from ratios, one of the simplest approach to nowcasting is to use a <span class="primary"><strong>linear regression model</strong></span>.</li>
<li>We model the relationship between provisional (predictor) data and response data.</li>
<li>This model helps us make <span class="primary"><strong>predictions</strong></span> for the finalized data based on the current (provisional) signals.</li>
</ul>
</section>
<section id="linear-regression" class="slide level2">
<h2>Linear regression</h2>
<ul>
<li><p><span class="primary"><strong>Goal</strong></span>: Estimate the coefficients <span class="math inline">\(\beta_0\)</span> and <span class="math inline">\(\beta_1\)</span> that describe the relationship between the predictor <span class="math inline">\(x_i\)</span> and the outcome <span class="math inline">\(y_i\)</span>.</p></li>
<li><p><span class="primary"><strong>Linear Model</strong></span>: The relationship is assumed to be:</p>
<p><span class="math display">\[y_i \approx \beta_0 + \beta_1 x_i \]</span></p>
<p>where <span class="math inline">\(\beta_0\)</span> is the intercept, <span class="math inline">\(\beta_1\)</span> is the slope.</p></li>
<li><p><strong>In R</strong>: Use <code>lm(y ~ x)</code> to estimate the coefficients, where <code>y</code> is the outcome variable and <code>x</code> is the predictor.</p></li>
</ul>
</section>
<section id="multiple-linear-regression" class="slide level2">
<h2>Multiple linear regression</h2>
<ul>
<li><p><span class="primary"><strong>Goal</strong></span>: Estimate coefficients <span class="math inline">\(\beta_0, \beta_1, \dots, \beta_p\)</span> that describe the relationship between multiple predictors <span class="math inline">\(x_{i1}, x_{i2}, \dots, x_{ip}\)</span> and the outcome <span class="math inline">\(y_i\)</span>.</p></li>
<li><p><span class="primary"><strong>Model</strong></span>: The relationship is assumed to be:</p>
<p><span class="math display">\[y_i \approx \beta_0 + \beta_1 x_{i1} + \beta_2 x_{i2} + \dots + \beta_p x_{ip}\]</span></p>
<p>where: <span class="math inline">\(\beta_0\)</span> is the intercept, <span class="math inline">\(\beta_1, \dots, \beta_p\)</span> are the coefficients.</p></li>
<li><p><span class="primary"><strong>In R</strong></span>: Use <code>lm(y ~ x1 + x2 + ... + xp)</code> to estimate the coefficients, where <code>y</code> is the outcome and <code>x1, x2, ..., xp</code> are the predictors.</p></li>
</ul>
</section>
<section id="multiple-linear-regression-model" class="slide level2">
<h2>Multiple linear regression model</h2>
<ul>
<li>A linear model is a good choice to describe the relationship between search trends and hospital admissions.</li>
<li>The model will include two predictors (s01 and s02).</li>
<li>We’ll use these two search trend signals to predict hospital admissions (response).</li>
</ul>
<!-- A linear regression model will be used to predict hospital admissions from search trends (s01 and s02). -->
</section>
<section id="multiple-linear-regression-model-1" class="slide level2">
<h2>Multiple linear regression model</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a></a><span class="co"># Define the function for lm model fit and prediction</span></span>
<span id="cb99-2"><a></a>lm_mod_pred <span class="ot">&lt;-</span> <span class="cf">function</span>(data, gk, rtv, ...) {</span>
<span id="cb99-3"><a></a>  </span>
<span id="cb99-4"><a></a>  <span class="co"># Fit the linear model</span></span>
<span id="cb99-5"><a></a>  model <span class="ot">&lt;-</span> <span class="fu">lm</span>(admissions <span class="sc">~</span> avg_search_vol_s01 <span class="sc">+</span> avg_search_vol_s02, <span class="at">data =</span> data)</span>
<span id="cb99-6"><a></a>  </span>
<span id="cb99-7"><a></a>  <span class="co"># Make predictions</span></span>
<span id="cb99-8"><a></a>  predictions <span class="ot">=</span> <span class="fu">predict</span>(model,</span>
<span id="cb99-9"><a></a>                        <span class="at">newdata =</span> data <span class="sc">|&gt;</span></span>
<span id="cb99-10"><a></a>                          <span class="co"># Use tidyr::fill() for LOCF if predictor data is incomplete </span></span>
<span id="cb99-11"><a></a>                          <span class="fu">fill</span>(avg_search_vol_s01, <span class="at">.direction =</span> <span class="st">"down"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb99-12"><a></a>                          <span class="fu">fill</span>(avg_search_vol_s02, <span class="at">.direction =</span> <span class="st">"down"</span>) <span class="sc">|&gt;</span></span>
<span id="cb99-13"><a></a>                          <span class="fu">filter</span>(time_value <span class="sc">==</span> <span class="fu">max</span>(time_value)),</span>
<span id="cb99-14"><a></a>                        <span class="at">interval =</span> <span class="st">"prediction"</span>, <span class="at">level =</span> <span class="fl">0.9</span></span>
<span id="cb99-15"><a></a>  )</span>
<span id="cb99-16"><a></a></span>
<span id="cb99-17"><a></a>  <span class="co"># Pull off true time value for comparison to target</span></span>
<span id="cb99-18"><a></a>  real_time_val <span class="ot">=</span> data <span class="sc">|&gt;</span> <span class="fu">filter</span>(time_value <span class="sc">==</span> <span class="fu">max</span>(time_value)) <span class="sc">|&gt;</span> <span class="fu">pull</span>(admissions)</span>
<span id="cb99-19"><a></a></span>
<span id="cb99-20"><a></a>  <span class="fu">return</span>(<span class="fu">data.frame</span>(predictions, <span class="at">actual_nowcast_date =</span> <span class="fu">max</span>(data<span class="sc">$</span>time_value), <span class="at">real_time_val =</span> real_time_val))</span>
<span id="cb99-21"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that this code is intentionally simple; while it can be refined to handle cases like negatives or other boundary conditions, we aim to avoid unnecessary complexity.</p>
</section>
<section id="nowcasting-with-epix_slide" class="slide level2">
<h2>Nowcasting with <code>epix_slide()</code></h2>
<ul>
<li>We will use <code>epix_slide()</code> to create a sliding window of training data.</li>
<li>The model will be trained on a 14-day window before the target date, and predictions will be made for the target date.</li>
<li>The beauty of this function is that it is version-aware - the sliding computation at any given reference time <span class="primary"><strong>t</strong></span> is performed on data that would have been available as of <span class="primary"><strong>t</strong></span> automatically.</li>
</ul>
</section>
<section id="nowcasting-with-epix_slide-1" class="slide level2">
<h2>Nowcasting with <code>epix_slide()</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a></a><span class="co"># Define the reference time points for nowcasting</span></span>
<span id="cb100-2"><a></a>targeted_nowcast_dates <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">as.Date</span>(<span class="st">"2023-04-15"</span>), <span class="fu">as.Date</span>(<span class="st">"2023-06-15"</span>), <span class="at">by =</span> <span class="st">"1 week"</span>)</span>
<span id="cb100-3"><a></a>ref_time_values <span class="ot">=</span> targeted_nowcast_dates <span class="sc">+</span> <span class="dv">2</span>  <span class="co"># Adjust for the systematic 2-day latency in the response</span></span>
<span id="cb100-4"><a></a><span class="co"># Determine this from revision_summary(y1, print_inform = TRUE) </span></span>
<span id="cb100-5"><a></a></span>
<span id="cb100-6"><a></a><span class="co"># Perform nowcasting using epix_slide</span></span>
<span id="cb100-7"><a></a>nowcast_res <span class="ot">&lt;-</span> archive <span class="sc">|&gt;</span></span>
<span id="cb100-8"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">|&gt;</span></span>
<span id="cb100-9"><a></a>  <span class="fu">epix_slide</span>(</span>
<span id="cb100-10"><a></a>    <span class="at">.f =</span> lm_mod_pred,</span>
<span id="cb100-11"><a></a>    <span class="at">.before =</span> <span class="dv">14</span>,  <span class="co"># 14-day training period</span></span>
<span id="cb100-12"><a></a>    <span class="at">.versions =</span> ref_time_values, </span>
<span id="cb100-13"><a></a>    <span class="at">.new_col_name =</span> <span class="st">"res"</span></span>
<span id="cb100-14"><a></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb100-15"><a></a>  <span class="fu">unnest</span>() <span class="sc">|&gt;</span> <span class="co"># Nesting creates a list-column of data frames; unnesting flattens it back out into regular columns. </span></span>
<span id="cb100-16"><a></a>  <span class="fu">mutate</span>(<span class="at">targeted_nowcast_date =</span> targeted_nowcast_dates, <span class="at">time_value =</span> actual_nowcast_date) <span class="sc">|&gt;</span></span>
<span id="cb100-17"><a></a>  <span class="fu">ungroup</span>()</span>
<span id="cb100-18"><a></a></span>
<span id="cb100-19"><a></a><span class="co"># View results</span></span>
<span id="cb100-20"><a></a><span class="fu">head</span>(nowcast_res, <span class="at">n=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 9
  geo_value version      fit    lwr   upr actual_nowcast_date real_time_val
  &lt;chr&gt;     &lt;date&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;date&gt;                      &lt;dbl&gt;
1 ga        2023-04-17  4.64 -0.122  9.39 2023-04-15                      4
2 ga        2023-04-24  7.36  1.56  13.2  2023-04-22                      4
# ℹ 2 more variables: targeted_nowcast_date &lt;date&gt;, time_value &lt;date&gt;</code></pre>
</div>
</div>
</section>
<section id="compare-with-the-actual-admissions" class="slide level2">
<h2>Compare with the actual admissions</h2>
<p>After making predictions, we compare them to the actual hospital admissions.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a></a><span class="co"># Left join with latest results </span></span>
<span id="cb102-2"><a></a><span class="co"># Latest snapshot of data (with the latest/finalized admissions)</span></span>
<span id="cb102-3"><a></a>x_latest <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(archive, <span class="fu">max</span>(archive<span class="sc">$</span>DT<span class="sc">$</span>version)) <span class="sc">|&gt;</span> <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(avg_search_vol_s01, avg_search_vol_s02))</span>
<span id="cb102-4"><a></a></span>
<span id="cb102-5"><a></a>res <span class="ot">&lt;-</span> nowcast_res <span class="sc">|&gt;</span> <span class="fu">left_join</span>(x_latest, <span class="at">by =</span> <span class="fu">join_by</span>(geo_value, time_value))</span>
<span id="cb102-6"><a></a><span class="fu">head</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 10
  geo_value version      fit    lwr   upr actual_nowcast_date real_time_val
  &lt;chr&gt;     &lt;date&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;date&gt;                      &lt;dbl&gt;
1 ga        2023-04-17  4.64 -0.122  9.39 2023-04-15                      4
2 ga        2023-04-24  7.36  1.56  13.2  2023-04-22                      4
3 ga        2023-05-01  6.06  1.57  10.5  2023-04-29                      5
4 ga        2023-05-08  5.01  1.28   8.74 2023-05-06                      6
5 ga        2023-05-15  8.14  5.69  10.6  2023-05-11                      8
6 ga        2023-05-22  3.43 -2.35   9.21 2023-05-20                      4
# ℹ 3 more variables: targeted_nowcast_date &lt;date&gt;, time_value &lt;date&gt;,
#   admissions &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="visualizing-the-nowcast-results" class="slide level2">
<h2>Visualizing the nowcast results</h2>
<p>We can then visualize the nowcast results alongside the true values using <code>ggplot2</code>:</p>

<img data-src="day1-afternoon_files/figure-revealjs/plot-multiple-lr-nowcast-res-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="evaluation-using-mae-2" class="slide level2">
<h2>Evaluation using MAE</h2>
<ul>
<li>As before, we can evaluate our point nowcasts numerically using MAE.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a></a><span class="co"># Calculate the absolute error between actual and nowcasted values</span></span>
<span id="cb104-2"><a></a>mae_data_admissions <span class="ot">&lt;-</span> res <span class="sc">|&gt;</span> </span>
<span id="cb104-3"><a></a>  <span class="fu">mutate</span>(<span class="at">nc_abs_error =</span> <span class="fu">abs</span>(admissions <span class="sc">-</span> fit),  <span class="co"># Nowcast vs Finalized admissions</span></span>
<span id="cb104-4"><a></a>         <span class="at">rt_abs_error =</span> <span class="fu">abs</span>(admissions <span class="sc">-</span> real_time_val))  <span class="co"># Real-Time vs Finalized admissions</span></span>
<span id="cb104-5"><a></a></span>
<span id="cb104-6"><a></a><span class="co"># Compute the MAE (mean of absolute errors)</span></span>
<span id="cb104-7"><a></a>mae_value_admissions <span class="ot">&lt;-</span> mae_data_admissions <span class="sc">|&gt;</span> </span>
<span id="cb104-8"><a></a>  <span class="fu">summarise</span>(<span class="at">nc_MAE =</span> <span class="fu">mean</span>(nc_abs_error),</span>
<span id="cb104-9"><a></a>            <span class="at">rt_MAE =</span> <span class="fu">mean</span>(rt_abs_error))</span>
<span id="cb104-10"><a></a>knitr<span class="sc">::</span><span class="fu">kable</span>(mae_value_admissions)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top">
<thead>
<tr class="header">
<th style="text-align: right;">nc_MAE</th>
<th style="text-align: right;">rt_MAE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">1.180965</td>
<td style="text-align: right;">2.333333</td>
</tr>
</tbody>
</table>
</div>
</div>
<ul>
<li>Based off of comparing these simple error measures, the nowcast MAE is clearly better.</li>
</ul>
</section>
<section id="evaluation-using-mae-3" class="slide level2">
<h2>Evaluation using MAE</h2>
<p>However, when we visualize the distribution of errors across time, it is not so cut-and-dry:</p>

<img data-src="day1-afternoon_files/figure-revealjs/abs-error-plot-hosp-1.svg" class="quarto-figure quarto-figure-center r-stretch"><p>The main driver behind the real-time MAE being greater is the “outlier-like” May 25 AE.</p>
<p>So visualizing can provide an important perspective that is missed from a simple numerical summary of error.</p>
</section>
<section id="key-takeaways-linear-regression-nowcasting-example" class="slide level2">
<h2>Key Takeaways: Linear regression nowcasting example</h2>
<ul>
<li><span class="primary"><strong>Provisional Data as Predictors</strong></span>: Using <span class="primary">Google symptom search trends</span> to predict <span class="primary">influenza hospital admissions</span>.</li>
<li><span class="primary"><strong>Simple Linear Model</strong></span>: A linear regression model captures the relationship between symptom searches and hospital admissions.</li>
<li><span class="primary"><strong>Actionable Predictions</strong></span>: Nowcasts provide <span class="primary">timely insights</span> for hospital admissions, even before data is finalized.</li>
<li><span class="primary"><strong>Sliding Window Approach</strong></span>: Predictions are based on <span class="primary">data up to the current time</span>, ensuring no future information influences the nowcast.</li>
<li><span class="primary"><strong>Evaluation</strong></span>: Predictions are compared with actual admissions using numerical and visual perspectives.</li>
</ul>
</section></section>
<section>
<section id="case-study---nowcasting-cases-using-cli" class="title-slide slide level1 center" data-number="6">
<h1><span class="header-section-number">6</span> Case Study - Nowcasting Cases Using %CLI</h1>

</section>
<section id="goal-of-this-case-study" class="slide level2">
<h2>Goal of this case study</h2>
<p><span class="primary"><strong>Goal</strong></span>: Nowcast COVID-19 Cases for MA using the estimated percentage of COVID-related doctor’s visits (%CLI), based on outpatient data from Optum.</p>
<ul>
<li>%CLI is contained in the Epidata API.</li>
<li>Cases by specimen collection date are not. They are from the MA gov website.</li>
<li>Cases in the API (JHU) are aligned by report date, not specimen collection/test date.</li>
<li>Working with cases aligned by <span class="primary"><strong>test date</strong></span> allows us to avoid the more unpredictable delays introduced by the <span class="primary"><strong>report date</strong></span>.</li>
</ul>
</section>
<section id="summary-of-main-steps" class="slide level2">
<h2>Summary of main steps</h2>
<p>The workflow is similar to the previous example where we nowcasted using two variables, only more involved. The main steps are…</p>
<ol type="1">
<li><p><span class="primary"><strong>Fetch Data</strong></span>: Retrieve %CLI and COVID-19 case data (by specimen collection date) for MA.</p></li>
<li><p><span class="primary"><strong>Merge Data</strong></span>: Align %CLI and case data using <code>epix_merge</code>, filling missing values via last observation carried forward (LOCF).</p></li>
<li><p><span class="primary"><strong>Model &amp; Prediction</strong></span>: Fit a linear model to predict cases based on %CLI, trained on a 30-day rolling window.</p></li>
<li><p><span class="primary"><strong>Nowcast Execution</strong></span>: Use <code>epix_slide</code> to nowcast the cases dynamically.</p></li>
<li><p><span class="primary"><strong>Visualization</strong></span>: Plot actual vs.&nbsp;nowcasted cases with confidence intervals to assess model accuracy.</p></li>
</ol>
<p>So the first step is to fetch the data…</p>
</section>
<section id="construct-an-epi_archive-from-scratch" class="slide level2">
<h2>Construct an <code>epi_archive</code> from scratch</h2>
<p><a href="&quot;https://www.mass.gov/info-details/archive-of-covid-19-cases-2020-2021&quot;">Here’s</a> the archive of COVID-19 case excel files from the MA gov website, which we’ll use to construct our own <code>epi_archive</code>. <br> <br> Brief summary of this data:</p>
<ul>
<li><p><span class="primary"><strong>First release</strong></span>: Raw .xlsx data was first released early January 2021.</p></li>
<li><p><span class="primary"><strong>Change in reporting</strong></span>: Starting <span class="primary"><strong>July 1, 2021</strong></span>, the dashboard shifted from <span class="primary"><strong>7 days/week</strong></span> to <span class="primary"><strong>5 days/week</strong></span> (Monday-Friday).</p></li>
<li><p><span class="primary"><strong>Friday, Saturday, and Sunday</strong></span> data is included in the <span class="primary"><strong>Monday</strong></span> dashboard.</p></li>
<li><p>When <span class="primary"><strong>Monday</strong></span> is a holiday, the <span class="primary"><strong>Friday through Monday</strong></span> data is posted on <span class="primary"><strong>Tuesday</strong></span>.</p></li>
</ul>
</section>
<section id="construct-an-epi_archive-from-scratch-1" class="slide level2">
<h2>Construct an <code>epi_archive</code> from scratch</h2>
<ul>
<li><span class="primary"><strong>Purpose</strong></span>: To create an <code>epi_archive</code> object for storing versioned time series data.</li>
<li><span class="primary"><strong>Required Columns</strong></span>:
<ul>
<li><code>geo_value</code>: Geographic data (e.g., region).</li>
<li><code>time_value</code>: Time-related data (e.g., date, time).</li>
<li><code>version</code>: Tracks when the data was available (enables version-aware forecasting).</li>
</ul></li>
<li><span class="primary"><strong>Constructor</strong></span>:
<ul>
<li><code>new_epi_archive()</code>: For manual construction of <code>epi_archive</code> (assumes validation of inputs).</li>
</ul></li>
<li><span class="primary"><strong>Recommended Method</strong></span>:
<ul>
<li><code>as_epi_archive()</code>: Simplifies the creation process, ensuring proper formatting and validation. We’ll use this one when we download some data from the MA gov website!</li>
</ul></li>
</ul>
</section>
<section id="main-steps-to-construct-the-epi_archive" class="slide level2">
<h2>Main steps to construct the <code>epi_archive</code></h2>
<ol type="1">
<li><span class="primary"><strong>Load necessary Libraries</strong></span>: Such as <code>tidyverse</code>, <code>readxl</code>, <code>epiprocess</code>.</li>
<li><span class="primary"><strong>Process Each Date’s Data</strong></span>:
<ul>
<li>A function we’ll make (<code>process_covid_data</code>) downloads and processes daily COVID-19 data from the MA gov Excel files on their website.</li>
<li>The data is cleaned and formatted with columns: <code>geo_value</code>, <code>time_value</code>, <code>version</code>, and values.</li>
</ul></li>
<li><span class="primary"><strong>Handle Missing Data</strong></span>: Checks if a date’s data is available (handle 404 errors).</li>
<li><span class="primary"><strong>Create <code>epi_archive</code></strong></span>:
<ul>
<li>Combine processed data into a tibble.</li>
<li>Convert the tibble to an <code>epi_archive</code> object using <code>as_epi_archive()</code>.</li>
</ul></li>
</ol>
</section>
<section id="fetch-data---code-for-one-date" class="slide level2">
<h2>Fetch Data - Code for one date</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a></a><span class="co"># Load required libraries</span></span>
<span id="cb105-2"><a></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb105-3"><a></a><span class="fu">library</span>(readxl)</span>
<span id="cb105-4"><a></a><span class="fu">library</span>(httr)</span>
<span id="cb105-5"><a></a><span class="fu">library</span>(tibble)</span>
<span id="cb105-6"><a></a><span class="fu">library</span>(epiprocess)</span>
<span id="cb105-7"><a></a></span>
<span id="cb105-8"><a></a><span class="co"># Function to download and process each Excel file for a given date</span></span>
<span id="cb105-9"><a></a>process_covid_data <span class="ot">&lt;-</span> <span class="cf">function</span>(Date) {</span>
<span id="cb105-10"><a></a>  <span class="co"># Generate the URL for the given date</span></span>
<span id="cb105-11"><a></a>  url <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"https://www.mass.gov/doc/covid-19-raw-data-"</span>, <span class="fu">tolower</span>(<span class="fu">gsub</span>(<span class="st">"-0"</span>, <span class="st">"-"</span>, <span class="fu">format</span>(Date, <span class="st">"%B-%d-%Y"</span>))), <span class="st">"/download"</span>) </span>
<span id="cb105-12"><a></a>  <span class="co"># Applies gsub("-0", "-", ...) to replace any occurrence of -0 (such as in "April-01") with just - (resulting in "April-1").</span></span>
<span id="cb105-13"><a></a>  </span>
<span id="cb105-14"><a></a>  <span class="co"># Check if the URL exists (handle the 404 error by skipping that date)</span></span>
<span id="cb105-15"><a></a>  response <span class="ot">&lt;-</span> <span class="fu">GET</span>(url)</span>
<span id="cb105-16"><a></a>  </span>
<span id="cb105-17"><a></a>  <span class="cf">if</span> (<span class="fu">status_code</span>(response) <span class="sc">!=</span> <span class="dv">200</span>) {</span>
<span id="cb105-18"><a></a>    <span class="fu">return</span>(<span class="cn">NULL</span>)  <span class="co"># Skip if URL doesn't exist (404)</span></span>
<span id="cb105-19"><a></a>  }</span>
<span id="cb105-20"><a></a>  </span>
<span id="cb105-21"><a></a>  <span class="co"># Define the destination file path for the Excel file</span></span>
<span id="cb105-22"><a></a>  file_path <span class="ot">&lt;-</span> <span class="fu">tempfile</span>(<span class="at">fileext =</span> <span class="st">".xlsx"</span>)</span>
<span id="cb105-23"><a></a>  </span>
<span id="cb105-24"><a></a>  <span class="co"># Download the Excel file</span></span>
<span id="cb105-25"><a></a>  <span class="fu">GET</span>(url, <span class="fu">write_disk</span>(file_path, <span class="at">overwrite =</span> <span class="cn">TRUE</span>))</span>
<span id="cb105-26"><a></a>  </span>
<span id="cb105-27"><a></a>  <span class="co"># Read the relevant sheet from the Excel file</span></span>
<span id="cb105-28"><a></a>  data <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(file_path, <span class="at">sheet =</span> <span class="st">"CasesByDate (Test Date)"</span>)</span>
<span id="cb105-29"><a></a>  </span>
<span id="cb105-30"><a></a>  <span class="co"># Process the data: rename columns and convert Date</span></span>
<span id="cb105-31"><a></a>  data <span class="ot">&lt;-</span> data <span class="sc">|&gt;</span></span>
<span id="cb105-32"><a></a>    <span class="fu">rename</span>(</span>
<span id="cb105-33"><a></a>      <span class="at">Date =</span> <span class="st">`</span><span class="at">Date</span><span class="st">`</span>,</span>
<span id="cb105-34"><a></a>      <span class="at">Positive_Total =</span> <span class="st">`</span><span class="at">Positive Total</span><span class="st">`</span>,</span>
<span id="cb105-35"><a></a>      <span class="at">Positive_New =</span> <span class="st">`</span><span class="at">Positive New</span><span class="st">`</span>,</span>
<span id="cb105-36"><a></a>      <span class="at">Case_Average_7day =</span> <span class="st">`</span><span class="at">7-day confirmed case average</span><span class="st">`</span></span>
<span id="cb105-37"><a></a>    ) <span class="sc">|&gt;</span></span>
<span id="cb105-38"><a></a>    <span class="fu">mutate</span>(<span class="at">Date =</span> <span class="fu">as.Date</span>(Date))  <span class="co"># Convert to Date class</span></span>
<span id="cb105-39"><a></a>  </span>
<span id="cb105-40"><a></a>  <span class="co"># Create a tibble with the required columns for the epi_archive</span></span>
<span id="cb105-41"><a></a>  tib <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb105-42"><a></a>    <span class="at">geo_value =</span> <span class="st">"ma"</span>,  <span class="co"># Massachusetts (geo_value)</span></span>
<span id="cb105-43"><a></a>    <span class="at">time_value =</span> data<span class="sc">$</span>Date,  <span class="co"># Date from the data</span></span>
<span id="cb105-44"><a></a>    <span class="at">version =</span> Date,  <span class="co"># The extracted version date</span></span>
<span id="cb105-45"><a></a>    <span class="at">case_rate_7d_av =</span> data<span class="sc">$</span>Case_Average_7day  <span class="co"># 7-day average case value</span></span>
<span id="cb105-46"><a></a>  )</span>
<span id="cb105-47"><a></a>  </span>
<span id="cb105-48"><a></a>  <span class="fu">return</span>(tib)</span>
<span id="cb105-49"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fetch-data---code-breakdown" class="slide level2">
<h2>Fetch Data - Code breakdown</h2>
<ul>
<li>This purpose of this function is to download and process each Excel file as of a date.</li>
<li><span class="primary"><strong>URL Creation</strong></span>: Dynamically generates the URL based on the date, removing leading zeros in day values (e.g., “April-01” → “April-1”).</li>
<li><span class="primary"><strong>Check URL</strong></span>: Sends a request (<code>GET(url)</code>) and skips the date if the URL returns a non-200 status (e.g., 404 error).</li>
<li><span class="primary"><strong>Download File</strong></span>: Saves the Excel file to a temporary path using <code>tempfile()</code> and <code>GET()</code>.</li>
<li><span class="primary"><strong>Read Data</strong></span>: Loads the relevant sheet (“CasesByDate”) from the Excel file using <code>read_excel()</code>.</li>
<li><span class="primary"><strong>Tibble Creation</strong></span>: Constructs a tibble with <code>geo_value</code>, <code>time_value</code>, <code>version</code>, and <code>case_rate_7d_av</code> to later compile into an <code>epi_archive</code> (you can think of an <code>epi_archive</code> as being a comprised of many <code>epi_df</code>s).</li>
</ul>
</section>
<section id="fetch-data---process-eange-of-dates" class="slide level2">
<h2>Fetch Data - Process eange of dates</h2>
<ul>
<li>Note that <code>process_covid_data()</code> works on one date at a time.</li>
<li>So now, we need a function that iterates over a date range and applies <code>process_covid_data()</code> to each date &amp; combines the resulting tibbles into an <code>epi_archive</code>.</li>
<li>We call this function <code>process_data_for_date_range()</code>…</li>
</ul>
</section>
<section id="fetch-data---process-range-of-dates" class="slide level2">
<h2>Fetch Data - Process range of dates</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a></a><span class="co"># Function to process data for a range of dates</span></span>
<span id="cb106-2"><a></a>process_data_for_date_range <span class="ot">&lt;-</span> <span class="cf">function</span>(start_date, end_date) {</span>
<span id="cb106-3"><a></a>  <span class="co"># Generate a sequence of dates between start_date and end_date</span></span>
<span id="cb106-4"><a></a>  date_sequence <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">as.Date</span>(start_date), <span class="fu">as.Date</span>(end_date), <span class="at">by =</span> <span class="st">"day"</span>)</span>
<span id="cb106-5"><a></a>  </span>
<span id="cb106-6"><a></a>  <span class="co"># Process data for each date and combine results</span></span>
<span id="cb106-7"><a></a>  covid_data_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(date_sequence, <span class="cf">function</span>(Date) {</span>
<span id="cb106-8"><a></a>    <span class="fu">process_covid_data</span>(Date)  <span class="co"># Skip over dates with no data (NULLs will be ignored)</span></span>
<span id="cb106-9"><a></a>  })</span>
<span id="cb106-10"><a></a>  </span>
<span id="cb106-11"><a></a>  <span class="co"># Combine all non-null individual tibbles into one data frame</span></span>
<span id="cb106-12"><a></a>  combined_data <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(covid_data_list[<span class="sc">!</span><span class="fu">sapply</span>(covid_data_list, is.null)])</span>
<span id="cb106-13"><a></a>  </span>
<span id="cb106-14"><a></a>  <span class="co"># Convert the combined data into an epi_archive object</span></span>
<span id="cb106-15"><a></a>  <span class="cf">if</span> (<span class="fu">nrow</span>(combined_data) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb106-16"><a></a>    epi_archive_data <span class="ot">&lt;-</span> combined_data <span class="sc">|&gt;</span></span>
<span id="cb106-17"><a></a>      <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">FALSE</span>)</span>
<span id="cb106-18"><a></a>    </span>
<span id="cb106-19"><a></a>    <span class="fu">return</span>(epi_archive_data)</span>
<span id="cb106-20"><a></a>  } <span class="cf">else</span> {</span>
<span id="cb106-21"><a></a>    <span class="fu">message</span>(<span class="st">"No valid data available for the given date range."</span>)</span>
<span id="cb106-22"><a></a>    <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb106-23"><a></a>  }</span>
<span id="cb106-24"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fetch-data---code-breakdown-1" class="slide level2">
<h2>Fetch Data - Code breakdown</h2>
<p>Here’s a summary of what <code>process_data_for_date_range()</code> does: 1. <span class="primary"><strong>Generates Date Range</strong></span>: Creates a sequence of dates between <code>start_date</code> and <code>end_date</code>.</p>
<ol start="2" type="1">
<li><p><span class="primary"><strong>Processes Data</strong></span>: Applies the <code>process_covid_data</code> function to each date in the range (skip over dates with no data).</p></li>
<li><p><span class="primary"><strong>Combines Results</strong></span>: Combines all valid (non-NULL) tibbles into one single data frame.</p></li>
<li><p><span class="primary"><strong>Creates <code>epi_archive</code></strong></span>: Converts the combined data into an <code>epi_archive</code> object.</p></li>
</ol>
</section>
<section id="fetch-data---run-the-function-inspect-archive" class="slide level2">
<h2>Fetch Data - Run the function &amp; inspect archive</h2>
<ul>
<li>Now, let’s run the function &amp; inspect the resulting <code>epi_archive</code> of 7-day avg. COVID-19 case counts:</li>
<li>Expect building the archive to some time (enough for a cup of coffee or to meditate on life).</li>
</ul>
<!-- To wonder why you chose Expect building the archive to take a nontrivial amount of time (enough for a cup of coffee or to wonder why you chose coding in the first place). -->
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a></a><span class="co"># Example usage: process data between Jan. 10, 2021, and Dec. 1, 2021</span></span>
<span id="cb107-2"><a></a>y <span class="ot">&lt;-</span> <span class="fu">process_data_for_date_range</span>(<span class="st">"2021-01-10"</span>, <span class="st">"2021-12-01"</span>)  <span class="co"># Raw .xlsx data is first released on Jan. 4, 2021</span></span>
<span id="cb107-3"><a></a>y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>→ An `epi_archive` object, with metadata:
ℹ Min/max time values: 2020-01-29 / 2021-11-30
ℹ First/last version with update: 2021-01-10 / 2021-12-01
ℹ Versions end: 2021-12-01
ℹ A preview of the table (135549 rows x 4 columns):
Key: &lt;geo_value, time_value, version&gt;
        geo_value time_value    version case_rate_7d_av
           &lt;char&gt;     &lt;Date&gt;     &lt;Date&gt;           &lt;num&gt;
     1:        ma 2020-01-29 2021-01-10              NA
     2:        ma 2020-01-29 2021-01-11              NA
     3:        ma 2020-01-29 2021-01-12              NA
     4:        ma 2020-01-29 2021-01-13              NA
     5:        ma 2020-01-29 2021-01-14              NA
    ---                                                
135545:        ma 2021-11-28 2021-11-30        2196.000
135546:        ma 2021-11-28 2021-12-01        2352.286
135547:        ma 2021-11-29 2021-11-30        1735.714
135548:        ma 2021-11-29 2021-12-01        2371.286
135549:        ma 2021-11-30 2021-12-01        1972.143</code></pre>
</div>
</div>
<ul>
<li>Alternatively, you may run the following to load <code>y</code> that was previously saved as an RDS file:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a></a>y <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"_data/ma_case_archive.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fetch-data---outpatient-doctors-visits-for-cli" class="slide level2">
<h2>Fetch Data - % Outpatient doctors visits for CLI</h2>
<ul>
<li>Now, from the Epidata API, let’s download the <a href="&quot;https://cmu-delphi.github.io/delphi-epidata/api/covidcast-signals/doctor-visits.html&quot;">estimated percentage of outpatient doctor visits</a> primarily for COVID-related symptoms, based on health system data.</li>
<li>Comes pre-smoothed in time using a Gaussian linear smoother</li>
<li>This will be the predictor when we nowcast COVID-19 cases in MA.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a></a><span class="co"># Step 1: Fetch Versioned Data </span></span>
<span id="cb110-2"><a></a>x <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb110-3"><a></a>  <span class="at">source =</span> <span class="st">"doctor-visits"</span>,</span>
<span id="cb110-4"><a></a>  <span class="at">signals =</span> <span class="st">"smoothed_adj_cli"</span>,</span>
<span id="cb110-5"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb110-6"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb110-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"ma"</span>, <span class="co"># Just for MA to keep it simple (&amp; to go with the case data by test date for that state)</span></span>
<span id="cb110-8"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">20210301</span>, <span class="dv">20211231</span>),</span>
<span id="cb110-9"><a></a>  <span class="at">issues =</span> <span class="fu">epirange</span>(<span class="dv">20210301</span>, <span class="dv">20211231</span>)</span>
<span id="cb110-10"><a></a>) <span class="sc">|&gt;</span></span>
<span id="cb110-11"><a></a>  <span class="fu">select</span>(geo_value, time_value,</span>
<span id="cb110-12"><a></a>         <span class="at">version =</span> issue,</span>
<span id="cb110-13"><a></a>         <span class="at">percent_cli =</span> value</span>
<span id="cb110-14"><a></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb110-15"><a></a>  <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="use-epix_merge-to-merge-the-two-archives" class="slide level2">
<h2>Use <code>epix_merge()</code> to merge the two archives</h2>
<p>Now we’ll use <code>epix_merge()</code> to combine the two <code>epi_archive</code>s that share the same <code>geo_value</code> &amp; <code>time_value</code>.</p>
<!-- LOCF is used to ensure missing data is handled by filling forward. -->
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a></a>archive <span class="ot">&lt;-</span> <span class="fu">epix_merge</span>(</span>
<span id="cb111-2"><a></a>  x, y,</span>
<span id="cb111-3"><a></a>  <span class="at">sync =</span> <span class="st">"locf"</span>,</span>
<span id="cb111-4"><a></a>  <span class="at">compactify =</span> <span class="cn">FALSE</span></span>
<span id="cb111-5"><a></a>)</span>
<span id="cb111-6"><a></a>archive</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>→ An `epi_archive` object, with metadata:
ℹ Min/max time values: 2020-01-29 / 2021-12-13
ℹ First/last version with update: 2021-01-10 / 2021-12-17
ℹ Versions end: 2021-12-17
ℹ A preview of the table (139190 rows x 5 columns):
Key: &lt;geo_value, time_value, version&gt;
        geo_value time_value    version percent_cli case_rate_7d_av
           &lt;char&gt;     &lt;Date&gt;     &lt;Date&gt;       &lt;num&gt;           &lt;num&gt;
     1:        ma 2020-01-29 2021-01-10          NA              NA
     2:        ma 2020-01-29 2021-01-11          NA              NA
     3:        ma 2020-01-29 2021-01-12          NA              NA
     4:        ma 2020-01-29 2021-01-13          NA              NA
     5:        ma 2020-01-29 2021-01-14          NA              NA
    ---                                                            
139186:        ma 2021-12-11 2021-12-16    2.306966              NA
139187:        ma 2021-12-11 2021-12-17    2.281141              NA
139188:        ma 2021-12-12 2021-12-16    2.333759              NA
139189:        ma 2021-12-12 2021-12-17    2.369756              NA
139190:        ma 2021-12-13 2021-12-17    2.256551              NA</code></pre>
</div>
</div>
</section>
<section id="fitting-and-predicting-with-linear-model" class="slide level2">
<h2>Fitting and predicting with linear model</h2>
<ul>
<li>Define <code>lm_mod_pred()</code>: A function that fits a linear model to forecast cases based on the <code>percent_cli</code> predictor.</li>
<li>Use <code>predict()</code> with a 90% prediction interval.</li>
<li>Save the actual cases to compare to the nowcasts later.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a></a>lm_mod_pred <span class="ot">&lt;-</span> <span class="cf">function</span>(data, ...) {</span>
<span id="cb113-2"><a></a>  <span class="co"># Linear model</span></span>
<span id="cb113-3"><a></a>  model <span class="ot">&lt;-</span> <span class="fu">lm</span>(case_rate_7d_av <span class="sc">~</span> percent_cli, <span class="at">data =</span> data)</span>
<span id="cb113-4"><a></a></span>
<span id="cb113-5"><a></a>  <span class="co"># Make predictions</span></span>
<span id="cb113-6"><a></a>  predictions <span class="ot">=</span> <span class="fu">predict</span>(model,</span>
<span id="cb113-7"><a></a>                        <span class="at">newdata =</span> data <span class="sc">|&gt;</span></span>
<span id="cb113-8"><a></a>                          <span class="fu">fill</span>(percent_cli, <span class="at">.direction =</span> <span class="st">"down"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb113-9"><a></a>                          <span class="fu">filter</span>(time_value <span class="sc">==</span> <span class="fu">max</span>(time_value)),</span>
<span id="cb113-10"><a></a>                        <span class="at">interval =</span> <span class="st">"prediction"</span>, <span class="at">level =</span> <span class="fl">0.9</span>)</span>
<span id="cb113-11"><a></a>  </span>
<span id="cb113-12"><a></a>  <span class="co"># Pull off real-time value for later comparison to the nowcast value</span></span>
<span id="cb113-13"><a></a>  real_time_val <span class="ot">=</span> data <span class="sc">|&gt;</span> <span class="fu">filter</span>(time_value <span class="sc">==</span> <span class="fu">max</span>(time_value)) <span class="sc">|&gt;</span> <span class="fu">pull</span>(case_rate_7d_av)</span>
<span id="cb113-14"><a></a>  </span>
<span id="cb113-15"><a></a>  <span class="co"># Could clip predictions and bounds at 0</span></span>
<span id="cb113-16"><a></a>  <span class="fu">return</span>(<span class="fu">data.frame</span>(predictions, <span class="at">actual_nowcast_date =</span> <span class="fu">max</span>(data<span class="sc">$</span>time_value), <span class="at">real_time_val =</span> real_time_val)) </span>
<span id="cb113-17"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="nowcasting-with-epix_slide-2" class="slide level2">
<h2>Nowcasting with <code>epix_slide()</code></h2>
<ul>
<li><span class="primary"><strong>Specify targets</strong></span>: Define the target dates for nowcasting (e.g., 1st of each month) &amp; adjust training data to include the lag for the latent case data.</li>
<li><span class="primary"><strong>Sliding window</strong></span>: Use <code>epix_slide()</code> to apply the linear model across a sliding window of data for each region.</li>
<li><span class="primary"><strong>Training-test split</strong></span>: Use the last 30 days of data to train and predict cases for each target nowcast date.</li>
</ul>
</section>
<section id="nowcasting-with-epix_slide-3" class="slide level2">
<h2>Nowcasting with <code>epix_slide()</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a></a><span class="co"># Define the reference time points (to give the training/test split)</span></span>
<span id="cb114-2"><a></a>targeted_nowcast_dates <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">as.Date</span>(<span class="st">"2021-04-01"</span>), <span class="fu">as.Date</span>(<span class="st">"2021-11-01"</span>), <span class="at">by =</span> <span class="st">"1 month"</span>) </span>
<span id="cb114-3"><a></a>ref_time_values <span class="ot">=</span> targeted_nowcast_dates <span class="sc">+</span> <span class="dv">1</span> <span class="co"># + 1 because the case data is 1 day latent. </span></span>
<span id="cb114-4"><a></a><span class="co"># Determine this from revision_summary(y)</span></span>
<span id="cb114-5"><a></a></span>
<span id="cb114-6"><a></a><span class="co"># Use epix_slide to perform the nowcasting with a training-test split</span></span>
<span id="cb114-7"><a></a>nowcast_res <span class="ot">&lt;-</span> archive <span class="sc">|&gt;</span></span>
<span id="cb114-8"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">|&gt;</span></span>
<span id="cb114-9"><a></a>  <span class="fu">epix_slide</span>(</span>
<span id="cb114-10"><a></a>    <span class="at">.f =</span> lm_mod_pred,  <span class="co"># Pass the function defined above</span></span>
<span id="cb114-11"><a></a>    <span class="at">.before =</span> <span class="dv">30</span>,   <span class="co"># Training period of 30 days</span></span>
<span id="cb114-12"><a></a>    <span class="at">.versions =</span> ref_time_values, <span class="co"># Determines the day where training data goes up to (not inclusive)</span></span>
<span id="cb114-13"><a></a>    <span class="at">.new_col_name =</span> <span class="st">"res"</span></span>
<span id="cb114-14"><a></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb114-15"><a></a>  <span class="fu">unnest</span>() <span class="sc">|&gt;</span></span>
<span id="cb114-16"><a></a>  <span class="fu">mutate</span>(<span class="at">targeted_nowcast_date =</span> targeted_nowcast_dates,</span>
<span id="cb114-17"><a></a>         <span class="at">time_value =</span> actual_nowcast_date)</span>
<span id="cb114-18"><a></a></span>
<span id="cb114-19"><a></a><span class="co"># Take a peek at the results</span></span>
<span id="cb114-20"><a></a><span class="fu">head</span>(nowcast_res, <span class="at">n =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 9
# Groups:   geo_value [1]
  geo_value version      fit   lwr   upr actual_nowcast_date real_time_val
  &lt;chr&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;date&gt;                      &lt;dbl&gt;
1 ma        2021-04-02 2114. 1975. 2254. 2021-04-01                  1556.
# ℹ 2 more variables: targeted_nowcast_date &lt;date&gt;, time_value &lt;date&gt;</code></pre>
</div>
</div>
</section>
<section id="visualizing-nowcasts-vs.-actual-values" class="slide level2">
<h2>Visualizing nowcasts vs.&nbsp;actual values</h2>
<p>Merge the nowcast results with the latest data for more direct comparison:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a></a>x_latest <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(archive, <span class="fu">max</span>(archive<span class="sc">$</span>DT<span class="sc">$</span>version)) <span class="sc">|&gt;</span></span>
<span id="cb116-2"><a></a>  <span class="fu">select</span>(<span class="sc">-</span>percent_cli) </span>
<span id="cb116-3"><a></a></span>
<span id="cb116-4"><a></a>res <span class="ot">&lt;-</span> nowcast_res <span class="sc">|&gt;</span> <span class="fu">left_join</span>(x_latest, <span class="at">by =</span> <span class="fu">join_by</span>(geo_value, time_value))</span>
<span id="cb116-5"><a></a></span>
<span id="cb116-6"><a></a>res</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 × 10
# Groups:   geo_value [1]
  geo_value version       fit    lwr   upr actual_nowcast_date real_time_val
  &lt;chr&gt;     &lt;date&gt;      &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;date&gt;                      &lt;dbl&gt;
1 ma        2021-04-02 2114.  1975.  2254. 2021-04-01                  1556.
2 ma        2021-05-02 1083.   851.  1316. 2021-05-01                   869.
3 ma        2021-06-02  353.   164.   541. 2021-06-01                   117.
4 ma        2021-07-02   57.1   11.3  103. 2021-07-01                    59 
5 ma        2021-08-02  513.   284.   742. 2021-08-01                   572.
6 ma        2021-09-02 1207.   888.  1527. 2021-09-01                  1099.
7 ma        2021-10-02 1575.  1357.  1793. 2021-09-30                  1069 
8 ma        2021-11-02 1299.  1257.  1340. 2021-11-01                   891.
# ℹ 3 more variables: targeted_nowcast_date &lt;date&gt;, time_value &lt;date&gt;,
#   case_rate_7d_av &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="visualizing-nowcasts-vs.-actual-values-1" class="slide level2">
<h2>Visualizing nowcasts vs.&nbsp;actual values</h2>
<p>Now, plot the predictions &amp; real-time values on top of latest COVID-19 cases using <code>ggplot2</code>:</p>

<img data-src="day1-afternoon_files/figure-revealjs/plot-simple-lr-nowcast-res-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="evaluation-using-mae-4" class="slide level2">
<h2>Evaluation using MAE</h2>
<ul>
<li><p>Finally, we numerically evaluate our nowcasts using MAE.</p></li>
<li><p>Shows that the nowcast errors are lower than those of the real-time estimates.</p></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a></a><span class="co"># Calculate the absolute error between actual and nowcasted COVID-19 cases</span></span>
<span id="cb118-2"><a></a>mae_data_cases <span class="ot">&lt;-</span> res <span class="sc">|&gt;</span> </span>
<span id="cb118-3"><a></a>  <span class="fu">mutate</span>(<span class="at">nc_abs_error =</span> <span class="fu">abs</span>(case_rate_7d_av <span class="sc">-</span> fit),  <span class="co"># Nowcast vs Finalized cases (7-day average)</span></span>
<span id="cb118-4"><a></a>         <span class="at">rt_abs_error =</span> <span class="fu">abs</span>(case_rate_7d_av <span class="sc">-</span> real_time_val))  <span class="co"># Real-Time vs Finalized cases</span></span>
<span id="cb118-5"><a></a></span>
<span id="cb118-6"><a></a><span class="co"># Compute the MAE (mean of absolute errors)</span></span>
<span id="cb118-7"><a></a>mae_value_cases <span class="ot">&lt;-</span> mae_data_cases <span class="sc">|&gt;</span> </span>
<span id="cb118-8"><a></a>  <span class="fu">summarise</span>(<span class="at">nc_MAE =</span> <span class="fu">mean</span>(nc_abs_error),</span>
<span id="cb118-9"><a></a>            <span class="at">rt_MAE =</span> <span class="fu">mean</span>(rt_abs_error))</span>
<span id="cb118-10"><a></a>knitr<span class="sc">::</span><span class="fu">kable</span>(mae_value_cases)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top">
<thead>
<tr class="header">
<th style="text-align: left;">geo_value</th>
<th style="text-align: right;">nc_MAE</th>
<th style="text-align: right;">rt_MAE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ma</td>
<td style="text-align: right;">152.5013</td>
<td style="text-align: right;">228.5357</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="evaluation-using-mae-5" class="slide level2">
<h2>Evaluation using MAE</h2>

<img data-src="day1-afternoon_files/figure-revealjs/abs-error-plot-cases-1.svg" class="quarto-figure quarto-figure-center r-stretch"><p>The elevated errors at both ends highlight periods where the discrepancies between the real-time and nowcast estimates are most pronounced.</p>
</section>
<section id="takeaways" class="slide level2">
<h2>Takeaways</h2>
<p><span class="primary"><strong>Goal</strong></span>: Predict COVID-19 cases using %CLI, overcoming delays in report data.</p>
<p>Main Steps:</p>
<ol type="1">
<li><p><span class="primary"><strong>Fetch Data</strong></span>: Collect case and %CLI data.</p></li>
<li><p><span class="primary"><strong>Merge Data</strong></span>: Align datasets with <code>epix_merge()</code> and fill missing values.</p></li>
<li><p><span class="primary"><strong>Model</strong></span>: Fit a linear model to predict cases.</p></li>
<li><p><span class="primary"><strong>Nowcast</strong></span>: Apply dynamic forecasting with <code>epix_slide()</code>.</p></li>
<li><p><span class="primary"><strong>Evaluate</strong></span>: Calculate error measures and numerically and visually assess the results.</p></li>
</ol>
<p>Overall, nowcasting, based on the linear model, provided a closer approximation of true cases compared to the real-time values.</p>
</section>
<section id="aside-on-nowcasting" class="slide level2">
<h2>Aside on nowcasting</h2>
<ul>
<li><p>To some Epis, “nowcasting” can be equated with “estimate the time-varying instantaneous reproduction number, <span class="math inline">\(R_t\)</span>”</p></li>
<li><p>Ex. using the number of reported COVID-19 cases in British Columbia between Jan.&nbsp;2020 and Apr.&nbsp;15, 2023.</p></li>
</ul>
<!-- This data is the number of reported COVID-19 cases in British Columbia between January 2020 and April 15, 2023. The values are.up-to-date as of August 2023. -->
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="day1-afternoon_files/figure-revealjs/nowcasting-1.svg" class="quarto-figure quarto-figure-center" height="400"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><p>Group built <a href="https://dajmcdon.github.io/rtestim"><code>{rtestim}</code></a> doing for this nonparametrically.</p></li>
<li><p>We may come back to this later…</p></li>
</ul>


</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<p><img src="gfx/delphi.jpg" class="slide-logo"></p>
<div class="footer footer-default">
<p>Nowcasting — <a href="https://github.com/cmu-delphi/insightnet-workshop-2024">cmu-delphi/insightnet-workshop-2024</a></p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: false,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp("https:\/\/cmu-delphi\.github\.io\/insightnet-workshop-2024\/");
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
              // target, if specified
              link.setAttribute("target", "_blank");
              if (link.getAttribute("rel") === null) {
                link.setAttribute("rel", "noopener");
              }
              // default icon
              link.classList.add("external");
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>