<!DOCTYPE html>
<html lang="en-CA"><head>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.5.57">

  <meta name="author" content="">
  <title>InsightNet EpiData Workshop 2024 – day1-afternoon</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto.css">
  <link rel="stylesheet" href="tachyons-minimal.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">


<section class="slide level2">

<p><!-- Set any of the above to "" to omit them --></p>
<p><!-- Or adjust the formatting in _titleslide.qmd --></p>
<div class="flex">
<div class="w-20">

</div>
<section id="section" class="slide level2 w-80" data-background-image="gfx/cover-art-1.svg" data-background-position="bottom">
<h2>Explore, clean &amp; transform data</h2>
<h3 id="section-1"></h3>
<p><br></p>
<h4 id="section-2"></h4>
<p></p>
<p></p>
</section>
</div>
</section>
<section id="outline" class="slide level2">
<h2>Outline</h2>
<ol type="1">
<li><p>Essentials of <code>dplyr</code> and <code>tidyr</code></p></li>
<li><p>Epiverse software ecosystem</p></li>
<li><p>Panel and versioned data in the epiverse</p></li>
<li><p>Basic Nowcasting using <code>epiprocess</code></p></li>
<li><p>Motivating case study</p></li>
</ol>
</section>
<section>
<section id="essentials-of-dplyr-and-tidyr" class="title-slide slide level1 center">
<h1>Essentials of <code>dplyr</code> and <code>tidyr</code></h1>

</section>
<section id="down-with-spreadsheets-for-data-manipulation" class="slide level2">
<h2>Down with Spreadsheets for Data Manipulation</h2>
<ul>
<li>Spreadsheets make it difficult to rerun analyses consistently.</li>
<li>Using R (and <code>dplyr</code>) allows for:
<ul>
<li>Reproducibility</li>
<li>Ease of modification</li>
</ul></li>
<li><strong>Recommendation</strong>: Avoid manual edits; instead, use code for transformations.</li>
</ul>
</section>
<section id="introduction-to-dplyr" class="slide level2">
<h2>Introduction to <code>dplyr</code></h2>
<ul>
<li><code>dplyr</code> is a powerful package in R for data manipulation.</li>
<li>It is part of the <strong>tidyverse</strong>, which includes a collection of packages designed to work together.</li>
<li>We focus on basic operations like selecting and filtering data.</li>
<li>Make sure to load the necessary libraries before using <code>dplyr</code>.</li>
</ul>
</section>
<section id="meet-the-palmers" class="slide level2">
<h2>Meet the Palmers</h2>

<img data-src="gfx/meet_the_palmers.png" style="width: 70%;" class="r-stretch"><p><small><a href="https://allisonhorst.github.io/palmerpenguins/">Illustration from the palmerpenguins website</a></small></p>
</section>
<section id="working-with-the-palmerpenguins-dataset" class="slide level2">
<h2>Working with the <code>palmerpenguins</code> Dataset</h2>
<ul>
<li>The <code>palmerpenguins</code> dataset is included in the <code>palmerpenguins</code> package.</li>
<li>Load both the <code>tidyverse</code> and <code>palmerpenguins</code> libraries to access and explore the dataset.</li>
<li>The dataset includes measurements of penguins such as species, bill length, flipper length, and body mass.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 8
   species island    bill_length_mm bill_depth_mm flipper_length_mm body_mass_g
   &lt;fct&gt;   &lt;fct&gt;              &lt;dbl&gt;         &lt;dbl&gt;             &lt;int&gt;       &lt;int&gt;
 1 Adelie  Torgersen           39.1          18.7               181        3750
 2 Adelie  Torgersen           39.5          17.4               186        3800
 3 Adelie  Torgersen           40.3          18                 195        3250
 4 Adelie  Torgersen           NA            NA                  NA          NA
 5 Adelie  Torgersen           36.7          19.3               193        3450
 6 Adelie  Torgersen           39.3          20.6               190        3650
 7 Adelie  Torgersen           38.9          17.8               181        3625
 8 Adelie  Torgersen           39.2          19.6               195        4675
 9 Adelie  Torgersen           34.1          18.1               193        3475
10 Adelie  Torgersen           42            20.2               190        4250
# ℹ 334 more rows
# ℹ 2 more variables: sex &lt;fct&gt;, year &lt;int&gt;</code></pre>
</div>
</div>
</section>
<section id="ways-to-inspect-the-dataset" class="slide level2">
<h2>Ways to Inspect the Dataset</h2>
<ul>
<li>Use <code>head()</code> to view the first 6 row of the data (<code>tail()</code> to view the last 6 rows)</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a></a><span class="fu">head</span>(penguins)  <span class="co"># First 6 rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  species island    bill_length_mm bill_depth_mm flipper_length_mm body_mass_g
  &lt;fct&gt;   &lt;fct&gt;              &lt;dbl&gt;         &lt;dbl&gt;             &lt;int&gt;       &lt;int&gt;
1 Adelie  Torgersen           39.1          18.7               181        3750
2 Adelie  Torgersen           39.5          17.4               186        3800
3 Adelie  Torgersen           40.3          18                 195        3250
4 Adelie  Torgersen           NA            NA                  NA          NA
5 Adelie  Torgersen           36.7          19.3               193        3450
6 Adelie  Torgersen           39.3          20.6               190        3650
# ℹ 2 more variables: sex &lt;fct&gt;, year &lt;int&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a></a><span class="co">#tail(penguins)  # Last 6 rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ways-to-inspect-the-dataset-1" class="slide level2">
<h2>Ways to Inspect the Dataset</h2>
<ul>
<li><code>glimpse()</code> to get a compact overview of the dataset.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a></a><span class="fu">glimpse</span>(penguins)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 344
Columns: 8
$ species           &lt;fct&gt; Adelie, Adelie, Adelie, Adelie, Adelie, Adelie, Adel…
$ island            &lt;fct&gt; Torgersen, Torgersen, Torgersen, Torgersen, Torgerse…
$ bill_length_mm    &lt;dbl&gt; 39.1, 39.5, 40.3, NA, 36.7, 39.3, 38.9, 39.2, 34.1, …
$ bill_depth_mm     &lt;dbl&gt; 18.7, 17.4, 18.0, NA, 19.3, 20.6, 17.8, 19.6, 18.1, …
$ flipper_length_mm &lt;int&gt; 181, 186, 195, NA, 193, 190, 181, 195, 193, 190, 186…
$ body_mass_g       &lt;int&gt; 3750, 3800, 3250, NA, 3450, 3650, 3625, 4675, 3475, …
$ sex               &lt;fct&gt; male, female, female, NA, female, male, female, male…
$ year              &lt;int&gt; 2007, 2007, 2007, 2007, 2007, 2007, 2007, 2007, 2007…</code></pre>
</div>
</div>
</section>
<section id="creating-tibbles" class="slide level2">
<h2>Creating Tibbles</h2>
<ul>
<li><strong>Tibbles</strong>: Modern data frames with enhanced features.</li>
<li>Rows represent <strong>observations</strong> (or cases).</li>
<li>Columns represent <strong>variables</strong> (or features).</li>
<li>You can create tibbles manually using the <code>tibble()</code> function.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a></a><span class="fu">tibble</span>(<span class="at">x =</span> letters, <span class="at">y =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">26</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 26 × 2
   x         y
   &lt;chr&gt; &lt;int&gt;
 1 a         1
 2 b         2
 3 c         3
 4 d         4
 5 e         5
 6 f         6
 7 g         7
 8 h         8
 9 i         9
10 j        10
# ℹ 16 more rows</code></pre>
</div>
</div>
</section>
<section id="selecting-columns-with-select" class="slide level2">
<h2>Selecting Columns with <code>select()</code></h2>
<ul>
<li>The <code>select()</code> function is used to pick specific columns from your dataset.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a></a><span class="fu">select</span>(penguins, species, body_mass_g)  <span class="co"># Select the 'species' and 'body_mass_g' columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 2
   species body_mass_g
   &lt;fct&gt;         &lt;int&gt;
 1 Adelie         3750
 2 Adelie         3800
 3 Adelie         3250
 4 Adelie           NA
 5 Adelie         3450
 6 Adelie         3650
 7 Adelie         3625
 8 Adelie         4675
 9 Adelie         3475
10 Adelie         4250
# ℹ 334 more rows</code></pre>
</div>
</div>
</section>
<section id="selecting-columns-with-select-1" class="slide level2">
<h2>Selecting Columns with <code>select()</code></h2>
<ul>
<li>You can exclude columns by prefixing the column names with a minus sign <code>-</code>.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a></a><span class="fu">select</span>(penguins, <span class="sc">-</span>species)  <span class="co"># Exclude the 'species' column from the dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 7
   island bill_length_mm bill_depth_mm flipper_length_mm body_mass_g sex    year
   &lt;fct&gt;           &lt;dbl&gt;         &lt;dbl&gt;             &lt;int&gt;       &lt;int&gt; &lt;fct&gt; &lt;int&gt;
 1 Torge…           39.1          18.7               181        3750 male   2007
 2 Torge…           39.5          17.4               186        3800 fema…  2007
 3 Torge…           40.3          18                 195        3250 fema…  2007
 4 Torge…           NA            NA                  NA          NA &lt;NA&gt;   2007
 5 Torge…           36.7          19.3               193        3450 fema…  2007
 6 Torge…           39.3          20.6               190        3650 male   2007
 7 Torge…           38.9          17.8               181        3625 fema…  2007
 8 Torge…           39.2          19.6               195        4675 male   2007
 9 Torge…           34.1          18.1               193        3475 &lt;NA&gt;   2007
10 Torge…           42            20.2               190        4250 &lt;NA&gt;   2007
# ℹ 334 more rows</code></pre>
</div>
</div>
<ul>
<li>So, this is useful when you want to keep only certain columns or remove unnecessary ones.</li>
</ul>
</section>
<section id="extracting-columns-with-pull" class="slide level2">
<h2>Extracting Columns with <code>pull()</code></h2>
<ul>
<li><code>pull()</code>: Extract a column as a vector.</li>
<li>Let’s try this with the <code>species</code> column…</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a></a><span class="fu">pull</span>(penguins, species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
  [8] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [15] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [22] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [29] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [36] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [43] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [50] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [57] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [64] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [71] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [78] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [85] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [92] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
 [99] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
[106] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
[113] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
[120] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
[127] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
[134] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
[141] Adelie    Adelie    Adelie    Adelie    Adelie    Adelie    Adelie   
[148] Adelie    Adelie    Adelie    Adelie    Adelie    Gentoo    Gentoo   
[155] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[162] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[169] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[176] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[183] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[190] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[197] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[204] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[211] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[218] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[225] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[232] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[239] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[246] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[253] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[260] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[267] Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo    Gentoo   
[274] Gentoo    Gentoo    Gentoo    Chinstrap Chinstrap Chinstrap Chinstrap
[281] Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap
[288] Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap
[295] Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap
[302] Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap
[309] Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap
[316] Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap
[323] Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap
[330] Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap
[337] Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap Chinstrap
[344] Chinstrap
Levels: Adelie Chinstrap Gentoo</code></pre>
</div>
</div>
</section>
<section id="filtering-rows-with-filter" class="slide level2">
<h2>Filtering Rows with <code>filter()</code></h2>
<ul>
<li>The <code>filter()</code> function allows you to select rows that meet specific conditions.</li>
<li>Conditions can involve column values, such as selecting only “Gentoo” penguins or filtering based on measurements like flipper length.</li>
<li>This enables you to narrow down your dataset to focus on relevant data.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a></a><span class="fu">filter</span>(penguins, species <span class="sc">==</span> <span class="st">"Gentoo"</span>, flipper_length_mm <span class="sc">&lt;</span> <span class="dv">208</span>)  <span class="co"># Filter Gentoo penguins with flipper length &lt; 208mm</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 8
  species island bill_length_mm bill_depth_mm flipper_length_mm body_mass_g
  &lt;fct&gt;   &lt;fct&gt;           &lt;dbl&gt;         &lt;dbl&gt;             &lt;int&gt;       &lt;int&gt;
1 Gentoo  Biscoe           45.1          14.5               207        5050
2 Gentoo  Biscoe           48.4          14.4               203        4625
# ℹ 2 more variables: sex &lt;fct&gt;, year &lt;int&gt;</code></pre>
</div>
</div>
</section>
<section id="combining-select-and-filter-functions" class="slide level2">
<h2>Combining <code>select()</code> and <code>filter()</code> Functions</h2>
<ul>
<li>You can combine <code>select()</code> and <code>filter()</code> functions to refine the dataset further.</li>
<li>Use <code>select()</code> to choose columns and <code>filter()</code> to narrow rows based on conditions.</li>
<li>This helps in extracting the exact data needed for analysis.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a></a><span class="fu">select</span>(<span class="fu">filter</span>(penguins, species <span class="sc">==</span> <span class="st">"Gentoo"</span>, flipper_length_mm <span class="sc">&lt;</span> <span class="dv">208</span>), species, flipper_length_mm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  species flipper_length_mm
  &lt;fct&gt;               &lt;int&gt;
1 Gentoo                207
2 Gentoo                203</code></pre>
</div>
</div>
</section>
<section id="using-the-pipe-operator" class="slide level2">
<h2>Using the Pipe Operator <code>%&gt;%</code></h2>
<ul>
<li>The pipe operator (<code>%&gt;%</code>) makes code more readable by chaining multiple operations together.</li>
<li>The output of one function is automatically passed to the next function.</li>
<li>This allows you to perform multiple steps (e.g., <code>select()</code> followed by <code>filter()</code>) in a clear and concise manner.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a></a><span class="co"># This code reads more like poetry!</span></span>
<span id="cb19-2"><a></a>penguins <span class="sc">%&gt;%</span> </span>
<span id="cb19-3"><a></a>  <span class="fu">select</span>(species, flipper_length_mm) <span class="sc">%&gt;%</span></span>
<span id="cb19-4"><a></a>  <span class="fu">filter</span>(species <span class="sc">==</span> <span class="st">"Gentoo"</span>, flipper_length_mm <span class="sc">&lt;</span> <span class="dv">208</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  species flipper_length_mm
  &lt;fct&gt;               &lt;int&gt;
1 Gentoo                207
2 Gentoo                203</code></pre>
</div>
</div>
</section>
<section id="key-practices-in-dplyr" class="slide level2">
<h2>Key Practices in <code>dplyr</code></h2>
<ul>
<li>Use <strong>tibbles</strong> for easier data handling.</li>
<li>Use <strong><code>select()</code></strong> and <strong><code>filter()</code></strong> for data manipulation.</li>
<li>Use <strong><code>pull()</code></strong> to extract columns as vectors.</li>
<li>Use <strong><code>head()</code></strong>, <strong><code>tail()</code></strong>, and <strong><code>glimpse()</code></strong> for quick data inspection.</li>
<li>Chain functions with <strong><code>%&gt;%</code></strong> for cleaner code.</li>
</ul>
</section>
<section id="grouping-data-with-group_by" class="slide level2">
<h2>Grouping Data with <code>group_by()</code></h2>
<ul>
<li>Use <code>group_by()</code> to group data by one or more columns.</li>
<li>Allows performing operations on specific groups of data.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb21-2"><a></a>  <span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span></span>
<span id="cb21-3"><a></a>  <span class="fu">filter</span>(body_mass_g <span class="sc">==</span> <span class="fu">min</span>(body_mass_g, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 8
# Groups:   species [3]
  species   island bill_length_mm bill_depth_mm flipper_length_mm body_mass_g
  &lt;fct&gt;     &lt;fct&gt;           &lt;dbl&gt;         &lt;dbl&gt;             &lt;int&gt;       &lt;int&gt;
1 Adelie    Biscoe           36.5          16.6               181        2850
2 Adelie    Biscoe           36.4          17.1               184        2850
3 Gentoo    Biscoe           42.7          13.7               208        3950
4 Chinstrap Dream            46.9          16.6               192        2700
# ℹ 2 more variables: sex &lt;fct&gt;, year &lt;int&gt;</code></pre>
</div>
</div>
</section>
<section id="penguin-bill-length-and-depth" class="slide level2">
<h2>Penguin bill length and depth</h2>
<div style="text-align: center;">
<p><img data-src="gfx/bill_length_depth.png" style="width: 60%;"></p>
<p><small><a href="https://allisonhorst.github.io/palmerpenguins/">Illustration from the palmerpenguins website</a></small></p>
</div>
</section>
<section id="creating-new-columns-with-mutate" class="slide level2">
<h2>Creating New Columns with <code>mutate()</code></h2>
<ul>
<li><code>mutate()</code> is used to create new columns.</li>
<li>Perform calculations using existing columns and assign to new columns.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb23-2"><a></a>  <span class="fu">mutate</span>(<span class="at">bill_size_mm2 =</span> bill_depth_mm <span class="sc">*</span> bill_length_mm) <span class="sc">%&gt;%</span> </span>
<span id="cb23-3"><a></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(flipper_length_mm, body_mass_g, sex)) <span class="co"># too many cols to print</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 6
   species island    bill_length_mm bill_depth_mm  year bill_size_mm2
   &lt;fct&gt;   &lt;fct&gt;              &lt;dbl&gt;         &lt;dbl&gt; &lt;int&gt;         &lt;dbl&gt;
 1 Adelie  Torgersen           39.1          18.7  2007          731.
 2 Adelie  Torgersen           39.5          17.4  2007          687.
 3 Adelie  Torgersen           40.3          18    2007          725.
 4 Adelie  Torgersen           NA            NA    2007           NA 
 5 Adelie  Torgersen           36.7          19.3  2007          708.
 6 Adelie  Torgersen           39.3          20.6  2007          810.
 7 Adelie  Torgersen           38.9          17.8  2007          692.
 8 Adelie  Torgersen           39.2          19.6  2007          768.
 9 Adelie  Torgersen           34.1          18.1  2007          617.
10 Adelie  Torgersen           42            20.2  2007          848.
# ℹ 334 more rows</code></pre>
</div>
</div>
</section>
<section id="creating-new-columns-with-mutate-1" class="slide level2">
<h2>Creating New Columns with <code>mutate()</code></h2>
<ul>
<li><code>mutate()</code> can create multiple new columns in one step.</li>
<li>Logical comparisons (e.g., <code>sex == "male"</code>) can be used within <code>mutate()</code>.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb25-2"><a></a>  <span class="fu">mutate</span>(<span class="at">bill_size_mm2 =</span> bill_depth_mm <span class="sc">*</span> bill_length_mm, </span>
<span id="cb25-3"><a></a>         <span class="at">TF =</span> sex <span class="sc">==</span> <span class="st">"male"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb25-4"><a></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(flipper_length_mm, body_mass_g, year)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 7
   species island    bill_length_mm bill_depth_mm sex    bill_size_mm2 TF   
   &lt;fct&gt;   &lt;fct&gt;              &lt;dbl&gt;         &lt;dbl&gt; &lt;fct&gt;          &lt;dbl&gt; &lt;lgl&gt;
 1 Adelie  Torgersen           39.1          18.7 male            731. TRUE 
 2 Adelie  Torgersen           39.5          17.4 female          687. FALSE
 3 Adelie  Torgersen           40.3          18   female          725. FALSE
 4 Adelie  Torgersen           NA            NA   &lt;NA&gt;             NA  NA   
 5 Adelie  Torgersen           36.7          19.3 female          708. FALSE
 6 Adelie  Torgersen           39.3          20.6 male            810. TRUE 
 7 Adelie  Torgersen           38.9          17.8 female          692. FALSE
 8 Adelie  Torgersen           39.2          19.6 male            768. TRUE 
 9 Adelie  Torgersen           34.1          18.1 &lt;NA&gt;            617. NA   
10 Adelie  Torgersen           42            20.2 &lt;NA&gt;            848. NA   
# ℹ 334 more rows</code></pre>
</div>
</div>
</section>
<section id="combining-group_by-and-mutate" class="slide level2">
<h2>Combining <code>group_by()</code> and <code>mutate()</code></h2>
<ul>
<li>First, group data using <code>group_by()</code>.</li>
<li>Then, use <code>drop_na()</code> from <code>tidyr</code> to exclude rows with missing values.</li>
<li>Finally, <code>mutate</code> to perform calculations for each group.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb27-2"><a></a>  <span class="fu">drop_na</span>() <span class="sc">%&gt;%</span> <span class="co"># Remove all non-complete rows</span></span>
<span id="cb27-3"><a></a>  <span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span></span>
<span id="cb27-4"><a></a>  <span class="fu">mutate</span>(<span class="at">body_mass_median =</span> <span class="fu">median</span>(body_mass_g)) <span class="sc">%&gt;%</span> </span>
<span id="cb27-5"><a></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(flipper_length_mm, body_mass_g, sex)) <span class="sc">%&gt;%</span> </span>
<span id="cb27-6"><a></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 6
# Groups:   species [1]
  species island    bill_length_mm bill_depth_mm  year body_mass_median
  &lt;fct&gt;   &lt;fct&gt;              &lt;dbl&gt;         &lt;dbl&gt; &lt;int&gt;            &lt;dbl&gt;
1 Adelie  Torgersen           39.1          18.7  2007             3700
2 Adelie  Torgersen           39.5          17.4  2007             3700
3 Adelie  Torgersen           40.3          18    2007             3700
4 Adelie  Torgersen           36.7          19.3  2007             3700
5 Adelie  Torgersen           39.3          20.6  2007             3700
6 Adelie  Torgersen           38.9          17.8  2007             3700</code></pre>
</div>
</div>
</section>
<section id="conditional-calculations-in-mutate-with-if_else" class="slide level2">
<h2>Conditional Calculations in <code>mutate()</code> with <code>if_else()</code></h2>
<ul>
<li><code>if_else()</code> allows conditional logic within <code>mutate()</code>.</li>
<li>Perform different operations depending on conditions, like “big” or “small.”</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a></a>t <span class="ot">&lt;-</span> <span class="dv">800</span></span>
<span id="cb29-2"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb29-3"><a></a>  <span class="fu">mutate</span>(<span class="at">bill_size_mm2 =</span> bill_depth_mm <span class="sc">*</span> bill_length_mm, </span>
<span id="cb29-4"><a></a>         <span class="at">TF =</span> sex <span class="sc">==</span> <span class="st">"male"</span>,</span>
<span id="cb29-5"><a></a>         <span class="at">bill_size_binary =</span> <span class="fu">if_else</span>(bill_size_mm2 <span class="sc">&gt;</span> t, <span class="st">"big"</span>, <span class="st">"small"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb29-6"><a></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(flipper_length_mm, body_mass_g, sex)) <span class="sc">%&gt;%</span> <span class="co"># too many cols to print</span></span>
<span id="cb29-7"><a></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  species island    bill_length_mm bill_depth_mm  year bill_size_mm2 TF   
  &lt;fct&gt;   &lt;fct&gt;              &lt;dbl&gt;         &lt;dbl&gt; &lt;int&gt;         &lt;dbl&gt; &lt;lgl&gt;
1 Adelie  Torgersen           39.1          18.7  2007          731. TRUE 
2 Adelie  Torgersen           39.5          17.4  2007          687. FALSE
3 Adelie  Torgersen           40.3          18    2007          725. FALSE
4 Adelie  Torgersen           NA            NA    2007           NA  NA   
5 Adelie  Torgersen           36.7          19.3  2007          708. FALSE
6 Adelie  Torgersen           39.3          20.6  2007          810. TRUE 
# ℹ 1 more variable: bill_size_binary &lt;chr&gt;</code></pre>
</div>
</div>
</section>
<section id="summarizing-data-with-summarise" class="slide level2">
<h2>Summarizing Data with <code>summarise()</code></h2>
<ul>
<li><code>summarise()</code> reduces data to summary statistics (e.g., mean, median).</li>
<li>Typically used after <code>group_by()</code> to summarize each group.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb31-2"><a></a>  <span class="fu">drop_na</span>() <span class="sc">%&gt;%</span></span>
<span id="cb31-3"><a></a>  <span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span></span>
<span id="cb31-4"><a></a>  <span class="fu">summarise</span>(<span class="at">body_mass_median =</span> <span class="fu">median</span>(body_mass_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  species   body_mass_median
  &lt;fct&gt;                &lt;dbl&gt;
1 Adelie                3700
2 Chinstrap             3700
3 Gentoo                5050</code></pre>
</div>
</div>
</section>
<section id="using-summarise-with-multiple-calculations" class="slide level2">
<h2>Using <code>summarise()</code> with Multiple Calculations</h2>
<ul>
<li>Use <code>summarise()</code> to calculate multiple summary statistics at once.</li>
<li>Include multiple columns and functions in a single call.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb33-2"><a></a>  <span class="fu">drop_na</span>() <span class="sc">%&gt;%</span></span>
<span id="cb33-3"><a></a>  <span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span></span>
<span id="cb33-4"><a></a>  <span class="fu">summarise</span>(<span class="at">body_mass_median =</span> <span class="fu">median</span>(body_mass_g), </span>
<span id="cb33-5"><a></a>            <span class="at">bill_depth_median =</span> <span class="fu">median</span>(bill_depth_mm),</span>
<span id="cb33-6"><a></a>            <span class="at">flipper_length_median =</span> <span class="fu">median</span>(flipper_length_mm),</span>
<span id="cb33-7"><a></a>            <span class="at">bill_length_mm =</span> <span class="fu">median</span>(bill_length_mm))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 5
  species   body_mass_median bill_depth_median flipper_length_median
  &lt;fct&gt;                &lt;dbl&gt;             &lt;dbl&gt;                 &lt;dbl&gt;
1 Adelie                3700              18.4                   190
2 Chinstrap             3700              18.4                   196
3 Gentoo                5050              15                     216
# ℹ 1 more variable: bill_length_mm &lt;dbl&gt;</code></pre>
</div>
</div>
<ul>
<li>Yikes! This can get long. Is there a way to condense this?</li>
</ul>
</section>
<section id="using-across-to-apply-functions-to-multiple-columns-in-one-swoop" class="slide level2">
<h2>Using <code>across()</code> to Apply Functions to Multiple Columns in one Swoop</h2>
<ul>
<li><code>across()</code> applies a function (e.g., median) to multiple columns.</li>
<li>It’s especially useful for summarizing multiple numeric columns in one step.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb35-2"><a></a>  <span class="fu">drop_na</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb35-3"><a></a>  <span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span></span>
<span id="cb35-4"><a></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="fu">where</span>(is.numeric), median))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 6
  species   bill_length_mm bill_depth_mm flipper_length_mm body_mass_g  year
  &lt;fct&gt;              &lt;dbl&gt;         &lt;dbl&gt;             &lt;dbl&gt;       &lt;dbl&gt; &lt;dbl&gt;
1 Adelie              38.8          18.4               190        3700  2008
2 Chinstrap           49.6          18.4               196        3700  2008
3 Gentoo              47.4          15                 216        5050  2008</code></pre>
</div>
</div>
</section>
<section id="using-count-to-aggregate-data" class="slide level2">
<h2>Using <code>count()</code> to Aggregate Data</h2>
<ul>
<li><code>count()</code> is a shortcut for grouping and summarizing the data:</li>
</ul>
<p>For example, if we want to count the number of penguins by species, then</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a></a>penguins_count <span class="ot">&lt;-</span> penguins <span class="sc">%&gt;%</span></span>
<span id="cb37-2"><a></a>  <span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span></span>
<span id="cb37-3"><a></a>  <span class="fu">summarize</span>(<span class="at">count =</span> <span class="fu">n</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>is equivalent to</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a></a>penguins_count <span class="ot">&lt;-</span> penguins <span class="sc">%&gt;%</span></span>
<span id="cb38-2"><a></a>  <span class="fu">count</span>(species)</span>
<span id="cb38-3"><a></a></span>
<span id="cb38-4"><a></a>penguins_count <span class="co"># Let's see what the counts are.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  species       n
  &lt;fct&gt;     &lt;int&gt;
1 Adelie      152
2 Chinstrap    68
3 Gentoo      124</code></pre>
</div>
</div>
</section>
<section id="tidy-data" class="slide level2">
<h2>Tidy Data</h2>
<blockquote>
<p>“Happy families are all alike; every unhappy family is unhappy in its own way.” — Leo Tolstoy</p>
</blockquote>
<ul>
<li><strong>Tidy datasets</strong> are like happy families: consistent, standardized, and easy to work with.<br>
</li>
<li><strong>Messy datasets</strong> are like unhappy families: each one messy in its own unique way.<br>
In this section:</li>
<li>We’ll define what makes data <em>tidy</em> and how to transform between the tidy and messy formats.</li>
</ul>
</section>
<section id="what-is-tidy-data" class="slide level2">
<h2>What is Tidy Data?</h2>
<ul>
<li>Tidy data follows a consistent structure: <strong>each row represents one observation, and each column represents one variable.</strong></li>
</ul>
<!-- * **Example:** Suppose we have the following messy/wide dataset of the counts of two species of penguins on two islands: -->
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a></a><span class="co"># Simple example of messy data (wide format)</span></span>
<span id="cb40-2"><a></a>penguins_wide <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb40-3"><a></a>  <span class="at">island =</span> <span class="fu">c</span>(<span class="st">"Biscoe"</span>, <span class="st">"Biscoe"</span>, <span class="st">"Dream"</span>),</span>
<span id="cb40-4"><a></a>  <span class="at">year =</span> <span class="fu">c</span>(<span class="dv">2007</span>, <span class="dv">2008</span>, <span class="dv">2007</span>),</span>
<span id="cb40-5"><a></a>  <span class="at">Adelie =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">18</span>, <span class="dv">20</span>),</span>
<span id="cb40-6"><a></a>  <span class="at">Gentoo =</span> <span class="fu">c</span>(<span class="dv">34</span>, <span class="dv">46</span>, <span class="dv">0</span>)</span>
<span id="cb40-7"><a></a>)</span>
<span id="cb40-8"><a></a>penguins_wide</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  island  year Adelie Gentoo
  &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 Biscoe  2007     10     34
2 Biscoe  2008     18     46
3 Dream   2007     20      0</code></pre>
</div>
</div>
</section>
<section id="tidying-messy-data-with-pivot_longer" class="slide level2">
<h2>Tidying Messy Data with <code>pivot_longer()</code></h2>
<ul>
<li>To turn messy data into tidy data, we often use the <code>tidyr</code> package in the tidyverse.</li>
<li>Use <code>pivot_longer()</code> to convert data from <strong>wide format</strong> (multiple columns for the same variable) to <strong>long format</strong> (one column per variable).</li>
<li>This makes it easier to perform group-based calculations or create visualizations.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  island  year species count
  &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt;
1 Biscoe  2007 Adelie     10
2 Biscoe  2007 Gentoo     34
3 Biscoe  2008 Adelie     18
4 Biscoe  2008 Gentoo     46
5 Dream   2007 Adelie     20
6 Dream   2007 Gentoo      0</code></pre>
</div>
</div>
</section>
<section id="making-data-wider-with-pivot_wider" class="slide level2">
<h2>Making Data Wider with <code>pivot_wider()</code></h2>
<ul>
<li>Sometimes, you need to convert data from long format to wide format using <code>pivot_wider()</code>.</li>
<li>This can be useful when you want to separate variables into individual columns.</li>
<li>Let’s try converting <code>penguins_tidy</code> back to <code>penguins_wide</code>!</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a></a><span class="co"># Pivoting long data back to wide format</span></span>
<span id="cb43-2"><a></a>penguins_wide_back <span class="ot">&lt;-</span> penguins_tidy <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> species, <span class="at">values_from =</span> count)</span>
<span id="cb43-4"><a></a></span>
<span id="cb43-5"><a></a>penguins_wide_back</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  island  year Adelie Gentoo
  &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 Biscoe  2007     10     34
2 Biscoe  2008     18     46
3 Dream   2007     20      0</code></pre>
</div>
</div>
</section>
<section id="complete-and-fill-to-handle-missing-data" class="slide level2">
<h2><code>complete()</code> and <code>fill()</code> to Handle Missing Data</h2>
<div style="font-size: 0.7em;">
<ol type="1">
<li><strong><code>complete()</code></strong>: Adds missing rows for combinations of specified variables.</li>
<li><strong><code>fill()</code></strong>: Fills missing values in columns, typically from previous or next available values (default is LOCF).</li>
</ol>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a></a><span class="co"># First, use complete() to add missing year (2008 for Dream)</span></span>
<span id="cb45-2"><a></a>penguins_complete <span class="ot">&lt;-</span> penguins_wide <span class="sc">%&gt;%</span></span>
<span id="cb45-3"><a></a>  <span class="fu">complete</span>(island, year)</span>
<span id="cb45-4"><a></a>penguins_complete</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 4
  island  year Adelie Gentoo
  &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 Biscoe  2007     10     34
2 Biscoe  2008     18     46
3 Dream   2007     20      0
4 Dream   2008     NA     NA</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a></a><span class="co"># Then, use fill() to fill the missing penguin counts</span></span>
<span id="cb47-2"><a></a>penguins_complete <span class="sc">%&gt;%</span></span>
<span id="cb47-3"><a></a>  <span class="fu">fill</span>(Adelie, Gentoo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 4
  island  year Adelie Gentoo
  &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 Biscoe  2007     10     34
2 Biscoe  2008     18     46
3 Dream   2007     20      0
4 Dream   2008     20      0</code></pre>
</div>
</div>
</section>
<section id="introduction-to-joins-in-dplyr" class="slide level2">
<h2>Introduction to Joins in <code>dplyr</code></h2>
<div style="font-size: 0.8em;">
<ul>
<li>Joining datasets is a powerful tool for combining info. from multiple sources.</li>
<li>In R, <code>dplyr</code> provides several functions to perform different types of joins.</li>
<li>We’ll demonstrate joining <code>penguins_complete</code> (our penguin counts dataset) with <code>island_info</code> (dataset containing additional info. about the islands).</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a></a><span class="co"># Island information dataset</span></span>
<span id="cb49-2"><a></a>island_info <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb49-3"><a></a>  <span class="at">island =</span> <span class="fu">c</span>(<span class="st">"Biscoe"</span>, <span class="st">"Dream"</span>, <span class="st">"Torgersen"</span>),</span>
<span id="cb49-4"><a></a>  <span class="at">location =</span> <span class="fu">c</span>(<span class="st">"Antarctica"</span>, <span class="st">"Antarctica"</span>, <span class="st">"Antarctica"</span>),</span>
<span id="cb49-5"><a></a>  <span class="at">region =</span> <span class="fu">c</span>(<span class="st">"West"</span>, <span class="st">"East"</span>, <span class="st">"East"</span>)</span>
<span id="cb49-6"><a></a>)</span>
<span id="cb49-7"><a></a></span>
<span id="cb49-8"><a></a>island_info</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 3
  island    location   region
  &lt;chr&gt;     &lt;chr&gt;      &lt;chr&gt; 
1 Biscoe    Antarctica West  
2 Dream     Antarctica East  
3 Torgersen Antarctica East  </code></pre>
</div>
</div>
<ul>
<li>Notice that the <code>island_info</code> dataset includes an island, Torgersen, that is not in <code>penguins_complete</code>.</li>
</ul>
</div>
</section>
<section id="left-join-keep-all-rows-from-the-first-dataset" class="slide level2">
<h2>Left Join: Keep All Rows from the First Dataset</h2>
<ul>
<li>A <strong>left join</strong> keeps all rows from the <strong>first dataset</strong> (<code>penguins_complete</code>), and adds matching data from the second dataset (<code>island_info</code>).</li>
<li>So <strong>all rows from the first dataset</strong> (<code>penguins_complete</code>) will be preserved.</li>
<li>The datasets are joined by matching the <code>island</code> column, specified by the by argument.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a></a><span class="co"># Left join: combining penguins data with island info</span></span>
<span id="cb51-2"><a></a>penguins_with_info <span class="ot">&lt;-</span> penguins_complete <span class="sc">%&gt;%</span></span>
<span id="cb51-3"><a></a>  <span class="fu">left_join</span>(island_info, <span class="at">by =</span> <span class="st">"island"</span>)</span>
<span id="cb51-4"><a></a></span>
<span id="cb51-5"><a></a>penguins_with_info</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 6
  island  year Adelie Gentoo location   region
  &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt; 
1 Biscoe  2007     10     34 Antarctica West  
2 Biscoe  2008     18     46 Antarctica West  
3 Dream   2007     20      0 Antarctica East  
4 Dream   2008     NA     NA Antarctica East  </code></pre>
</div>
</div>
</section>
<section id="right-join-keep-all-rows-from-the-second-dataset" class="slide level2">
<h2>Right Join: Keep All Rows from the Second Dataset</h2>
<ul>
<li>A <strong>right join</strong> keeps all rows from the <strong>second dataset</strong> (<code>island_info</code>), and adds matching data from the first dataset (<code>penguins_complete</code>).</li>
<li>If a row in the second dataset doesn’t have a match in the first, then the columns from the first will be filled with NA.</li>
<li>We can see this for the <code>Torgersen</code> row from <code>island_info</code>…</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a></a><span class="co"># Right join: keep all rows from island_info</span></span>
<span id="cb53-2"><a></a>penguins_right_join <span class="ot">&lt;-</span> penguins_complete <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a></a>  <span class="fu">right_join</span>(island_info, <span class="at">by =</span> <span class="st">"island"</span>)</span>
<span id="cb53-4"><a></a></span>
<span id="cb53-5"><a></a>penguins_right_join</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 6
  island     year Adelie Gentoo location   region
  &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt; 
1 Biscoe     2007     10     34 Antarctica West  
2 Biscoe     2008     18     46 Antarctica West  
3 Dream      2007     20      0 Antarctica East  
4 Dream      2008     NA     NA Antarctica East  
5 Torgersen    NA     NA     NA Antarctica East  </code></pre>
</div>
</div>
</section>
<section id="inner-join-only-keeping-matching-rows" class="slide level2">
<h2>Inner Join: Only Keeping Matching Rows</h2>
<ul>
<li>An inner join will only keep rows where there is a match in both datasets.</li>
<li>So, if an island in <code>island_info</code> does not have a corresponding entry in <code>penguins_complete</code>, then that row will be excluded.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a></a><span class="co"># Inner join: only matching rows are kept</span></span>
<span id="cb55-2"><a></a>penguins_inner_join <span class="ot">&lt;-</span> penguins_complete <span class="sc">%&gt;%</span></span>
<span id="cb55-3"><a></a>  <span class="fu">inner_join</span>(island_info, <span class="at">by =</span> <span class="st">"island"</span>)</span>
<span id="cb55-4"><a></a></span>
<span id="cb55-5"><a></a>penguins_inner_join</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 6
  island  year Adelie Gentoo location   region
  &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt; 
1 Biscoe  2007     10     34 Antarctica West  
2 Biscoe  2008     18     46 Antarctica West  
3 Dream   2007     20      0 Antarctica East  
4 Dream   2008     NA     NA Antarctica East  </code></pre>
</div>
</div>
</section>
<section id="full-join-keeping-all-rows-from-both-datasets" class="slide level2">
<h2>Full Join: Keeping All Rows from Both Datasets</h2>
<ul>
<li>A full join will keep all rows from both datasets.</li>
<li>If an island in either dataset has no match in the other, the missing values will be filled with NA.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a></a><span class="co"># Full join: keep all rows from both datasets</span></span>
<span id="cb57-2"><a></a>penguins_full_join <span class="ot">&lt;-</span> penguins_complete <span class="sc">%&gt;%</span></span>
<span id="cb57-3"><a></a>  <span class="fu">full_join</span>(island_info, <span class="at">by =</span> <span class="st">"island"</span>)</span>
<span id="cb57-4"><a></a></span>
<span id="cb57-5"><a></a>penguins_full_join</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 6
  island     year Adelie Gentoo location   region
  &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt; 
1 Biscoe     2007     10     34 Antarctica West  
2 Biscoe     2008     18     46 Antarctica West  
3 Dream      2007     20      0 Antarctica East  
4 Dream      2008     NA     NA Antarctica East  
5 Torgersen    NA     NA     NA Antarctica East  </code></pre>
</div>
</div>
</section>
<section id="summary-of-the-four-join-functions" class="slide level2">
<h2>Summary of the Four Join Functions</h2>
<ul>
<li><strong>Left join:</strong> All rows from the left dataset and matching rows from the right dataset.</li>
<li><strong>Right join:</strong> All rows from the right dataset and matching rows from the left dataset.</li>
<li><strong>Inner join:</strong> Only matching rows from both datasets.</li>
<li><strong>Full join:</strong> All rows from both datasets, with NA where no match exists.</li>
</ul>
</section>
<section id="final-thoughts-on-joins" class="slide level2">
<h2>Final thoughts on joins</h2>
<ul>
<li>Joins are an essential part of data wrangling in R.</li>
<li>The choice of join depends on the analysis you need to perform:
<ul>
<li>Use <strong>left joins</strong> when you want to keep all data from the first dataset.</li>
<li>Use <strong>right joins</strong> when you want to keep all data from the second dataset.</li>
<li>Use <strong>inner joins</strong> when you’re only interested in matching rows.</li>
<li>Use <strong>full joins</strong> when you want to preserve all information from both datasets.</li>
</ul></li>
</ul>
</section>
<section id="goodbye-palmer-penguins" class="slide level2">
<h2>Goodbye palmer penguins</h2>
<!-- Penguins are great at `group_by()` – they always know how to stick together in a `summarize()`d form! 🐧 -->
<p><strong>What’s a penguin’s favorite tool?</strong></p>
<p>%&gt;% — to keep the fish moving, from one catch to the next! 🐧</p>
<div style="text-align: center;">
<p><img data-src="gfx/penguins_logo.png" style="width: 30%;"></p>
<p><small><a href="https://allisonhorst.github.io/palmerpenguins/">Logo from the palmerpenguins website</a></small></p>
</div>
</section></section>
<section>
<section id="epiverse-software-ecosystem" class="title-slide slide level1 center">
<h1>Epiverse software ecosystem</h1>

</section>
<section id="the-epiverse-ecosystem" class="slide level2">
<h2>The epiverse ecosystem</h2>
<p>Interworking, community-driven, packages for epi tracking &amp; forecasting.</p>

<!-- 1. Fetch data: epidatr, epidatpy, and other sources, 2. Explore, clean, transform & backtest 3. Pre-built forecasters, modular forecasting framework: epipredict -->
<img data-src="gfx/epiverse_packages_flow.jpg" style="width: 60%;" class="r-stretch"></section></section>
<section>
<section id="panel-and-versioned-data-in-the-epiverse" class="title-slide slide level1 center">
<h1>Panel and versioned data in the epiverse</h1>

</section>
<section id="what-is-panel-data" class="slide level2">
<h2>What is panel data?</h2>
<div style="font-size: 0.8em;">
<ul>
<li>Recall that <a href="https://en.wikipedia.org/wiki/Panel_data">panel data</a>, or longitudinal data, contain cross-sectional measurements of subjects over time.</li>
<li>Built-in example: <a href="https://cmu-delphi.github.io/epidatasets/reference/covid_case_death_rates.html"><code>covid_case_death_rates</code></a> dataset, which is a snapshot <strong>as of</strong> May 31, 2022 that contains daily state-wise measures of <code>case_rate</code> and <code>death_rate</code> for COVID-19 in 2021:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>An `epi_df` object, 6 x 4 with metadata:
* geo_type  = state
* time_type = day
* as_of     = 2022-05-31

# A tibble: 6 × 4
  geo_value time_value case_rate death_rate
* &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 ak        2020-12-31      35.9      0.158
2 al        2020-12-31      65.1      0.438
3 ar        2020-12-31      66.0      1.27 
4 as        2020-12-31       0        0    
5 az        2020-12-31      76.8      1.10 
6 ca        2020-12-31      96.0      0.751</code></pre>
</div>
</div>
<ul>
<li>How do we store &amp; work with such snapshots in the epiverse software ecosystem?</li>
</ul>
</div>
</section>
<section id="epi_df-snapshot-of-a-data-set" class="slide level2">
<h2><code>epi_df</code>: snapshot of a data set</h2>
<ul>
<li>You can convert panel data into an <code>epi_df</code> with the required <code>geo_value</code> and <code>time_value</code> columns</li>
</ul>
<p>Therefore, an <code>epi_df</code> is… * a tibble that requires columns <code>geo_value</code> and <code>time_value</code>. * arbitrary additional columns containing <span class="primary">measured values</span> * additional <span class="primary">keys</span> to index (<code>age_group</code>, <code>ethnicity</code>, etc.)</p>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong><code>epi_df</code></strong></p>
</div>
<div class="callout-content">
<p>Represents a <span class="primary">snapshot</span> that contains the most <span class="primary">up-to-date values</span> of the signal variables, <span class="primary">as of</span> a given time.</p>
</div>
</div>
</div>
</section>
<section id="epi_df-snapshot-of-a-data-set-1" class="slide level2">
<h2><code>epi_df</code>: snapshot of a data set</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a></a>edf <span class="ot">&lt;-</span> covid_case_death_rates</span>
<span id="cb60-2"><a></a>edf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An `epi_df` object, 20,496 x 4 with metadata:
* geo_type  = state
* time_type = day
* as_of     = 2022-05-31

# A tibble: 20,496 × 4
   geo_value time_value case_rate death_rate
 * &lt;chr&gt;     &lt;date&gt;         &lt;dbl&gt;      &lt;dbl&gt;
 1 ak        2020-12-31      35.9      0.158
 2 al        2020-12-31      65.1      0.438
 3 ar        2020-12-31      66.0      1.27 
 4 as        2020-12-31       0        0    
 5 az        2020-12-31      76.8      1.10 
 6 ca        2020-12-31      96.0      0.751
 7 co        2020-12-31      35.8      0.649
 8 ct        2020-12-31      52.1      0.819
 9 dc        2020-12-31      31.0      0.601
10 de        2020-12-31      64.3      0.912
# ℹ 20,486 more rows</code></pre>
</div>
</div>
</section>
<section id="sliding-examples-on-epi_df" class="slide level2">
<h2>Sliding examples on <code>epi_df</code></h2>
<h3 id="growth-rates">Growth rates</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a></a>edf <span class="ot">&lt;-</span> <span class="fu">filter</span>(edf, geo_value <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"ut"</span>, <span class="st">"ca"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb62-2"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">%&gt;%</span></span>
<span id="cb62-3"><a></a>  <span class="fu">mutate</span>(<span class="at">gr_cases =</span> <span class="fu">growth_rate</span>(time_value, case_rate, <span class="at">method =</span> <span class="st">"trend_filter"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<img data-src="day1-afternoon_files/figure-revealjs/unnamed-chunk-5-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="epi_archive-collection-of-epi_dfs" class="slide level2">
<h2><code>epi_archive</code>: collection of <code>epi_df</code>s</h2>
<ul>
<li>full version history of a data set</li>
<li>acts like a bunch of <code>epi_df</code>s — but stored <span class="primary">compactly</span></li>
<li>allows similar functionality as <code>epi_df</code> but using only <span class="primary">data that would have been available at the time</span></li>
</ul>
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Revisions</strong></p>
</div>
<div class="callout-content">
<p>Epidemiology data gets revised frequently.</p>
<ul>
<li>We may want to use the data <span class="primary">as it looked in the past</span></li>
<li>or we may want to examine <span class="primary">the history of revisions</span>.</li>
</ul>
</div>
</div>
</div>
</section>
<section id="epi_archive-collection-of-epi_dfs-1" class="slide level2">
<h2><code>epi_archive</code>: collection of <code>epi_df</code>s</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a></a>archive_cases_dv_subset</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>→ An `epi_archive` object, with metadata:
ℹ Min/max time values: 2020-06-01 / 2021-11-30
ℹ First/last version with update: 2020-06-02 / 2021-12-01
ℹ Versions end: 2021-12-01
ℹ A preview of the table (129638 rows x 5 columns):
Key: &lt;geo_value, time_value, version&gt;
        geo_value time_value    version percent_cli case_rate_7d_av
           &lt;char&gt;     &lt;Date&gt;     &lt;Date&gt;       &lt;num&gt;           &lt;num&gt;
     1:        ca 2020-06-01 2020-06-02          NA        6.628329
     2:        ca 2020-06-01 2020-06-06    2.140116        6.628329
     3:        ca 2020-06-01 2020-06-07    2.140116        6.628329
     4:        ca 2020-06-01 2020-06-08    2.140379        6.628329
     5:        ca 2020-06-01 2020-06-09    2.114430        6.628329
    ---                                                            
129634:        tx 2021-11-26 2021-11-29    1.858596        7.957657
129635:        tx 2021-11-27 2021-11-28          NA        7.174299
129636:        tx 2021-11-28 2021-11-29          NA        6.834681
129637:        tx 2021-11-29 2021-11-30          NA        8.841247
129638:        tx 2021-11-30 2021-12-01          NA        9.566218</code></pre>
</div>
</div>
</section>
<section id="revision-patterns" class="slide level2">
<h2>Revision patterns</h2>

<img data-src="day1-afternoon_files/figure-revealjs/unnamed-chunk-7-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="finalized-data" class="slide level2">
<h2>Finalized data</h2>
<div style="font-size: 0.9em;">
<ul>
<li>Counts are revised as time proceeds</li>
<li>Want to know the <span class="primary">final</span> value</li>
<li>Often not available until weeks/months later</li>
</ul>
</div>
<dl>
<dt>Forecasting</dt>
<dd>
At time <span class="math inline">\(t\)</span>, predict the final value for time <span class="math inline">\(t+h\)</span>, <span class="math inline">\(h &gt; 0\)</span>
</dd>
</dl>
<p><br></p>
<dl>
<dt>Backcasting</dt>
<dd>
At time <span class="math inline">\(t\)</span>, predict the final value for time <span class="math inline">\(t-h\)</span>, <span class="math inline">\(h &lt; 0\)</span>
</dd>
</dl>
<p><br></p>
<dl>
<dt>Nowcasting</dt>
<dd>
At time <span class="math inline">\(t\)</span>, predict the final value for time <span class="math inline">\(t\)</span>
</dd>
</dl>
</section></section>
<section>
<section id="basic-nowcasting-in-the-epiverse" class="title-slide slide level1 center">
<h1>Basic Nowcasting in the epiverse</h1>
<!-- predicting a finalized value from a provisional value and making predictions. -->
</section>
<section id="backfill-canadian-edition" class="slide level2">
<h2>Backfill Canadian edition</h2>
<ul>
<li><p>Every week the BC CDC releases COVID-19 hospitalization data.</p></li>
<li><p>Following week they revise the number upward (by ~25%) due to lagged reports.</p></li>
</ul>
<p><img data-src="gfx/bc_hosp_admissions_ex.jpg" style="width: 60%;"> <!-- Newest iteration of "backfill”, Canada edition. Every week the BC CDC releases hospitalization data. The following week they revise the number upward (by about 25%) due to lagging reports. Every single week, the newspaper says “hospitalizations have declined”. This week the BC CDC’s own report said “hospitalizations have declined”. The takeaway in the news is that hospitalizations ALWAYS fall from the previous week, but once backfilled, they’re rarely down --></p>
<ul>
<li><strong>Takeaway</strong>: Once the data is backfilled, hospitalizations rarely show a decline, challenging the common media narrative.</li>
</ul>
</section>
<section id="aside-on-nowcasting" class="slide level2">
<h2>Aside on Nowcasting</h2>
<ul>
<li><p>To some Epis, “nowcasting” can be equated with “estimate the time-varying instantaneous reproduction number, <span class="math inline">\(R_t\)</span>”</p></li>
<li><p>Example using the number of reported COVID-19 cases in British Columbia between January 2020 and April 15, 2023. <!-- This data is the number of reported COVID-19 cases in British Columbia between January 2020 and April 15, 2023. The values are.up-to-date as of August 2023. --></p></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="day1-afternoon_files/figure-revealjs/nowcasting-1.svg" class="quarto-figure quarto-figure-center" height="400"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><p>Group built <a href="https://dajmcdon.github.io/rtestim"><code>{rtestim}</code></a> doing for this nonparametrically.</p></li>
<li><p>We may come back to this later…</p></li>
</ul>
</section>
<section id="mathematical-setup" class="slide level2">
<h2>Mathematical setup</h2>
<ul>
<li><p>Suppose today is time <span class="math inline">\(t\)</span></p></li>
<li><p>Let <span class="math inline">\(y_i\)</span> denote a series of interest observed at times <span class="math inline">\(i=1,\ldots, t\)</span>.</p></li>
</ul>
<div class="callout callout-important no-icon callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<p><strong>Our goal</strong></p>
</div>
<div class="callout-content">
<ul>
<li>Produce a <strong>point nowcast</strong> for the finalized values of <span class="math inline">\(y_t\)</span>.</li>
<li>Accompany with time-varying prediction intervals</li>
</ul>
</div>
</div>
</div>
<ul>
<li><p>We also have access to <span class="math inline">\(p\)</span> other time series <span class="math inline">\(x_{ij},\; i=1,\ldots,t, \; j = 1,\ldots,p\)</span></p></li>
<li><p>All may be subject to revisions.</p></li>
</ul>
</section></section>
<section>
<section id="nowcasting-with-one-variable" class="title-slide slide level1 center">
<h1>Nowcasting with one variable</h1>

</section>
<section id="nowcasting-simple-ratio-ex-hhs-admissions" class="slide level2">
<h2>Nowcasting Simple Ratio Ex: HHS Admissions</h2>
<ul>
<li>In this example, we will demonstrate the concept of nowcasting using <strong>HHS influenza admissions data</strong> (All confirmed influenza hospital admissions occurring each day).</li>
<li>We will work with <strong>provisional</strong> data (real-time reports) and compare them to <strong>finalized</strong> data (final reports).</li>
<li>The goal is to estimate or <strong>nowcast</strong> the admissions values for days when only provisional data is available.</li>
</ul>
</section>
<section id="fetch-versioned-data" class="slide level2">
<h2>Fetch Versioned Data</h2>
<p>Let’s fetch versioned admissions data from the API (<code>pub_covidcast</code>) for Georgia (<code>geo_values = "ga"</code>) and the signal of interest (<code>confirmed_admissions_influenza_1d</code>) over February 2023.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a></a><span class="co"># Set the start and end dates for the analysis</span></span>
<span id="cb65-2"><a></a>start_time <span class="ot">=</span> <span class="fu">as.Date</span>(<span class="st">"2023-02-01"</span>)</span>
<span id="cb65-3"><a></a>end_time <span class="ot">=</span> <span class="fu">as.Date</span>(<span class="st">"2023-02-28"</span>)</span>
<span id="cb65-4"><a></a></span>
<span id="cb65-5"><a></a>admissions_archive <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb65-6"><a></a>  <span class="at">source =</span> <span class="st">"hhs"</span>,</span>
<span id="cb65-7"><a></a>  <span class="at">signals =</span> <span class="st">"confirmed_admissions_influenza_1d"</span>,</span>
<span id="cb65-8"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb65-9"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb65-10"><a></a>  <span class="at">geo_values =</span> <span class="st">"ga"</span>,</span>
<span id="cb65-11"><a></a>  <span class="at">time_values =</span> <span class="st">"*"</span>, <span class="co"># all time values</span></span>
<span id="cb65-12"><a></a>  <span class="at">issues =</span> <span class="st">"*"</span></span>
<span id="cb65-13"><a></a>) <span class="sc">%&gt;%</span> </span>
<span id="cb65-14"><a></a>  <span class="fu">filter</span>(time_value <span class="sc">&gt;=</span> start_time <span class="sc">&amp;</span> time_value <span class="sc">&lt;=</span> end_time) <span class="sc">%&gt;%</span></span>
<span id="cb65-15"><a></a>  <span class="fu">select</span>(geo_value, time_value, <span class="at">version =</span> issue, <span class="at">admissions =</span> value) <span class="sc">%&gt;%</span></span>
<span id="cb65-16"><a></a>  <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="latency-in-reporting---minimum-lag" class="slide level2">
<h2>Latency in Reporting - Minimum Lag</h2>
<ul>
<li>A quick inspection reveals that admissions are systematically 2 days latent (<strong>fixed lag</strong>).</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a></a>admissions_revision_inspect <span class="ot">=</span> admissions_archive<span class="sc">$</span>DT <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">version_time_diff =</span> version <span class="sc">-</span> time_value)</span>
<span id="cb66-2"><a></a>admissions_revision_inspect <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(time_value) <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="dv">1</span>) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
# Groups:   time_value [6]
  geo_value time_value version    admissions version_time_diff
  &lt;chr&gt;     &lt;date&gt;     &lt;date&gt;          &lt;dbl&gt; &lt;drtn&gt;           
1 ga        2023-02-01 2023-02-03          6 2 days           
2 ga        2023-02-02 2023-02-04         68 2 days           
3 ga        2023-02-03 2023-02-05         11 2 days           
4 ga        2023-02-04 2023-02-06         74 2 days           
5 ga        2023-02-05 2023-02-07          9 2 days           
6 ga        2023-02-06 2023-02-08          9 2 days           </code></pre>
</div>
</div>
<ul>
<li>Use <code>revision_summary()</code> from <code>epiprocess</code> to generate basic statistics about the revision behavior for the dataset.</li>
</ul>
</section>
<section id="latency-in-reporting---finalized-value-attainment" class="slide level2">
<h2>Latency in Reporting - Finalized Value Attainment</h2>
<div style="font-size: 0.8em;">
<ul>
<li>When is the <strong>finalized value</strong> first attained for each date? (we wouldn’t have access to this in real-time)</li>
<li>How fast are the final values attained &amp; what’s the pattern for these times, if any?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  geo_value time_value min_version diff   
  &lt;chr&gt;     &lt;date&gt;     &lt;date&gt;      &lt;drtn&gt; 
1 ga        2023-02-01 2023-02-26  25 days
2 ga        2023-02-02 2023-02-26  24 days
3 ga        2023-02-03 2023-02-26  23 days
4 ga        2023-02-04 2023-02-07   3 days
5 ga        2023-02-05 2023-02-26  21 days
6 ga        2023-02-06 2023-02-26  20 days</code></pre>
</div>
</div>
<p>And here’s a numerical summary:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   2.00    7.00   10.00   22.54   18.25  320.00 </code></pre>
</div>
</div>
<ul>
<li><strong>Conclusion</strong>: Varies somewhat. Even for this relatively small time period… Could be as low as 2 days or as high as 320 days. Yikes.</li>
<li>So if we were doing this in real-time, then we don’t necessarily have access to finalized data until much later in time.</li>
</ul>
</div>
</section>
<section id="comparison-of-final-vs.-multiple-revisions" class="slide level2">
<h2>Comparison of Final vs.&nbsp;Multiple Revisions</h2>
<p>This figure shows the finalized admissions in comparison to <strong>multiple revisions</strong> to see how the data changes over time:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="day1-afternoon_files/figure-revealjs/final-vs-revisions-plot-1.svg" class="quarto-figure quarto-figure-center" height="500"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="comparison-of-final-vs.-one-revision" class="slide level2">
<h2>Comparison of Final vs.&nbsp;One Revision</h2>
<div style="font-size: 0.8em;">
<p>The below figure compares the finalized admissions (in black) to <strong>one revision</strong> (in yellow) for Feb.&nbsp;16, 2023.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="day1-afternoon_files/figure-revealjs/one-revision-final-plot-1.svg" class="quarto-figure quarto-figure-center" height="400"></p>
</figure>
</div>
</div>
</div>
<p>The real-time data is biased downwards (systematically below the true value). That is, the signal tends to get scaled up with future revisions.</p>
</div>
</section>
<section id="calculate-one-ratio-provisional-vs.-finalized-data" class="slide level2">
<h2>Calculate One Ratio: Provisional vs.&nbsp;Finalized Data</h2>
<!-- * Let's start simple with computing one ratio. -->
<div style="font-size: 0.8em;">
<ul>
<li>Suppose that the day is February 16, 2023. Then, because the data is 2 days latent, we can compute the ratio between provisional and finalized data for <strong>February 14, 2023</strong>.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a></a>date_under_investigation <span class="ot">=</span> <span class="fu">as.Date</span>(<span class="st">"2023-02-14"</span>)</span>
<span id="cb70-2"><a></a></span>
<span id="cb70-3"><a></a><span class="co"># Load the finalized influenza data for GA</span></span>
<span id="cb70-4"><a></a>ga_finalized <span class="ot">&lt;-</span> admissions_latest <span class="sc">%&gt;%</span></span>
<span id="cb70-5"><a></a>  <span class="fu">filter</span>(time_value <span class="sc">==</span> date_under_investigation) <span class="sc">%&gt;%</span></span>
<span id="cb70-6"><a></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(admissions)</span>
<span id="cb70-7"><a></a></span>
<span id="cb70-8"><a></a><span class="co"># Look at the same data but "as_of" "2023-02-16"</span></span>
<span id="cb70-9"><a></a><span class="co"># Load the provisional case data (reported on "2023-02-16" but for "2022-02-14")</span></span>
<span id="cb70-10"><a></a>admissions_old <span class="ot">=</span> <span class="fu">epix_as_of</span>(admissions_archive, <span class="at">max_version =</span> (date_under_investigation <span class="sc">+</span> <span class="dv">2</span>L))</span>
<span id="cb70-11"><a></a></span>
<span id="cb70-12"><a></a>ga_provisional <span class="ot">&lt;-</span> admissions_old <span class="sc">%&gt;%</span></span>
<span id="cb70-13"><a></a>  <span class="fu">filter</span>(time_value <span class="sc">==</span> date_under_investigation) <span class="sc">%&gt;%</span></span>
<span id="cb70-14"><a></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(admissions)</span>
<span id="cb70-15"><a></a></span>
<span id="cb70-16"><a></a><span class="co"># Calculate ratio between provisional and finalized cases for 2022-02-14</span></span>
<span id="cb70-17"><a></a>ratio <span class="ot">&lt;-</span> ga_provisional<span class="sc">$</span>admissions <span class="sc">/</span> ga_finalized<span class="sc">$</span>admissions</span>
<span id="cb70-18"><a></a>ratio</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3571429</code></pre>
</div>
</div>
<p><strong>Conclusion</strong>: The real-time counts tend to be far below the finalized counts for this time (5 vs 14 here).</p>
<p><strong>Question</strong>: Can we generalize this over many days?</p>
</div>
</section>
<section id="calculating-the-ratio-over-multiple-dates" class="slide level2">
<h2>Calculating the Ratio over Multiple Dates</h2>
<ul>
<li>Let’s move from calculating the ratio for one day to multiple day with the goal to use it to nowcast for Feb.&nbsp;27, which has a <strong>provisional value</strong> of 6</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a></a>as_of_date <span class="ot">=</span> <span class="fu">as.Date</span>(<span class="st">"2023-03-01"</span>)</span>
<span id="cb72-2"><a></a></span>
<span id="cb72-3"><a></a>provisional_feb_27 <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(admissions_archive, <span class="at">max_version =</span> as_of_date) <span class="sc">%&gt;%</span></span>
<span id="cb72-4"><a></a>  <span class="fu">filter</span>(time_value <span class="sc">==</span> as_of_date <span class="sc">-</span> <span class="dv">2</span>L) <span class="sc">%&gt;%</span></span>
<span id="cb72-5"><a></a>  <span class="fu">pull</span>(admissions)</span>
<span id="cb72-6"><a></a>provisional_feb_27</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6</code></pre>
</div>
</div>
<p>and a <strong>finalized value</strong> of 8</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a></a>finalized_value <span class="ot">&lt;-</span> admissions_latest <span class="sc">%&gt;%</span></span>
<span id="cb74-2"><a></a>  <span class="fu">filter</span>(time_value <span class="sc">==</span> as_of_date <span class="sc">-</span> <span class="dv">2</span>L) <span class="sc">%&gt;%</span></span>
<span id="cb74-3"><a></a>  <span class="fu">pull</span>(admissions)</span>
<span id="cb74-4"><a></a>finalized_value</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8</code></pre>
</div>
</div>
</section>
<section id="calculating-the-ratio-over-multiple-dates-1" class="slide level2">
<h2>Calculating the Ratio over Multiple Dates</h2>
<p>First, let’s download the real-time number of cases for GA in February 2023, and compare them to their finalized version.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a></a>dates <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">as.Date</span>(<span class="st">"2023-02-01"</span>), <span class="fu">as.Date</span>(<span class="st">"2023-02-26"</span>), <span class="at">by =</span> <span class="st">"day"</span>)</span>
<span id="cb76-2"><a></a>admissions_real_time <span class="ot">&lt;-</span> <span class="cf">function</span>(date) {</span>
<span id="cb76-3"><a></a>  <span class="fu">epix_as_of</span>(admissions_archive, <span class="at">max_version =</span> (date <span class="sc">+</span> <span class="dv">2</span>L)) <span class="sc">%&gt;%</span></span>
<span id="cb76-4"><a></a>    <span class="fu">filter</span>(time_value <span class="sc">==</span> date)</span>
<span id="cb76-5"><a></a>}</span>
<span id="cb76-6"><a></a>admissions_ga_real_time <span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(dates, admissions_real_time)</span>
<span id="cb76-7"><a></a><span class="fu">head</span>(admissions_ga_real_time)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An `epi_df` object, 6 x 3 with metadata:
* geo_type  = state
* time_type = day
* as_of     = 2023-02-03

# A tibble: 6 × 3
  geo_value time_value admissions
* &lt;chr&gt;     &lt;date&gt;          &lt;dbl&gt;
1 ga        2023-02-01          6
2 ga        2023-02-02         68
3 ga        2023-02-03         11
4 ga        2023-02-04         74
5 ga        2023-02-05          9
6 ga        2023-02-06          9</code></pre>
</div>
</div>
</section>
<section id="calculating-the-ratio-over-multiple-dates-2" class="slide level2">
<h2>Calculating the Ratio over Multiple Dates</h2>
<p>Now, let’s plot the real-time vs the finalized number of admissions:</p>

<img data-src="day1-afternoon_files/figure-revealjs/real-time-vs-finalized-1.svg" class="quarto-figure quarto-figure-center r-stretch"><ul>
<li><strong>Takeaways</strong>: Aside from two major overreporting spikes near the beginning, the real-time counts are generally biased <strong>below</strong> the finalized counts.</li>
<li>Systematic underreporting begins around Feb.&nbsp;6, where the bias becomes more consistent.</li>
</ul>
</section>
<section id="calculating-the-ratio-over-multiple-dates-3" class="slide level2">
<h2>Calculating the Ratio over Multiple Dates</h2>
<p>Since we expect that this pattern in bias continues going forward (and the massive overreporting near the beginning may drown this out), we’ll start estimating the ratio from Feb.&nbsp;6 and beyond.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a></a>admissions_ga_real_time <span class="ot">=</span> admissions_ga_real_time <span class="sc">%&gt;%</span> <span class="fu">filter</span>(time_value <span class="sc">&gt;=</span> <span class="st">"2023-02-06"</span>)</span>
<span id="cb78-2"><a></a><span class="fu">head</span>(admissions_ga_real_time)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An `epi_df` object, 6 x 3 with metadata:
* geo_type  = state
* time_type = day
* as_of     = 2023-02-03

# A tibble: 6 × 3
  geo_value time_value admissions
* &lt;chr&gt;     &lt;date&gt;          &lt;dbl&gt;
1 ga        2023-02-06          9
2 ga        2023-02-07          5
3 ga        2023-02-08          4
4 ga        2023-02-09          8
5 ga        2023-02-10          6
6 ga        2023-02-11          7</code></pre>
</div>
</div>
</section>
<section id="realistic-limitation-of-nowcasting---finalized-data" class="slide level2">
<h2>Realistic Limitation of Nowcasting - Finalized Data</h2>
<ul>
<li>Recall that real-time access to finalized data is limited as finalized values can take months to report (e.g., Feb 10 finalized 320 days later).</li>
<li>To nowcast accurately, we must rely on the <strong>best available approximation of finalized data</strong> at the time of estimation (March 1).</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a></a>admissions_as_of_mar_1 <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(admissions_archive, <span class="at">max_version =</span> as_of_date) <span class="sc">%&gt;%</span></span>
<span id="cb80-2"><a></a>  <span class="fu">filter</span>(time_value <span class="sc">&gt;=</span> <span class="st">"2023-02-06"</span> <span class="sc">&amp;</span> time_value <span class="sc">&lt;=</span> <span class="st">"2023-02-26"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ratio-calculation-summary" class="slide level2">
<h2>Ratio calculation &amp; summary</h2>
<p>We then use the “finalized” and real-time counts to compute the mean ratio.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a></a>ratio_real_time_to_mar1 <span class="ot">&lt;-</span> admissions_ga_real_time<span class="sc">$</span>admissions <span class="sc">/</span> admissions_as_of_mar_1<span class="sc">$</span>admissions</span>
<span id="cb81-2"><a></a><span class="fu">summary</span>(ratio_real_time_to_mar1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.2667  0.4000  0.5000  0.7811  0.7778  3.3333 </code></pre>
</div>
</div>
<p>On average, the real-time counts are ~78.1% of the finalized counts.</p>

<img data-src="day1-afternoon_files/figure-revealjs/boxplot-ratio-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="nowcasting-on-march-1" class="slide level2">
<h2>Nowcasting on March 1</h2>
<p>Since the <strong>average ratio</strong> between real-time and finalized values is <strong>0.781</strong> (i.e., real-time counts are typically 78% of the finalized counts), then the nowcast is</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a></a><span class="co"># Now we can nowcast properly:</span></span>
<span id="cb83-2"><a></a>nowcast <span class="ot">&lt;-</span> provisional_feb_27 <span class="sc">*</span></span>
<span id="cb83-3"><a></a>  <span class="dv">1</span> <span class="sc">/</span> <span class="fu">mean</span>(ratio_real_time_to_mar1)</span>
<span id="cb83-4"><a></a>nowcast</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 7.681254</code></pre>
</div>
</div>
<p>So, this <strong>nowcast is 7.68</strong>, which is closer to the true finalized value of 8 than the provisional value of 6.</p>
</section>
<section id="summary-of-three-main-steps" class="slide level2">
<h2>Summary of three main steps</h2>
<p>So the main steps for this type of fixed lag nowcasting are…</p>
<ol type="1">
<li><p>Obtain the <strong>provisional value</strong> for the target.</p></li>
<li><p>Estimate the ratio using the <strong>real-time</strong> and <strong>“finalized”</strong> data (for all previous dates that follow a consistent pattern in reporting).</p></li>
<li><p>Profit.</p></li>
</ol>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Expand for the accompanying code</summary>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a></a><span class="co"># Today</span></span>
<span id="cb85-2"><a></a>as_of_date <span class="ot">=</span> <span class="fu">as.Date</span>(<span class="st">"2023-03-01"</span>)</span>
<span id="cb85-3"><a></a></span>
<span id="cb85-4"><a></a><span class="co"># 1. Obtain the provisional value</span></span>
<span id="cb85-5"><a></a>provisional_feb_27 <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(admissions_archive, <span class="at">max_version =</span> as_of_date) <span class="sc">%&gt;%</span></span>
<span id="cb85-6"><a></a>  <span class="fu">filter</span>(time_value <span class="sc">==</span> as_of_date <span class="sc">-</span> <span class="dv">2</span>L) <span class="sc">%&gt;%</span></span>
<span id="cb85-7"><a></a>  <span class="fu">pull</span>(admissions)</span>
<span id="cb85-8"><a></a>provisional_feb_27</span>
<span id="cb85-9"><a></a></span>
<span id="cb85-10"><a></a><span class="co"># 2. Estimate the ratio </span></span>
<span id="cb85-11"><a></a>admissions_ga_real_time <span class="ot">=</span> admissions_ga_real_time <span class="sc">%&gt;%</span> <span class="fu">filter</span>(time_value <span class="sc">&gt;=</span> <span class="st">"2023-02-06"</span>) <span class="co"># Real-time data</span></span>
<span id="cb85-12"><a></a>admissions_as_of_mar_1 <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(admissions_archive, <span class="at">max_version =</span> as_of_date) <span class="sc">%&gt;%</span> <span class="co"># "Finalized" data</span></span>
<span id="cb85-13"><a></a>  <span class="fu">filter</span>(time_value <span class="sc">&gt;=</span> <span class="st">"2023-02-06"</span> <span class="sc">&amp;</span> time_value <span class="sc">&lt;=</span> <span class="st">"2023-02-26"</span>)</span>
<span id="cb85-14"><a></a></span>
<span id="cb85-15"><a></a>mean_ratio <span class="ot">&lt;-</span> <span class="fu">mean</span>(admissions_ga_real_time<span class="sc">$</span>admissions <span class="sc">/</span> admissions_as_of_mar_1<span class="sc">$</span>admissions) <span class="co"># Mean ratio</span></span>
<span id="cb85-16"><a></a></span>
<span id="cb85-17"><a></a><span class="co"># 3. Profit.</span></span>
<span id="cb85-18"><a></a>nowcast <span class="ot">&lt;-</span> provisional_feb_27 <span class="sc">*</span></span>
<span id="cb85-19"><a></a>  <span class="dv">1</span> <span class="sc">/</span> <span class="fu">mean</span>(ratio_real_time_to_mar1)</span>
<span id="cb85-20"><a></a>nowcast</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="nowcasting-admissions-data-for-multiple-dates" class="slide level2">
<h2>Nowcasting Admissions Data for Multiple Dates</h2>
<ul>
<li><strong>Define Nowcast Function</strong>:
<ul>
<li><strong>Input</strong>: Takes in the dates to nowcast and the fixed lag</li>
<li><strong>Output</strong>: The nowcasted admissions based on the ratio of real-time to finalized data.</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a></a>nowcast_function <span class="ot">&lt;-</span> <span class="cf">function</span>(nowcast_date, fixed_lag) {</span>
<span id="cb86-2"><a></a>  as_of_date <span class="ot">=</span> nowcast_date <span class="sc">+</span> fixed_lag</span>
<span id="cb86-3"><a></a>  </span>
<span id="cb86-4"><a></a>  <span class="co"># 1. Obtain the provisional value for the target.</span></span>
<span id="cb86-5"><a></a>  provisional <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(admissions_archive, <span class="at">max_version =</span> as_of_date) <span class="sc">%&gt;%</span></span>
<span id="cb86-6"><a></a>  <span class="fu">filter</span>(time_value <span class="sc">==</span> as_of_date <span class="sc">-</span> <span class="dv">2</span>L) <span class="sc">%&gt;%</span></span>
<span id="cb86-7"><a></a>  <span class="fu">pull</span>(admissions)</span>
<span id="cb86-8"><a></a>  </span>
<span id="cb86-9"><a></a>  <span class="co">#2. Estimate the ratio multiplier using</span></span>
<span id="cb86-10"><a></a>  <span class="co"># real-time</span></span>
<span id="cb86-11"><a></a>  dates <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">as.Date</span>(<span class="st">"2023-02-06"</span>), nowcast_date <span class="sc">-</span> <span class="dv">1</span>, <span class="at">by =</span> <span class="st">"day"</span>)</span>
<span id="cb86-12"><a></a>  admissions_ga_real_time <span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(dates, admissions_real_time)</span>
<span id="cb86-13"><a></a>  </span>
<span id="cb86-14"><a></a>  <span class="co"># and "finalized" data</span></span>
<span id="cb86-15"><a></a>  finalized <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(admissions_archive, <span class="at">max_version =</span> as_of_date) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(time_value <span class="sc">&gt;=</span> <span class="st">"2023-02-06"</span> <span class="sc">&amp;</span> time_value <span class="sc">&lt;=</span> (nowcast_date <span class="sc">-</span> <span class="dv">1</span>)) </span>
<span id="cb86-16"><a></a></span>
<span id="cb86-17"><a></a>  ratios <span class="ot">&lt;-</span> admissions_ga_real_time<span class="sc">$</span>admissions <span class="sc">/</span> finalized<span class="sc">$</span>admissions</span>
<span id="cb86-18"><a></a>  </span>
<span id="cb86-19"><a></a>  <span class="co"># Remove infinite or NaN ratios (i.e., keep only finite values)</span></span>
<span id="cb86-20"><a></a>  mean_ratio <span class="ot">&lt;-</span> <span class="fu">mean</span>(ratios[<span class="fu">is.finite</span>(ratios)])</span>
<span id="cb86-21"><a></a></span>
<span id="cb86-22"><a></a>  <span class="co">#3. Profit.</span></span>
<span id="cb86-23"><a></a>  nowcast <span class="ot">&lt;-</span> provisional <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">/</span> mean_ratio)</span>
<span id="cb86-24"><a></a>  </span>
<span id="cb86-25"><a></a>  <span class="co"># Return a dataframe with the nowcast and date</span></span>
<span id="cb86-26"><a></a>  <span class="fu">tibble</span>(</span>
<span id="cb86-27"><a></a>    <span class="at">time_value =</span> nowcast_date,</span>
<span id="cb86-28"><a></a>    <span class="at">nowcast_admissions =</span> nowcast</span>
<span id="cb86-29"><a></a>  )</span>
<span id="cb86-30"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="map-nowcast-over-multiple-dates" class="slide level2">
<h2>Map Nowcast Over Multiple Dates</h2>
<ul>
<li>We can use <code>map2_df()</code> to apply the function to a series of dates (e.g., Feb 27–28).</li>
<li>Returns a <strong>dataframe</strong> with nowcasted results.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a></a><span class="co"># Apply Nowcast Function Over Multiple Dates</span></span>
<span id="cb87-2"><a></a>nowcast_dates <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">as.Date</span>(<span class="st">"2023-02-27"</span>), <span class="fu">as.Date</span>(<span class="st">"2023-02-28"</span>), <span class="at">by =</span> <span class="st">"day"</span>)</span>
<span id="cb87-3"><a></a>fixed_lag <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb87-4"><a></a>nowcast_results_df <span class="ot">&lt;-</span> <span class="fu">map2_df</span>(nowcast_dates, fixed_lag, nowcast_function)</span>
<span id="cb87-5"><a></a></span>
<span id="cb87-6"><a></a><span class="co"># View Nowcast Results</span></span>
<span id="cb87-7"><a></a>nowcast_results_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  time_value nowcast_admissions
  &lt;date&gt;                  &lt;dbl&gt;
1 2023-02-27               7.68
2 2023-02-28               8.49</code></pre>
</div>
</div>
</section>
<section id="visualize-nowcast-real-time-and-finalized-values" class="slide level2">
<h2>Visualize nowcast, real-time, and finalized values</h2>
<p>Finally, we can compare these nowcast results to the real-time and finalized values:</p>

<img data-src="day1-afternoon_files/figure-revealjs/nowcast-fun-plot-results-1.svg" class="quarto-figure quarto-figure-center r-stretch"><p>The real-time counts tend to be biased slightly below the finalized counts. Nowcasted values tend to provide a better approximation of the truth (at least for these dates).</p>
</section></section>
<section>
<section id="nowcasting-with-two-variables" class="title-slide slide level1 center">
<h1>Nowcasting with two variables</h1>

</section>
<section id="nowcasting-moving-from-one-signal-to-two" class="slide level2">
<h2>Nowcasting: Moving from One Signal to Two</h2>
<ul>
<li>Recall that in nowcasting the goal is to predict a finalized value from a provisional value.</li>
<li>Now, we’ll move from one signal to two, creating a simple linear model to nowcast.</li>
<li>Exogenous features (predictors) could include relevant signals, such as Google symptom search trends.</li>
<li>We will use these signals to nowcast hospital admissions related to influenza.</li>
</ul>
</section>
<section id="data-sources-google-search-trends-hospital-admissions" class="slide level2">
<h2>Data Sources: Google Search Trends &amp; Hospital Admissions</h2>
<ul>
<li><p><strong>Google Search Trends</strong>: Symptoms like cough, fever, and shortness of breath.</p>
<ul>
<li><strong>s01</strong>: Cough, Phlegm, Sputum, Upper respiratory tract infection</li>
<li><strong>s03</strong>: Fever, Hyperthermia, Chills, Shivering</li>
</ul></li>
<li><p><strong>Hospital Admissions</strong>: Data from the Department of Health &amp; Human Services on confirmed influenza admissions.</p></li>
<li><p>Using these, we will <strong>nowcast</strong> hospital admissions by using Google symptom search trends for GA from April to June 2023.</p></li>
<li><p>The first step is to fetch this data…</p></li>
</ul>
</section>
<section id="data-sources-google-search-trends-hospital-admissions-1" class="slide level2">
<h2>Data Sources: Google Search Trends &amp; Hospital Admissions</h2>
<div class="cell" data-layout-align="center">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a></a><span class="co"># Fetch Google symptom data for s01 and s03</span></span>
<span id="cb89-2"><a></a>x1 <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb89-3"><a></a>  <span class="at">source =</span> <span class="st">"google-symptoms"</span>,</span>
<span id="cb89-4"><a></a>  <span class="at">signals =</span> <span class="st">"s01_smoothed_search"</span>, </span>
<span id="cb89-5"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb89-6"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb89-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"ga"</span>,</span>
<span id="cb89-8"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">20230401</span>, <span class="dv">20230701</span>),</span>
<span id="cb89-9"><a></a>  <span class="at">issues =</span> <span class="st">"*"</span></span>
<span id="cb89-10"><a></a>) <span class="sc">%&gt;%</span></span>
<span id="cb89-11"><a></a>  <span class="fu">select</span>(geo_value, time_value, <span class="at">version =</span> issue, <span class="at">avg_search_vol_s01 =</span> value) <span class="sc">%&gt;%</span></span>
<span id="cb89-12"><a></a>  <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">FALSE</span>)</span>
<span id="cb89-13"><a></a></span>
<span id="cb89-14"><a></a>x2 <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb89-15"><a></a>  <span class="at">source =</span> <span class="st">"google-symptoms"</span>,</span>
<span id="cb89-16"><a></a>  <span class="at">signals =</span> <span class="st">"s03_smoothed_search"</span>,</span>
<span id="cb89-17"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb89-18"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb89-19"><a></a>  <span class="at">geo_values =</span> <span class="st">"ga"</span>,</span>
<span id="cb89-20"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">20230401</span>, <span class="dv">20230701</span>),</span>
<span id="cb89-21"><a></a>  <span class="at">issues =</span> <span class="st">"*"</span></span>
<span id="cb89-22"><a></a>) <span class="sc">%&gt;%</span></span>
<span id="cb89-23"><a></a>  <span class="fu">select</span>(geo_value, time_value, <span class="at">version =</span> issue, <span class="at">avg_search_vol_s03 =</span> value) <span class="sc">%&gt;%</span></span>
<span id="cb89-24"><a></a>  <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">FALSE</span>)</span>
<span id="cb89-25"><a></a></span>
<span id="cb89-26"><a></a><span class="co"># Fetch hospital admissions data</span></span>
<span id="cb89-27"><a></a>y1 <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb89-28"><a></a>  <span class="at">source =</span> <span class="st">"hhs"</span>,</span>
<span id="cb89-29"><a></a>  <span class="at">signals =</span> <span class="st">"confirmed_admissions_influenza_1d"</span>,</span>
<span id="cb89-30"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb89-31"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb89-32"><a></a>  <span class="at">geo_values =</span> <span class="st">"ga"</span>,</span>
<span id="cb89-33"><a></a>  <span class="at">time_values =</span> <span class="fu">epirange</span>(<span class="dv">20230401</span>, <span class="dv">20230701</span>),</span>
<span id="cb89-34"><a></a>  <span class="at">issues =</span> <span class="st">"*"</span></span>
<span id="cb89-35"><a></a>) <span class="sc">%&gt;%</span></span>
<span id="cb89-36"><a></a>  <span class="fu">select</span>(geo_value, time_value, <span class="at">version =</span> issue, <span class="at">admissions =</span> value) <span class="sc">%&gt;%</span></span>
<span id="cb89-37"><a></a>  <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="merging-the-archives" class="slide level2">
<h2>Merging the Archives</h2>
<ul>
<li>We’ll merge the symptom search trends (<code>x1</code>, <code>x2</code>) with hospital admissions data (<code>y</code>) using <code>epix_merge()</code> from <code>epiprocess</code>.</li>
<li>This allows us to match data by time and geography, &amp; fill any missing values with the most recent observation (LOCF).</li>
</ul>
</section>
<section id="linear-model-a-simple-approach-for-nowcasting" class="slide level2">
<h2>Linear Model: A Simple Approach for Nowcasting</h2>
<ul>
<li>Aside from ratios, one of the simplest approach to nowcasting is to use a <strong>linear regression model</strong>.</li>
<li>We model the relationship between provisional (predictor) data and response data.</li>
<li>This model helps us make <strong>predictions</strong> for the finalized data based on the current (provisional) signals.</li>
</ul>
</section>
<section id="linear-regression" class="slide level2">
<h2>Linear Regression</h2>
<ul>
<li><p><strong>Goal</strong>: Estimate the coefficients <span class="math inline">\(\beta_0\)</span> and <span class="math inline">\(\beta_1\)</span> that describe the relationship between the predictor <span class="math inline">\(x_i\)</span> and the outcome <span class="math inline">\(y_i\)</span>.</p></li>
<li><p><strong>Linear Model</strong>: The relationship is assumed to be:</p>
<p><span class="math display">\[y_i \approx \beta_0 + \beta_1 x_i \]</span></p>
<p>where <span class="math inline">\(\beta_0\)</span> is the intercept, <span class="math inline">\(\beta_1\)</span> is the slope.</p></li>
<li><p><strong>In R</strong>: Use <code>lm(y ~ x)</code> to estimate the coefficients, where <code>y</code> is the outcome variable and <code>x</code> is the predictor.</p></li>
</ul>
</section>
<section id="multiple-linear-regression" class="slide level2">
<h2>Multiple Linear Regression</h2>
<ul>
<li><p><strong>Goal</strong>: Estimate coefficients <span class="math inline">\(\beta_0, \beta_1, \dots, \beta_p\)</span> that describe the relationship between multiple predictors <span class="math inline">\(x_{i1}, x_{i2}, \dots, x_{ip}\)</span> and the outcome <span class="math inline">\(y_i\)</span>.</p></li>
<li><p><strong>Model</strong>: The relationship is assumed to be:</p>
<p><span class="math display">\[y_i \approx \beta_0 + \beta_1 x_{i1} + \beta_2 x_{i2} + \dots + \beta_p x_{ip}\]</span></p>
<p>where: <span class="math inline">\(\beta_0\)</span> is the intercept, <span class="math inline">\(\beta_1, \dots, \beta_p\)</span> are the coefficients.</p></li>
<li><p><strong>In R</strong>: Use <code>lm(y ~ x1 + x2 + ... + xp)</code> to estimate the coefficients, where <code>y</code> is the outcome and <code>x1, x2, ..., xp</code> are the predictors.</p></li>
</ul>
</section>
<section id="multiple-linear-regression-model" class="slide level2">
<h2>Multiple Linear Regression Model</h2>
<ul>
<li>A linear model is a good choice to describe the relationship between search trends and hospital admissions.</li>
<li>The model will include two predictors (s01 and s03).</li>
<li>We’ll use these two search trend signals to predict hospital admissions (response).</li>
</ul>
<!-- A linear regression model will be used to predict hospital admissions from search trends (s01 and s03). -->
</section>
<section id="multiple-linear-regression-model-1" class="slide level2">
<h2>Multiple Linear Regression Model</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a></a><span class="co"># Define the function for lm model fit and prediction</span></span>
<span id="cb90-2"><a></a>lm_mod_pred <span class="ot">&lt;-</span> <span class="cf">function</span>(data, gk, rtv, ...) {</span>
<span id="cb90-3"><a></a>  </span>
<span id="cb90-4"><a></a>  <span class="co"># Fit the linear model</span></span>
<span id="cb90-5"><a></a>  model <span class="ot">&lt;-</span> <span class="fu">lm</span>(admissions <span class="sc">~</span> avg_search_vol_s01 <span class="sc">+</span> avg_search_vol_s03, <span class="at">data =</span> data)</span>
<span id="cb90-6"><a></a>  </span>
<span id="cb90-7"><a></a>  <span class="co"># Make predictions</span></span>
<span id="cb90-8"><a></a>  predictions <span class="ot">=</span> <span class="fu">predict</span>(model,</span>
<span id="cb90-9"><a></a>                        <span class="at">newdata =</span> data <span class="sc">%&gt;%</span></span>
<span id="cb90-10"><a></a>                          <span class="co"># Use tidyr::fill() for LOCF if predictor data is incomplete </span></span>
<span id="cb90-11"><a></a>                          <span class="fu">fill</span>(avg_search_vol_s01, <span class="at">.direction =</span> <span class="st">"down"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb90-12"><a></a>                          <span class="fu">fill</span>(avg_search_vol_s03, <span class="at">.direction =</span> <span class="st">"down"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb90-13"><a></a>                          <span class="fu">filter</span>(time_value <span class="sc">==</span> <span class="fu">max</span>(time_value)),</span>
<span id="cb90-14"><a></a>                        <span class="at">interval =</span> <span class="st">"prediction"</span>, <span class="at">level =</span> <span class="fl">0.9</span></span>
<span id="cb90-15"><a></a>  )</span>
<span id="cb90-16"><a></a></span>
<span id="cb90-17"><a></a>  <span class="co"># Pull off true time value for comparison to target</span></span>
<span id="cb90-18"><a></a>  real_time_val <span class="ot">=</span> data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(time_value <span class="sc">==</span> <span class="fu">max</span>(time_value)) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(admissions)</span>
<span id="cb90-19"><a></a></span>
<span id="cb90-20"><a></a>  <span class="fu">return</span>(<span class="fu">data.frame</span>(predictions, <span class="at">actual_nowcast_date =</span> <span class="fu">max</span>(data<span class="sc">$</span>time_value), <span class="at">real_time_val =</span> real_time_val))</span>
<span id="cb90-21"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that this code is intentionally simple; while it can be refined to handle cases like negatives or other boundary conditions, we aim to avoid unnecessary complexity.</p>
</section>
<section id="nowcasting-with-epix_slide" class="slide level2">
<h2>Nowcasting with <code>epix_slide()</code></h2>
<ul>
<li>We will use <code>epix_slide()</code> to create a sliding window of training data.</li>
<li>The model will be trained on a 14-day window before the target date, and predictions will be made for the target date.</li>
<li>The beauty of this function is that it is version-aware - the sliding computation at any given reference time <strong>t</strong> is performed on data that would have been available as of <strong>t</strong> automatically.</li>
</ul>
</section>
<section id="nowcasting-with-epix_slide-1" class="slide level2">
<h2>Nowcasting with <code>epix_slide()</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a></a><span class="co"># Define the reference time points for nowcasting</span></span>
<span id="cb91-2"><a></a>targeted_nowcast_dates <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">as.Date</span>(<span class="st">"2023-04-15"</span>), <span class="fu">as.Date</span>(<span class="st">"2023-06-15"</span>), <span class="at">by =</span> <span class="st">"1 week"</span>)</span>
<span id="cb91-3"><a></a>ref_time_values <span class="ot">=</span> targeted_nowcast_dates <span class="sc">+</span> <span class="dv">2</span>  <span class="co"># Adjust for the systematic 2-day latency in the response</span></span>
<span id="cb91-4"><a></a><span class="co"># Determine this from revision_summary(y1, print_inform = TRUE) </span></span>
<span id="cb91-5"><a></a></span>
<span id="cb91-6"><a></a><span class="co"># Perform nowcasting using epix_slide</span></span>
<span id="cb91-7"><a></a>nowcast_res <span class="ot">&lt;-</span> archive <span class="sc">%&gt;%</span></span>
<span id="cb91-8"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">%&gt;%</span></span>
<span id="cb91-9"><a></a>  <span class="fu">epix_slide</span>(</span>
<span id="cb91-10"><a></a>    <span class="at">.f =</span> lm_mod_pred,</span>
<span id="cb91-11"><a></a>    <span class="at">.before =</span> <span class="dv">14</span>,  <span class="co"># 14-day training period</span></span>
<span id="cb91-12"><a></a>    <span class="at">.versions =</span> ref_time_values, </span>
<span id="cb91-13"><a></a>    <span class="at">.new_col_name =</span> <span class="st">"res"</span></span>
<span id="cb91-14"><a></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb91-15"><a></a>  <span class="fu">unnest</span>() <span class="sc">%&gt;%</span> <span class="co"># Nesting creates a list-column of data frames; unnesting flattens it back out into regular columns. </span></span>
<span id="cb91-16"><a></a>  <span class="fu">mutate</span>(<span class="at">targeted_nowcast_date =</span> targeted_nowcast_dates, <span class="at">time_value =</span> actual_nowcast_date) <span class="sc">%&gt;%</span></span>
<span id="cb91-17"><a></a>  <span class="fu">ungroup</span>()</span>
<span id="cb91-18"><a></a></span>
<span id="cb91-19"><a></a><span class="co"># View results</span></span>
<span id="cb91-20"><a></a><span class="fu">head</span>(nowcast_res, <span class="at">n=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 9
  geo_value version      fit    lwr   upr actual_nowcast_date real_time_val
  &lt;chr&gt;     &lt;date&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;date&gt;                      &lt;dbl&gt;
1 ga        2023-04-17  4.53 -0.331  9.39 2023-04-15                      4
2 ga        2023-04-24  7.58  1.64  13.5  2023-04-22                      4
# ℹ 2 more variables: targeted_nowcast_date &lt;date&gt;, time_value &lt;date&gt;</code></pre>
</div>
</div>
</section>
<section id="compare-with-the-actual-admissions" class="slide level2">
<h2>Compare with the Actual Admissions</h2>
<ul>
<li>After making predictions, we compare them to the actual hospital admissions.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a></a><span class="co"># Left join with latest results </span></span>
<span id="cb93-2"><a></a><span class="co"># Latest snapshot of data (with the latest/finalized admissions)</span></span>
<span id="cb93-3"><a></a>x_latest <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(archive, <span class="at">max_version =</span> <span class="fu">max</span>(archive<span class="sc">$</span>DT<span class="sc">$</span>version)) <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(avg_search_vol_s01, avg_search_vol_s03))</span>
<span id="cb93-4"><a></a></span>
<span id="cb93-5"><a></a>res <span class="ot">&lt;-</span> nowcast_res <span class="sc">%&gt;%</span> <span class="fu">left_join</span>(x_latest, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"geo_value"</span>, <span class="st">"time_value"</span>))</span>
<span id="cb93-6"><a></a><span class="fu">head</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 10
  geo_value version      fit    lwr   upr actual_nowcast_date real_time_val
  &lt;chr&gt;     &lt;date&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;date&gt;                      &lt;dbl&gt;
1 ga        2023-04-17  4.53 -0.331  9.39 2023-04-15                      4
2 ga        2023-04-24  7.58  1.64  13.5  2023-04-22                      4
3 ga        2023-05-01  5.89  1.88   9.91 2023-04-29                      5
4 ga        2023-05-08  4.57  0.450  8.70 2023-05-06                      6
5 ga        2023-05-15  7.95  4.91  11.0  2023-05-11                      8
6 ga        2023-05-22  2.84 -2.00   7.68 2023-05-20                      4
# ℹ 3 more variables: targeted_nowcast_date &lt;date&gt;, time_value &lt;date&gt;,
#   admissions &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="visualizing-the-nowcast-results" class="slide level2">
<h2>Visualizing the Nowcast Results</h2>
<div style="font-size: 0.9em;">
<p>We can then visualize the nowcast results alongside the true values using <code>ggplot2</code>:</p>
</div>

<img data-src="day1-afternoon_files/figure-revealjs/plot-lr-nowcast-res-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="key-takeaways-linear-regression-nowcasting-example" class="slide level2">
<h2>Key Takeaways: Linear Regression Nowcasting Example</h2>
<ul>
<li><strong>Provisional Data as Predictors</strong>: Using Google symptom search trends (<code>s01</code>, <code>s03</code>) to predict <strong>influenza hospital admissions</strong>.</li>
<li><strong>Simple Linear Model</strong>: A linear regression model captures the relationship between symptom searches and hospital admissions.</li>
<li><strong>Actionable Predictions</strong>: Nowcasts provide <strong>timely insights</strong> for hospital admissions, even before data is finalized.</li>
<li><strong>Sliding Window Approach</strong>: Predictions are based on <strong>data up to the current time</strong>, ensuring no future information influences the nowcast.</li>
<li><strong>Evaluation</strong>: Predictions are compared with actual admissions visually.</li>
</ul>
</section></section>
<section>
<section id="case-study---nowcasting-case-rates-using-cli" class="title-slide slide level1 center">
<h1>Case Study - Nowcasting case rates using %CLI</h1>

</section>
<section id="goal-of-this-case-study" class="slide level2">
<h2>Goal of this Case Study</h2>
<p><strong>Goal</strong>: Nowcast COVID-19 Case Rates using %CLI for Massachusetts.</p>
<ul>
<li>%CLI is contained in the Epidata API.</li>
<li>Case rates by specimen collection date are not. They are from the MA gov website.</li>
<li>Case rates in the API (JHU) are aligned by report date, not specimen collection/test date.</li>
<li>Aligning by <strong>test date</strong> allows us to avoid the more unpredictable delays introduced by the <strong>report date</strong>.</li>
</ul>
</section>
<section id="summary-of-main-steps" class="slide level2">
<h2>Summary of main steps</h2>
<div style="font-size: 0.8em;">
<p>The workflow is similar to the previous example where we nowcasted using two variables, only more involved. The main steps are…</p>
<ol type="1">
<li><p><strong>Fetch Data</strong>: Retrieve %CLI and COVID-19 case data (by specimen collection date) for MA.</p></li>
<li><p><strong>Merge Data</strong>: Align %CLI and case rate data using <code>epix_merge</code>, filling missing values via last observation carried forward (LOCF).</p></li>
<li><p><strong>Model &amp; Prediction</strong>: Fit a linear model to predict case rates based on %CLI, trained on a 30-day rolling window.</p></li>
<li><p><strong>Nowcast Execution</strong>: Use <code>epix_slide</code> to nowcast the case rates dynamically.</p></li>
<li><p><strong>Visualization</strong>: Plot actual vs.&nbsp;nowcasted case rates with confidence intervals to assess model accuracy.</p></li>
</ol>
<p>So the first step is to fetch the data…</p>
</div>
</section>
<section id="construct-an-epi_archive-from-scratch" class="slide level2">
<h2>Construct an <code>epi_archive</code> from scratch</h2>
<div style="font-size: 0.9em;">
<p><a href="&quot;https://www.mass.gov/info-details/archive-of-covid-19-cases-2020-2021&quot;">Here’s</a> the archive of COVID-19 case excel files from the MA gov website, which we’ll use to construct our own <code>epi_archive</code>. <br> Brief summary of this data:</p>
<ul>
<li><p><strong>First release</strong>: Raw .xlsx data was first released early January 2021.</p></li>
<li><p><strong>Change in reporting</strong>: Starting <strong>July 1, 2021</strong>, the dashboard shifted from <strong>7 days/week</strong> to <strong>5 days/week</strong> (Monday-Friday).</p></li>
<li><p><strong>Friday, Saturday, and Sunday</strong> data is included in the <strong>Monday</strong> dashboard.</p></li>
<li><p>When <strong>Monday</strong> is a holiday, the <strong>Friday through Monday</strong> data is posted on <strong>Tuesday</strong>.</p></li>
</ul>
</div>
</section>
<section id="construct-an-epi_archive-from-scratch-1" class="slide level2">
<h2>Construct an <code>epi_archive</code> from scratch</h2>
<div style="font-size: 0.8em;">
<ul>
<li><strong>Purpose</strong>: To create an <code>epi_archive</code> object for storing versioned time series data.</li>
<li><strong>Required Columns</strong>:
<ul>
<li><code>geo_value</code>: Geographic data (e.g., region).</li>
<li><code>time_value</code>: Time-related data (e.g., date, time).</li>
<li><code>version</code>: Tracks when the data was available (enables version-aware forecasting).</li>
</ul></li>
<li><strong>Constructor</strong>:
<ul>
<li><code>new_epi_archive()</code>: For manual construction of <code>epi_archive</code> (assumes validation of inputs).</li>
</ul></li>
<li><strong>Recommended Method</strong>:
<ul>
<li><code>as_epi_archive()</code>: Simplifies the creation process, ensuring proper formatting and validation. We’ll use this one when we download some data from the MA gov website!</li>
</ul></li>
</ul>
</div>
</section>
<section id="main-steps-to-construct-the-epi_archive" class="slide level2">
<h2>Main steps to construct the <code>epi_archive</code></h2>
<div style="font-size: 0.8em;">
<ol type="1">
<li><strong>Load necessary Libraries</strong>: Such as <code>tidyverse</code>, <code>readxl</code>, <code>epiprocess</code>.</li>
<li><strong>Process Each Date’s Data</strong>:
<ul>
<li>A function we’ll make (<code>process_covid_data</code>) downloads and processes daily COVID-19 data from the MA gov Excel files on their website.</li>
<li>The data is cleaned and formatted with columns: <code>geo_value</code>, <code>time_value</code>, <code>version</code>, and values.</li>
</ul></li>
<li><strong>Handle Missing Data</strong>: Checks if a date’s data is available (handle 404 errors).</li>
<li><strong>Create <code>epi_archive</code></strong>:
<ul>
<li>Combine processed data into a tibble.</li>
<li>Convert the tibble to an <code>epi_archive</code> object using <code>as_epi_archive()</code>.</li>
</ul></li>
</ol>
</div>
</section>
<section id="fetch-data---code-for-one-date" class="slide level2">
<h2>Fetch Data - Code for one date</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a></a><span class="co"># Load required libraries</span></span>
<span id="cb95-2"><a></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb95-3"><a></a><span class="fu">library</span>(readxl)</span>
<span id="cb95-4"><a></a><span class="fu">library</span>(httr)</span>
<span id="cb95-5"><a></a><span class="fu">library</span>(tibble)</span>
<span id="cb95-6"><a></a><span class="fu">library</span>(epiprocess)</span>
<span id="cb95-7"><a></a></span>
<span id="cb95-8"><a></a><span class="co"># Function to download and process each Excel file for a given date</span></span>
<span id="cb95-9"><a></a>process_covid_data <span class="ot">&lt;-</span> <span class="cf">function</span>(Date) {</span>
<span id="cb95-10"><a></a>  <span class="co"># Generate the URL for the given date</span></span>
<span id="cb95-11"><a></a>  url <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"https://www.mass.gov/doc/covid-19-raw-data-"</span>, <span class="fu">tolower</span>(<span class="fu">gsub</span>(<span class="st">"-0"</span>, <span class="st">"-"</span>, <span class="fu">format</span>(Date, <span class="st">"%B-%d-%Y"</span>))), <span class="st">"/download"</span>) </span>
<span id="cb95-12"><a></a>  <span class="co"># Applies gsub("-0", "-", ...) to replace any occurrence of -0 (such as in "April-01") with just - (resulting in "April-1").</span></span>
<span id="cb95-13"><a></a>  </span>
<span id="cb95-14"><a></a>  <span class="co"># Check if the URL exists (handle the 404 error by skipping that date)</span></span>
<span id="cb95-15"><a></a>  response <span class="ot">&lt;-</span> <span class="fu">GET</span>(url)</span>
<span id="cb95-16"><a></a>  </span>
<span id="cb95-17"><a></a>  <span class="cf">if</span> (<span class="fu">status_code</span>(response) <span class="sc">!=</span> <span class="dv">200</span>) {</span>
<span id="cb95-18"><a></a>    <span class="fu">return</span>(<span class="cn">NULL</span>)  <span class="co"># Skip if URL doesn't exist (404)</span></span>
<span id="cb95-19"><a></a>  }</span>
<span id="cb95-20"><a></a>  </span>
<span id="cb95-21"><a></a>  <span class="co"># Define the destination file path for the Excel file</span></span>
<span id="cb95-22"><a></a>  file_path <span class="ot">&lt;-</span> <span class="fu">tempfile</span>(<span class="at">fileext =</span> <span class="st">".xlsx"</span>)</span>
<span id="cb95-23"><a></a>  </span>
<span id="cb95-24"><a></a>  <span class="co"># Download the Excel file</span></span>
<span id="cb95-25"><a></a>  <span class="fu">GET</span>(url, <span class="fu">write_disk</span>(file_path, <span class="at">overwrite =</span> <span class="cn">TRUE</span>))</span>
<span id="cb95-26"><a></a>  </span>
<span id="cb95-27"><a></a>  <span class="co"># Read the relevant sheet from the Excel file</span></span>
<span id="cb95-28"><a></a>  data <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(file_path, <span class="at">sheet =</span> <span class="st">"CasesByDate (Test Date)"</span>)</span>
<span id="cb95-29"><a></a>  </span>
<span id="cb95-30"><a></a>  <span class="co"># Process the data: rename columns and convert Date</span></span>
<span id="cb95-31"><a></a>  data <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span></span>
<span id="cb95-32"><a></a>    <span class="fu">rename</span>(</span>
<span id="cb95-33"><a></a>      <span class="at">Date =</span> <span class="st">`</span><span class="at">Date</span><span class="st">`</span>,</span>
<span id="cb95-34"><a></a>      <span class="at">Positive_Total =</span> <span class="st">`</span><span class="at">Positive Total</span><span class="st">`</span>,</span>
<span id="cb95-35"><a></a>      <span class="at">Positive_New =</span> <span class="st">`</span><span class="at">Positive New</span><span class="st">`</span>,</span>
<span id="cb95-36"><a></a>      <span class="at">Case_Average_7day =</span> <span class="st">`</span><span class="at">7-day confirmed case average</span><span class="st">`</span></span>
<span id="cb95-37"><a></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb95-38"><a></a>    <span class="fu">mutate</span>(<span class="at">Date =</span> <span class="fu">as.Date</span>(Date))  <span class="co"># Convert to Date class</span></span>
<span id="cb95-39"><a></a>  </span>
<span id="cb95-40"><a></a>  <span class="co"># Create a tibble with the required columns for the epi_archive</span></span>
<span id="cb95-41"><a></a>  tib <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb95-42"><a></a>    <span class="at">geo_value =</span> <span class="st">"ma"</span>,  <span class="co"># Massachusetts (geo_value)</span></span>
<span id="cb95-43"><a></a>    <span class="at">time_value =</span> data<span class="sc">$</span>Date,  <span class="co"># Date from the data</span></span>
<span id="cb95-44"><a></a>    <span class="at">version =</span> Date,  <span class="co"># The extracted version date</span></span>
<span id="cb95-45"><a></a>    <span class="at">case_rate_7d_av =</span> data<span class="sc">$</span>Case_Average_7day  <span class="co"># 7-day average case value</span></span>
<span id="cb95-46"><a></a>  )</span>
<span id="cb95-47"><a></a>  </span>
<span id="cb95-48"><a></a>  <span class="fu">return</span>(tib)</span>
<span id="cb95-49"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fetch-data---code-breakdown" class="slide level2">
<h2>Fetch Data - Code breakdown</h2>
<div style="font-size: 0.8em;">
<ul>
<li>This purpose of this function is to download and process each Excel file as of a date.</li>
<li><strong>URL Creation</strong>: Dynamically generates the URL based on the date, removing leading zeros in day values (e.g., “April-01” → “April-1”).</li>
<li><strong>Check URL</strong>: Sends a request (<code>GET(url)</code>) and skips the date if the URL returns a non-200 status (e.g., 404 error).</li>
<li><strong>Download File</strong>: Saves the Excel file to a temporary path using <code>tempfile()</code> and <code>GET()</code>.</li>
<li><strong>Read Data</strong>: Loads the relevant sheet (“CasesByDate”) from the Excel file using <code>read_excel()</code>.</li>
<li><strong>Tibble Creation</strong>: Constructs a tibble with <code>geo_value</code>, <code>time_value</code>, <code>version</code>, and <code>case_rate_7d_av</code> to later compile into an <code>epi_archive</code> (you can think of an <code>epi_archive</code> as being a comprised of many <code>epi_df</code>s).</li>
</ul>
</div>
</section>
<section id="fetch-data---process-range-of-dates" class="slide level2">
<h2>Fetch Data - Process range of dates</h2>
<ul>
<li>Note that <code>process_covid_data()</code> works on one date at a time.</li>
<li>So now, we need a function that iterates over a date range and applies <code>process_covid_data()</code> to each date &amp; combines the resulting tibbles into an <code>epi_archive</code>.</li>
<li>We call this function <code>process_data_for_date_range()</code>…</li>
</ul>
</section>
<section id="fetch-data---process-range-of-dates-1" class="slide level2">
<h2>Fetch Data - Process range of dates</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a></a><span class="co"># Function to process data for a range of dates</span></span>
<span id="cb96-2"><a></a>process_data_for_date_range <span class="ot">&lt;-</span> <span class="cf">function</span>(start_date, end_date) {</span>
<span id="cb96-3"><a></a>  <span class="co"># Generate a sequence of dates between start_date and end_date</span></span>
<span id="cb96-4"><a></a>  date_sequence <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">as.Date</span>(start_date), <span class="fu">as.Date</span>(end_date), <span class="at">by =</span> <span class="st">"day"</span>)</span>
<span id="cb96-5"><a></a>  </span>
<span id="cb96-6"><a></a>  <span class="co"># Process data for each date and combine results</span></span>
<span id="cb96-7"><a></a>  covid_data_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(date_sequence, <span class="cf">function</span>(Date) {</span>
<span id="cb96-8"><a></a>    <span class="fu">process_covid_data</span>(Date)  <span class="co"># Skip over dates with no data (NULLs will be ignored)</span></span>
<span id="cb96-9"><a></a>  })</span>
<span id="cb96-10"><a></a>  </span>
<span id="cb96-11"><a></a>  <span class="co"># Combine all non-null individual tibbles into one data frame</span></span>
<span id="cb96-12"><a></a>  combined_data <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(covid_data_list[<span class="sc">!</span><span class="fu">sapply</span>(covid_data_list, is.null)])</span>
<span id="cb96-13"><a></a>  </span>
<span id="cb96-14"><a></a>  <span class="co"># Convert the combined data into an epi_archive object</span></span>
<span id="cb96-15"><a></a>  <span class="cf">if</span> (<span class="fu">nrow</span>(combined_data) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb96-16"><a></a>    epi_archive_data <span class="ot">&lt;-</span> combined_data <span class="sc">%&gt;%</span></span>
<span id="cb96-17"><a></a>      <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">FALSE</span>)</span>
<span id="cb96-18"><a></a>    </span>
<span id="cb96-19"><a></a>    <span class="fu">return</span>(epi_archive_data)</span>
<span id="cb96-20"><a></a>  } <span class="cf">else</span> {</span>
<span id="cb96-21"><a></a>    <span class="fu">message</span>(<span class="st">"No valid data available for the given date range."</span>)</span>
<span id="cb96-22"><a></a>    <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb96-23"><a></a>  }</span>
<span id="cb96-24"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fetch-data---code-breakdown-1" class="slide level2">
<h2>Fetch Data - Code breakdown</h2>
<p>Here’s a summary of what <code>process_data_for_date_range()</code> does: 1. <strong>Generates Date Range</strong>: Creates a sequence of dates between <code>start_date</code> and <code>end_date</code>.</p>
<ol start="2" type="1">
<li><p><strong>Processes Data</strong>: Applies the <code>process_covid_data</code> function to each date in the range (skip over dates with no data).</p></li>
<li><p><strong>Combines Results</strong>: Combines all valid (non-NULL) tibbles into one single data frame.</p></li>
<li><p><strong>Creates <code>epi_archive</code></strong>: Converts the combined data into an <code>epi_archive</code> object.</p></li>
</ol>
</section>
<section id="fetch-data---run-the-function-inspect-archive" class="slide level2">
<h2>Fetch Data - Run the function &amp; inspect archive</h2>
<div style="font-size: 0.75em;">
<ul>
<li>Now, let’s run the function &amp; inspect the resulting <code>epi_archive</code> of 7-day averaged COVID-19 case counts:</li>
<li>Expect building the archive to take a nontrivial amount of time (enough for a cup of coffee or to meditate on life).</li>
</ul>
</div>
<!-- To wonder why you chose Expect building the archive to take a nontrivial amount of time (enough for a cup of coffee or to wonder why you chose coding in the first place). -->
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a></a><span class="co"># Example usage: process data between Jan. 10, 2021, and Dec. 1, 2021</span></span>
<span id="cb97-2"><a></a>y <span class="ot">&lt;-</span> <span class="fu">process_data_for_date_range</span>(<span class="st">"2021-01-10"</span>, <span class="st">"2021-12-01"</span>)  <span class="co"># Raw .xlsx data is first released on Jan. 4, 2021</span></span>
<span id="cb97-3"><a></a>y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>→ An `epi_archive` object, with metadata:
ℹ Min/max time values: 2020-01-29 / 2021-11-30
ℹ First/last version with update: 2021-01-10 / 2021-12-01
ℹ Versions end: 2021-12-01
ℹ A preview of the table (135549 rows x 4 columns):
Key: &lt;geo_value, time_value, version&gt;
        geo_value time_value    version case_rate_7d_av
           &lt;char&gt;     &lt;Date&gt;     &lt;Date&gt;           &lt;num&gt;
     1:        ma 2020-01-29 2021-01-10              NA
     2:        ma 2020-01-29 2021-01-11              NA
     3:        ma 2020-01-29 2021-01-12              NA
     4:        ma 2020-01-29 2021-01-13              NA
     5:        ma 2020-01-29 2021-01-14              NA
    ---                                                
135545:        ma 2021-11-28 2021-11-30        2196.000
135546:        ma 2021-11-28 2021-12-01        2352.286
135547:        ma 2021-11-29 2021-11-30        1735.714
135548:        ma 2021-11-29 2021-12-01        2371.286
135549:        ma 2021-11-30 2021-12-01        1972.143</code></pre>
</div>
</div>
</section>
<section id="fetch-data---outpatient-doctors-visits-for-cli" class="slide level2">
<h2>Fetch Data - % Outpatient Doctors Visits for CLI</h2>
<div style="font-size: 0.9em;">
<ul>
<li>Now, from the Epidata API, let’s download the <a href="&quot;https://cmu-delphi.github.io/delphi-epidata/api/covidcast-signals/doctor-visits.html&quot;">estimated percentage of outpatient doctor visits</a> primarily for COVID-related symptoms, based on health system data.</li>
<li>Comes pre-smoothed in time using a Gaussian linear smoother</li>
<li>This will be the predictor when we nowcast COVID-19 Case Rates in MA.</li>
</ul>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a></a><span class="co"># Step 1: Fetch Versioned Data </span></span>
<span id="cb99-2"><a></a>x <span class="ot">&lt;-</span> <span class="fu">pub_covidcast</span>(</span>
<span id="cb99-3"><a></a>  <span class="at">source =</span> <span class="st">"doctor-visits"</span>,</span>
<span id="cb99-4"><a></a>  <span class="at">signals =</span> <span class="st">"smoothed_adj_cli"</span>,</span>
<span id="cb99-5"><a></a>  <span class="at">geo_type =</span> <span class="st">"state"</span>,</span>
<span id="cb99-6"><a></a>  <span class="at">time_type =</span> <span class="st">"day"</span>,</span>
<span id="cb99-7"><a></a>  <span class="at">geo_values =</span> <span class="st">"ma"</span>, <span class="co"># Just for MA to keep it simple (&amp; to go with the case data by test date for that state)</span></span>
<span id="cb99-8"><a></a>  <span class="at">time_value =</span> <span class="fu">epirange</span>(<span class="dv">20210301</span>, <span class="dv">20212101</span>), </span>
<span id="cb99-9"><a></a>  <span class="at">issues =</span> <span class="fu">epirange</span>(<span class="dv">20210301</span>, <span class="dv">20212101</span>)</span>
<span id="cb99-10"><a></a>) <span class="sc">%&gt;%</span></span>
<span id="cb99-11"><a></a>  <span class="fu">select</span>(geo_value, time_value,</span>
<span id="cb99-12"><a></a>         <span class="at">version =</span> issue,</span>
<span id="cb99-13"><a></a>         <span class="at">percent_cli =</span> value</span>
<span id="cb99-14"><a></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb99-15"><a></a>  <span class="fu">as_epi_archive</span>(<span class="at">compactify =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="use-epix_merge-to-merge-the-two-archives" class="slide level2">
<h2>Use <code>epix_merge()</code> to merge the two archives</h2>
<p>Now we’ll use <code>epix_merge()</code> to combine the two <code>epi_archive</code>s that share the same <code>geo_value</code> &amp; <code>time_value</code>.</p>
<!-- LOCF is used to ensure missing data is handled by filling forward. -->
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a></a>archive <span class="ot">&lt;-</span> <span class="fu">epix_merge</span>(</span>
<span id="cb100-2"><a></a>  x, y,</span>
<span id="cb100-3"><a></a>  <span class="at">sync =</span> <span class="st">"locf"</span>,</span>
<span id="cb100-4"><a></a>  <span class="at">compactify =</span> <span class="cn">FALSE</span></span>
<span id="cb100-5"><a></a>)</span>
<span id="cb100-6"><a></a>archive</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>→ An `epi_archive` object, with metadata:
ℹ Min/max time values: 2020-01-29 / 2021-12-13
ℹ First/last version with update: 2021-01-10 / 2021-12-17
ℹ Versions end: 2021-12-17
ℹ A preview of the table (139190 rows x 5 columns):
Key: &lt;geo_value, time_value, version&gt;
        geo_value time_value    version percent_cli case_rate_7d_av
           &lt;char&gt;     &lt;Date&gt;     &lt;Date&gt;       &lt;num&gt;           &lt;num&gt;
     1:        ma 2020-01-29 2021-01-10          NA              NA
     2:        ma 2020-01-29 2021-01-11          NA              NA
     3:        ma 2020-01-29 2021-01-12          NA              NA
     4:        ma 2020-01-29 2021-01-13          NA              NA
     5:        ma 2020-01-29 2021-01-14          NA              NA
    ---                                                            
139186:        ma 2021-12-11 2021-12-16    2.306966              NA
139187:        ma 2021-12-11 2021-12-17    2.281141              NA
139188:        ma 2021-12-12 2021-12-16    2.333759              NA
139189:        ma 2021-12-12 2021-12-17    2.369756              NA
139190:        ma 2021-12-13 2021-12-17    2.256551              NA</code></pre>
</div>
</div>
</section>
<section id="fitting-and-predicting-with-linear-model" class="slide level2">
<h2>Fitting and Predicting with Linear Model</h2>
<div style="font-size: 0.8em;">
<ul>
<li>Define <code>lm_mod_pred()</code>: A function that fits a linear model to forecast case rates based on the <code>percent_cli</code> predictor.</li>
<li>Use <code>predict()</code> with a 90% prediction interval.</li>
<li>Save the actual case rates to compare to the nowcasts later.</li>
</ul>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a></a>lm_mod_pred <span class="ot">&lt;-</span> <span class="cf">function</span>(data, ...) {</span>
<span id="cb102-2"><a></a>  <span class="co"># Linear model</span></span>
<span id="cb102-3"><a></a>  model <span class="ot">&lt;-</span> <span class="fu">lm</span>(case_rate_7d_av <span class="sc">~</span> percent_cli, <span class="at">data =</span> data)</span>
<span id="cb102-4"><a></a></span>
<span id="cb102-5"><a></a>  <span class="co"># Make predictions</span></span>
<span id="cb102-6"><a></a>  predictions <span class="ot">=</span> <span class="fu">predict</span>(model,</span>
<span id="cb102-7"><a></a>                        <span class="at">newdata =</span> data <span class="sc">%&gt;%</span></span>
<span id="cb102-8"><a></a>                          <span class="fu">fill</span>(percent_cli, <span class="at">.direction =</span> <span class="st">"down"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb102-9"><a></a>                          <span class="fu">filter</span>(time_value <span class="sc">==</span> <span class="fu">max</span>(time_value)),</span>
<span id="cb102-10"><a></a>                        <span class="at">interval =</span> <span class="st">"prediction"</span>, <span class="at">level =</span> <span class="fl">0.9</span>)</span>
<span id="cb102-11"><a></a>  </span>
<span id="cb102-12"><a></a>  <span class="co"># Pull off real-time value for later comparison to the nowcast value</span></span>
<span id="cb102-13"><a></a>  real_time_val <span class="ot">=</span> data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(time_value <span class="sc">==</span> <span class="fu">max</span>(time_value)) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(case_rate_7d_av)</span>
<span id="cb102-14"><a></a>  </span>
<span id="cb102-15"><a></a>  <span class="co"># Could clip predictions and bounds at 0</span></span>
<span id="cb102-16"><a></a>  <span class="fu">return</span>(<span class="fu">data.frame</span>(predictions, <span class="at">actual_nowcast_date =</span> <span class="fu">max</span>(data<span class="sc">$</span>time_value), <span class="at">real_time_val =</span> real_time_val)) </span>
<span id="cb102-17"><a></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="nowcasting-with-epix_slide-2" class="slide level2">
<h2>Nowcasting with <code>epix_slide()</code></h2>
<ul>
<li><strong>Specify targets</strong>: Define the target dates for nowcasting (e.g., 1st of each month) &amp; adjust training data to include the lag for the latent case data.</li>
<li><strong>Sliding window</strong>: Use <code>epix_slide()</code> to apply the linear model across a sliding window of data for each region.</li>
<li><strong>Training-test split</strong>: Use the last 30 days of data to train and predict case rates for each target nowcast date.</li>
</ul>
</section>
<section id="nowcasting-with-epix_slide-3" class="slide level2">
<h2>Nowcasting with <code>epix_slide()</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a></a><span class="co"># Define the reference time points (to give the training/test split)</span></span>
<span id="cb103-2"><a></a>targeted_nowcast_dates <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">as.Date</span>(<span class="st">"2021-04-01"</span>), <span class="fu">as.Date</span>(<span class="st">"2021-11-01"</span>), <span class="at">by =</span> <span class="st">"1 month"</span>) </span>
<span id="cb103-3"><a></a>ref_time_values <span class="ot">=</span> targeted_nowcast_dates <span class="sc">+</span> <span class="dv">1</span> <span class="co"># + 1 because the case data is 1 day latent. </span></span>
<span id="cb103-4"><a></a><span class="co"># Determine this from revision_summary(y)</span></span>
<span id="cb103-5"><a></a></span>
<span id="cb103-6"><a></a><span class="co"># Use epix_slide to perform the nowcasting with a training-test split</span></span>
<span id="cb103-7"><a></a>nowcast_res <span class="ot">&lt;-</span> archive <span class="sc">%&gt;%</span></span>
<span id="cb103-8"><a></a>  <span class="fu">group_by</span>(geo_value) <span class="sc">%&gt;%</span></span>
<span id="cb103-9"><a></a>  <span class="fu">epix_slide</span>(</span>
<span id="cb103-10"><a></a>    <span class="at">.f =</span> lm_mod_pred,  <span class="co"># Pass the function defined above</span></span>
<span id="cb103-11"><a></a>    <span class="at">.before =</span> <span class="dv">30</span>,   <span class="co"># Training period of 30 days</span></span>
<span id="cb103-12"><a></a>    <span class="at">.versions =</span> ref_time_values, <span class="co"># Determines the day where training data goes up to (not inclusive)</span></span>
<span id="cb103-13"><a></a>    <span class="at">.new_col_name =</span> <span class="st">"res"</span></span>
<span id="cb103-14"><a></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb103-15"><a></a>  <span class="fu">unnest</span>() <span class="sc">%&gt;%</span></span>
<span id="cb103-16"><a></a>  <span class="fu">mutate</span>(<span class="at">targeted_nowcast_date =</span> targeted_nowcast_dates,</span>
<span id="cb103-17"><a></a>         <span class="at">time_value =</span> actual_nowcast_date)</span>
<span id="cb103-18"><a></a></span>
<span id="cb103-19"><a></a><span class="co"># Take a peek at the results</span></span>
<span id="cb103-20"><a></a><span class="fu">head</span>(nowcast_res, <span class="at">n =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 9
# Groups:   geo_value [1]
  geo_value version      fit   lwr   upr actual_nowcast_date real_time_val
  &lt;chr&gt;     &lt;date&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;date&gt;                      &lt;dbl&gt;
1 ma        2021-04-02 2114. 1975. 2254. 2021-04-01                  1556.
# ℹ 2 more variables: targeted_nowcast_date &lt;date&gt;, time_value &lt;date&gt;</code></pre>
</div>
</div>
</section>
<section id="visualizing-nowcasts-vs.-actual-values" class="slide level2">
<h2>Visualizing Nowcasts vs.&nbsp;Actual Values</h2>
<p>Merge the nowcast results with the latest data for more direct comparison:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a></a>x_latest <span class="ot">&lt;-</span> <span class="fu">epix_as_of</span>(archive, <span class="at">max_version =</span> <span class="fu">max</span>(archive<span class="sc">$</span>DT<span class="sc">$</span>version)) <span class="sc">%&gt;%</span></span>
<span id="cb105-2"><a></a>  <span class="fu">select</span>(<span class="sc">-</span>percent_cli) </span>
<span id="cb105-3"><a></a></span>
<span id="cb105-4"><a></a>res <span class="ot">&lt;-</span> nowcast_res <span class="sc">%&gt;%</span> <span class="fu">left_join</span>(x_latest, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"geo_value"</span>, <span class="st">"time_value"</span>))</span>
<span id="cb105-5"><a></a></span>
<span id="cb105-6"><a></a>res</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 × 10
# Groups:   geo_value [1]
  geo_value version       fit    lwr   upr actual_nowcast_date real_time_val
  &lt;chr&gt;     &lt;date&gt;      &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;date&gt;                      &lt;dbl&gt;
1 ma        2021-04-02 2114.  1975.  2254. 2021-04-01                  1556.
2 ma        2021-05-02 1083.   851.  1316. 2021-05-01                   869.
3 ma        2021-06-02  353.   164.   541. 2021-06-01                   117.
4 ma        2021-07-02   57.1   11.3  103. 2021-07-01                    59 
5 ma        2021-08-02  513.   284.   742. 2021-08-01                   572.
6 ma        2021-09-02 1207.   888.  1527. 2021-09-01                  1099.
7 ma        2021-10-02 1575.  1357.  1793. 2021-09-30                  1069 
8 ma        2021-11-02 1299.  1257.  1340. 2021-11-01                   891.
# ℹ 3 more variables: targeted_nowcast_date &lt;date&gt;, time_value &lt;date&gt;,
#   case_rate_7d_av &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="visualizing-nowcasts-vs.-actual-values-1" class="slide level2">
<h2>Visualizing Nowcasts vs.&nbsp;Actual Values</h2>
<div style="font-size: 0.9em;">
<p>Finally, plot the predictions &amp; real-time values on top of latest COVID-19 case rates using <code>ggplot2</code>:</p>
</div>

<img data-src="day1-afternoon_files/figure-revealjs/unnamed-chunk-21-1.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="takeaways" class="slide level2">
<h2>Takeaways</h2>
<p><strong>Goal</strong>: Predict COVID-19 case rates using %CLI, overcoming delays in report data.</p>
<p>Main Steps:</p>
<ol type="1">
<li><p><strong>Fetch Data</strong>: Collect case rates and %CLI data.</p></li>
<li><p><strong>Merge Data</strong>: Align datasets with epix_merge() and fill missing values.</p></li>
<li><p><strong>Model</strong>: Fit a linear model to predict case rates.</p></li>
<li><p><strong>Nowcast</strong>: Apply dynamic forecasting with epix_slide().</p></li>
<li><p><strong>Visualize</strong>: Plot nowcasts vs.&nbsp;actual case rates with confidence intervals.</p></li>
</ol>
<p>Overall, nowcasting, based on the linear model, provided a closer approximation of true case rates compared to the real-time values.</p>
</section>
<section id="final-slide" class="slide level2 smaller">
<h2>Final slide</h2>
<h3 id="thanks">Thanks:</h3>
<ul>
<li>The whole <a href="https://delphi.cmu.edu/about/team/">CMU Delphi Team</a> (across many institutions)</li>
<li>Optum/UnitedHealthcare, Change Healthcare.</li>
<li>Google, Facebook, Amazon Web Services.</li>
<li>Quidel, SafeGraph, Qualtrics.</li>
<li>Centers for Disease Control and Prevention.</li>
<li>Council of State and Territorial Epidemiologists</li>
</ul>
<div data-layout-row="1" data-fig-align="center">
<p><img data-src="gfx/delphi.jpg" height="100"> <img data-src="gfx/berkeley.jpg" height="100"> <img data-src="gfx/cmu.jpg" height="100"> <img data-src="gfx/ubc.jpg" width="250"> <img data-src="gfx/stanford.jpg" width="250"></p>
</div>

<div class="quarto-auto-generated-content">
<p><img src="gfx/delphi.jpg" class="slide-logo"></p>
<div class="footer footer-default">
<p>Explore, clean &amp; transform data — cmu-delphi/insightnet-workshop-2024</p>
</div>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: false,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        function fireSlideChanged(previousSlide, currentSlide) {

          // dispatch for htmlwidgets
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for reveal
        if (window.Reveal) {
          window.Reveal.addEventListener("slidechanged", function(event) {
            fireSlideChanged(event.previousSlide, event.currentSlide);
          });
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        // For code content inside modals, clipBoardJS needs to be initialized with a container option
        // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp("https:\/\/cmu-delphi\.github\.io\/insightnet-workshop-2024\/");
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
              // target, if specified
              link.setAttribute("target", "_blank");
              if (link.getAttribute("rel") === null) {
                link.setAttribute("rel", "noopener");
              }
              // default icon
              link.classList.add("external");
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>